{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2fb41230-099e-4695-9bb7-902b8b0348c9",
   "metadata": {},
   "source": [
    "# Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40899a15-d9d7-4bc5-8907-595b2ddf40bf",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8b1cd363-d4f2-4805-91b9-5ff4cbf5e83b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD LIBRARIES\n",
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(future)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(presto)\n",
    "library(cowplot)\n",
    "library(tictoc)\n",
    "\n",
    "library(enrichR)\n",
    "library(GPTCelltype)\n",
    "library(openai)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c0ef4f3e-6c74-4181-bdc5-43316dd1ed54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load\n",
    "load.data <- function(\n",
    "    file_name,\n",
    "    output = F,\n",
    "    reduced.output = T \n",
    ") {\n",
    "    if (output | reduced.output) {\n",
    "        print(paste(\"Loading data for time point:\", file_name))\n",
    "    }\n",
    "\n",
    "    # Load the data\n",
    "    sc_data <- Read10X(data.dir = paste(path_to_data, \"expression_\", file_name, sep = \"\"), gene.column = 1)\n",
    "\n",
    "    # Create Seurat object\n",
    "    sc_data <- CreateSeuratObject(counts = sc_data, min.cells = 3, min.features = 500, project = file_name, names.delim = \"-\", names.field = 2)\n",
    "\n",
    "    # Normalize the data\n",
    "    sc_data <- NormalizeData(sc_data, normalization.method = \"LogNormalize\", scale.factor = 1e6, verbose = output)\n",
    "\n",
    "    # Find variable features\n",
    "    sc_data <- FindVariableFeatures(sc_data, selection.method = \"mvp\", nfeatures = 2000, verbose = output)\n",
    "\n",
    "    # Scale the data\n",
    "    sc_data <- ScaleData(sc_data, verbose = output)\n",
    "\n",
    "    return(sc_data)\n",
    "}\n",
    "\n",
    "# Cluster\n",
    "PCA.cluster <- function(\n",
    "    data = sc_data_scaled, \n",
    "    file_name = timepoints[time_point], \n",
    "    res = 1, \n",
    "    n_dim = 40, \n",
    "    save = F,\n",
    "    output = F,\n",
    "    reduced.output = T    \n",
    ") {  \n",
    "    if (output | reduced.output) {\n",
    "        print(paste(\"Running PCA and clustering for time point:\", file_name))\n",
    "        print(paste(\"- Resolution:\", res))\n",
    "        print(paste(\"- Dimensions:\", n_dim))\n",
    "    }\n",
    "        \n",
    "    # PCA\n",
    "    data <- RunPCA(data, npcs = n_dim, verbose = output)\n",
    "    #print(ElbowPlot(object = data, ndims = 50))\n",
    "\n",
    "    # Cluster the cells\n",
    "    data <- FindNeighbors(data, dims = 1:n_dim, verbose = output)\n",
    "    data <- FindClusters(data, resolution = res, verbose = output)\n",
    "    \n",
    "    #print(table(Idents(data)))\n",
    "\n",
    "    # Save the PCA plot\n",
    "    if (save) {\n",
    "        name_new_dir <- paste(name_new_dir_partial, \"/\", file_name, \"/cluster\", param, sep=\"\")\n",
    "        if (!dir.exists(name_new_dir)) {dir.create(name_new_dir)} \n",
    "    \n",
    "        print(paste(\"Saving PCA for time point\", file_name, \"in\", name_new_dir))\n",
    "        save(data, file = paste(name_new_dir, \"/PCA_res_\", res, \"_dim_\", n_dim, \"_\", file_name, \".Robj\", sep=\"\"))\n",
    "    }\n",
    "    return(data)\n",
    "}\n",
    "\n",
    "# FIND ALL MARKERS\n",
    "cluster.markers <- function(\n",
    "    data, \n",
    "    file_name = timepoints[time_point],\n",
    "    output = F,\n",
    "    reduced.output = T\n",
    ") {\n",
    "    if (output | reduced.output) {\n",
    "        print(paste(\"Finding all markers for time point:\", file_name))\n",
    "    }\n",
    "\n",
    "    # Find all markers for every cluster compared to all remaining cells\n",
    "    markers <- FindAllMarkers(data,\n",
    "                              only.pos = TRUE,   # Considera solo i marker espressi positivamente\n",
    "                              min.pct = 0.25,    # Percentuale minima di espressione nelle cellule del cluster\n",
    "                              logfc.threshold = 0.25,  # Soglia minima di LogFC\n",
    "                              verbose = output)\n",
    "        \n",
    "    return(markers)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8ba24db1-07cf-4523-8411-a788ed62a637",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Loading data for time point: 23days\"\n",
      "[1] \"Running PCA and clustering for time point: 23days\"\n",
      "[1] \"- Resolution: 0.5\"\n",
      "[1] \"- Dimensions: 20\"\n",
      "[1] \"Finding all markers for time point: 23days\"\n",
      "Setup Clusterization and Markers: 95.965 sec elapsed\n"
     ]
    }
   ],
   "source": [
    "tic(\"Setup Clusterization and Markers\")\n",
    "\n",
    "# SET UP NAMES\n",
    "timepoints <- c(\"23days\", \"1month\", \"1.5month\", \"2month\", \"3month\", \"4month\", \"5month\", \"6month\")\n",
    "housekeeping_genes <- c(\"ACTB\", \"DLG4\")\n",
    "genes_of_interest <- c(\"SRCIN1\", \"KIAA1217\", \"CIT\")\n",
    "path_to_data <- \"/sharedFolder/Data/\"\n",
    "\n",
    "dir_annotations <- \"Results/Annotations\"\n",
    "if (!dir.exists(dir_annotations)) {dir.create(dir_annotations)}\n",
    "\n",
    "f_name <- \"23days\"\n",
    "clustering_resolution <- 0.5\n",
    "n_of_dimesnions <- 20\n",
    "\n",
    "# Load data\n",
    "sc_data_scaled <- load.data(file_name = f_name, reduced.output = T)\n",
    "\n",
    "# Clusterize\n",
    "sc_data <- PCA.cluster(\n",
    "    data = sc_data_scaled, \n",
    "    file_name = f_name, \n",
    "    res = clustering_resolution, \n",
    "    n_dim = n_of_dimesnions,\n",
    "    reduced.output = T\n",
    ")\n",
    "\n",
    "# Find Markers\n",
    "cluster_markers <- cluster.markers(data = sc_data, file_name = f_name, output = F, reduced.output = T)\n",
    "\n",
    "toc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "440bac5c-c718-4cd4-a724-6742c4311a0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15 \n",
      "6742 2736 2618 2448 2042 2037 1941 1810 1800 1792 1203 1025  727  508  116  106 \n",
      "  16 \n",
      "  85 \n"
     ]
    }
   ],
   "source": [
    "print(table(Idents(sc_data)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c0a57868-5337-4169-99fe-e3e5470abf9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "17"
      ],
      "text/latex": [
       "17"
      ],
      "text/markdown": [
       "17"
      ],
      "text/plain": [
       "[1] 17"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(table(Idents(sc_data)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77f49aa5-7717-4bf3-8a9f-5005a4e185f5",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8148a21-f233-419e-9f0f-6ddf4d4b07af",
   "metadata": {},
   "source": [
    "### Base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8ffb3fe-f40d-4421-8e51-f1933d43dfb8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     p_val             avg_log2FC          pct.1            pct.2       \n",
       " Min.   :0.0000000   Min.   : 0.2500   Min.   :0.1670   Min.   :0.0000  \n",
       " 1st Qu.:0.0000000   1st Qu.: 0.3978   1st Qu.:0.3250   1st Qu.:0.1770  \n",
       " Median :0.0000000   Median : 0.6083   Median :0.4350   Median :0.2760  \n",
       " Mean   :0.0001703   Mean   : 0.9779   Mean   :0.5015   Mean   :0.3412  \n",
       " 3rd Qu.:0.0000000   3rd Qu.: 1.0635   3rd Qu.:0.6370   3rd Qu.:0.4470  \n",
       " Max.   :0.0099990   Max.   :13.8465   Max.   :1.0000   Max.   :0.9990  \n",
       "                                                                        \n",
       "   p_val_adj            cluster         gene          \n",
       " Min.   :0.0000000   7      :1339   Length:13730      \n",
       " 1st Qu.:0.0000000   9      :1232   Class :character  \n",
       " Median :0.0000000   8      :1209   Mode  :character  \n",
       " Mean   :0.1003849   15     :1197                     \n",
       " 3rd Qu.:0.0000002   16     :1140                     \n",
       " Max.   :1.0000000   12     : 965                     \n",
       "                     (Other):6648                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>DCT</th><td>0</td><td>1.780295</td><td>0.383</td><td>0.161</td><td>0</td><td>0</td><td>DCT      </td></tr>\n",
       "\t<tr><th scope=row>AMBN</th><td>0</td><td>1.767947</td><td>0.431</td><td>0.175</td><td>0</td><td>0</td><td>AMBN     </td></tr>\n",
       "\t<tr><th scope=row>HEY1</th><td>0</td><td>1.708512</td><td>0.464</td><td>0.202</td><td>0</td><td>0</td><td>HEY1     </td></tr>\n",
       "\t<tr><th scope=row>LINC01551</th><td>0</td><td>1.623801</td><td>0.767</td><td>0.311</td><td>0</td><td>0</td><td>LINC01551</td></tr>\n",
       "\t<tr><th scope=row>OLFM3</th><td>0</td><td>1.543925</td><td>0.396</td><td>0.189</td><td>0</td><td>0</td><td>OLFM3    </td></tr>\n",
       "\t<tr><th scope=row>SFRP1</th><td>0</td><td>1.514941</td><td>0.968</td><td>0.510</td><td>0</td><td>0</td><td>SFRP1    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & p\\_val & avg\\_log2FC & pct.1 & pct.2 & p\\_val\\_adj & cluster & gene\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr>\\\\\n",
       "\\hline\n",
       "\tDCT & 0 & 1.780295 & 0.383 & 0.161 & 0 & 0 & DCT      \\\\\n",
       "\tAMBN & 0 & 1.767947 & 0.431 & 0.175 & 0 & 0 & AMBN     \\\\\n",
       "\tHEY1 & 0 & 1.708512 & 0.464 & 0.202 & 0 & 0 & HEY1     \\\\\n",
       "\tLINC01551 & 0 & 1.623801 & 0.767 & 0.311 & 0 & 0 & LINC01551\\\\\n",
       "\tOLFM3 & 0 & 1.543925 & 0.396 & 0.189 & 0 & 0 & OLFM3    \\\\\n",
       "\tSFRP1 & 0 & 1.514941 & 0.968 & 0.510 & 0 & 0 & SFRP1    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | p_val &lt;dbl&gt; | avg_log2FC &lt;dbl&gt; | pct.1 &lt;dbl&gt; | pct.2 &lt;dbl&gt; | p_val_adj &lt;dbl&gt; | cluster &lt;fct&gt; | gene &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| DCT | 0 | 1.780295 | 0.383 | 0.161 | 0 | 0 | DCT       |\n",
       "| AMBN | 0 | 1.767947 | 0.431 | 0.175 | 0 | 0 | AMBN      |\n",
       "| HEY1 | 0 | 1.708512 | 0.464 | 0.202 | 0 | 0 | HEY1      |\n",
       "| LINC01551 | 0 | 1.623801 | 0.767 | 0.311 | 0 | 0 | LINC01551 |\n",
       "| OLFM3 | 0 | 1.543925 | 0.396 | 0.189 | 0 | 0 | OLFM3     |\n",
       "| SFRP1 | 0 | 1.514941 | 0.968 | 0.510 | 0 | 0 | SFRP1     |\n",
       "\n"
      ],
      "text/plain": [
       "          p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     \n",
       "DCT       0     1.780295   0.383 0.161 0         0       DCT      \n",
       "AMBN      0     1.767947   0.431 0.175 0         0       AMBN     \n",
       "HEY1      0     1.708512   0.464 0.202 0         0       HEY1     \n",
       "LINC01551 0     1.623801   0.767 0.311 0         0       LINC01551\n",
       "OLFM3     0     1.543925   0.396 0.189 0         0       OLFM3    \n",
       "SFRP1     0     1.514941   0.968 0.510 0         0       SFRP1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(cluster_markers)\n",
    "head(cluster_markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f164ee80-2e86-4944-8f15-e8e8fb7b05dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "cl_0 <- cluster_markers %>% filter(cluster %in% 0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a321e7f9-70bc-444f-b2bc-83c4c281ff18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "679"
      ],
      "text/latex": [
       "679"
      ],
      "text/markdown": [
       "679"
      ],
      "text/plain": [
       "[1] 679"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cl_0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dcf23e98-386b-4f95-a970-829ffc09a5e8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 17 × 1</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>summary(cluster_markers$cluster)</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>0</th><td> 679</td></tr>\n",
       "\t<tr><th scope=row>1</th><td> 885</td></tr>\n",
       "\t<tr><th scope=row>2</th><td> 638</td></tr>\n",
       "\t<tr><th scope=row>3</th><td> 453</td></tr>\n",
       "\t<tr><th scope=row>4</th><td> 666</td></tr>\n",
       "\t<tr><th scope=row>5</th><td> 837</td></tr>\n",
       "\t<tr><th scope=row>6</th><td> 510</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>1339</td></tr>\n",
       "\t<tr><th scope=row>8</th><td>1209</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>1232</td></tr>\n",
       "\t<tr><th scope=row>10</th><td> 479</td></tr>\n",
       "\t<tr><th scope=row>11</th><td> 514</td></tr>\n",
       "\t<tr><th scope=row>12</th><td> 965</td></tr>\n",
       "\t<tr><th scope=row>13</th><td> 471</td></tr>\n",
       "\t<tr><th scope=row>14</th><td> 516</td></tr>\n",
       "\t<tr><th scope=row>15</th><td>1197</td></tr>\n",
       "\t<tr><th scope=row>16</th><td>1140</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 17 × 1\n",
       "\\begin{tabular}{r|l}\n",
       "  & summary(cluster\\_markers\\$cluster)\\\\\n",
       "  & <int>\\\\\n",
       "\\hline\n",
       "\t0 &  679\\\\\n",
       "\t1 &  885\\\\\n",
       "\t2 &  638\\\\\n",
       "\t3 &  453\\\\\n",
       "\t4 &  666\\\\\n",
       "\t5 &  837\\\\\n",
       "\t6 &  510\\\\\n",
       "\t7 & 1339\\\\\n",
       "\t8 & 1209\\\\\n",
       "\t9 & 1232\\\\\n",
       "\t10 &  479\\\\\n",
       "\t11 &  514\\\\\n",
       "\t12 &  965\\\\\n",
       "\t13 &  471\\\\\n",
       "\t14 &  516\\\\\n",
       "\t15 & 1197\\\\\n",
       "\t16 & 1140\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 17 × 1\n",
       "\n",
       "| <!--/--> | summary(cluster_markers$cluster) &lt;int&gt; |\n",
       "|---|---|\n",
       "| 0 |  679 |\n",
       "| 1 |  885 |\n",
       "| 2 |  638 |\n",
       "| 3 |  453 |\n",
       "| 4 |  666 |\n",
       "| 5 |  837 |\n",
       "| 6 |  510 |\n",
       "| 7 | 1339 |\n",
       "| 8 | 1209 |\n",
       "| 9 | 1232 |\n",
       "| 10 |  479 |\n",
       "| 11 |  514 |\n",
       "| 12 |  965 |\n",
       "| 13 |  471 |\n",
       "| 14 |  516 |\n",
       "| 15 | 1197 |\n",
       "| 16 | 1140 |\n",
       "\n"
      ],
      "text/plain": [
       "   summary(cluster_markers$cluster)\n",
       "0   679                            \n",
       "1   885                            \n",
       "2   638                            \n",
       "3   453                            \n",
       "4   666                            \n",
       "5   837                            \n",
       "6   510                            \n",
       "7  1339                            \n",
       "8  1209                            \n",
       "9  1232                            \n",
       "10  479                            \n",
       "11  514                            \n",
       "12  965                            \n",
       "13  471                            \n",
       "14  516                            \n",
       "15 1197                            \n",
       "16 1140                            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a <- as.data.frame(summary(cluster_markers$cluster))\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d104681b-600e-499a-9170-534677c5496b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1700"
      ],
      "text/latex": [
       "1700"
      ],
      "text/markdown": [
       "1700"
      ],
      "text/plain": [
       "[1] 1700"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "top_genes <- cluster_markers %>% group_by(cluster) %>% top_n(n = 100, wt = avg_log2FC) %>% as.data.frame()\n",
    "nrow(top_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "4c2342f6-ab21-4a95-9b6e-39ba57d46715",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>1.780295</td><td>0.383</td><td>0.161</td><td>0</td><td>0</td><td>DCT      </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>0</td><td>1.767947</td><td>0.431</td><td>0.175</td><td>0</td><td>0</td><td>AMBN     </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>0</td><td>1.708512</td><td>0.464</td><td>0.202</td><td>0</td><td>0</td><td>HEY1     </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>0</td><td>1.623801</td><td>0.767</td><td>0.311</td><td>0</td><td>0</td><td>LINC01551</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>0</td><td>1.543925</td><td>0.396</td><td>0.189</td><td>0</td><td>0</td><td>OLFM3    </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>0</td><td>1.514941</td><td>0.968</td><td>0.510</td><td>0</td><td>0</td><td>SFRP1    </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & p\\_val & avg\\_log2FC & pct.1 & pct.2 & p\\_val\\_adj & cluster & gene\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 0 & 1.780295 & 0.383 & 0.161 & 0 & 0 & DCT      \\\\\n",
       "\t2 & 0 & 1.767947 & 0.431 & 0.175 & 0 & 0 & AMBN     \\\\\n",
       "\t3 & 0 & 1.708512 & 0.464 & 0.202 & 0 & 0 & HEY1     \\\\\n",
       "\t4 & 0 & 1.623801 & 0.767 & 0.311 & 0 & 0 & LINC01551\\\\\n",
       "\t5 & 0 & 1.543925 & 0.396 & 0.189 & 0 & 0 & OLFM3    \\\\\n",
       "\t6 & 0 & 1.514941 & 0.968 & 0.510 & 0 & 0 & SFRP1    \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | p_val &lt;dbl&gt; | avg_log2FC &lt;dbl&gt; | pct.1 &lt;dbl&gt; | pct.2 &lt;dbl&gt; | p_val_adj &lt;dbl&gt; | cluster &lt;fct&gt; | gene &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 1 | 0 | 1.780295 | 0.383 | 0.161 | 0 | 0 | DCT       |\n",
       "| 2 | 0 | 1.767947 | 0.431 | 0.175 | 0 | 0 | AMBN      |\n",
       "| 3 | 0 | 1.708512 | 0.464 | 0.202 | 0 | 0 | HEY1      |\n",
       "| 4 | 0 | 1.623801 | 0.767 | 0.311 | 0 | 0 | LINC01551 |\n",
       "| 5 | 0 | 1.543925 | 0.396 | 0.189 | 0 | 0 | OLFM3     |\n",
       "| 6 | 0 | 1.514941 | 0.968 | 0.510 | 0 | 0 | SFRP1     |\n",
       "\n"
      ],
      "text/plain": [
       "  p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     \n",
       "1 0     1.780295   0.383 0.161 0         0       DCT      \n",
       "2 0     1.767947   0.431 0.175 0         0       AMBN     \n",
       "3 0     1.708512   0.464 0.202 0         0       HEY1     \n",
       "4 0     1.623801   0.767 0.311 0         0       LINC01551\n",
       "5 0     1.543925   0.396 0.189 0         0       OLFM3    \n",
       "6 0     1.514941   0.968 0.510 0         0       SFRP1    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(top_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6045cd2-ee33-47e3-82fe-5d4fc8c514d4",
   "metadata": {},
   "source": [
    "### EnrichR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6b29e13d-fc0a-42f0-8082-6153936a004c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Genome_Browser_PWMs'</li><li>'TRANSFAC_and_JASPAR_PWMs'</li><li>'Transcription_Factor_PPIs'</li><li>'ChEA_2013'</li><li>'Drug_Perturbations_from_GEO_2014'</li><li>'ENCODE_TF_ChIP-seq_2014'</li><li>'BioCarta_2013'</li><li>'WikiPathways_2013'</li><li>'Disease_Signatures_from_GEO_up_2014'</li><li>'KEGG_2013'</li><li>'TF-LOF_Expression_from_GEO'</li><li>'TargetScan_microRNA'</li><li>'PPI_Hub_Proteins'</li><li>'GeneSigDB'</li><li>'Chromosome_Location'</li><li>'Human_Gene_Atlas'</li><li>'Mouse_Gene_Atlas'</li><li>'Human_Phenotype_Ontology'</li><li>'Epigenomics_Roadmap_HM_ChIP-seq'</li><li>'KEA_2013'</li><li>'NURSA_Human_Endogenous_Complexome'</li><li>'CORUM'</li><li>'SILAC_Phosphoproteomics'</li><li>'Old_CMAP_up'</li><li>'Old_CMAP_down'</li><li>'OMIM_Disease'</li><li>'OMIM_Expanded'</li><li>'VirusMINT'</li><li>'MSigDB_Computational'</li><li>'MSigDB_Oncogenic_Signatures'</li><li>'Disease_Signatures_from_GEO_down_2014'</li><li>'Virus_Perturbations_from_GEO_up'</li><li>'Virus_Perturbations_from_GEO_down'</li><li>'Cancer_Cell_Line_Encyclopedia'</li><li>'NCI-60_Cancer_Cell_Lines'</li><li>'Tissue_Protein_Expression_from_ProteomicsDB'</li><li>'Tissue_Protein_Expression_from_Human_Proteome_Map'</li><li>'HMDB_Metabolites'</li><li>'Pfam_InterPro_Domains'</li><li>'Allen_Brain_Atlas_up'</li><li>'ENCODE_TF_ChIP-seq_2015'</li><li>'ENCODE_Histone_Modifications_2015'</li><li>'Phosphatase_Substrates_from_DEPOD'</li><li>'Allen_Brain_Atlas_down'</li><li>'ENCODE_Histone_Modifications_2013'</li><li>'Achilles_fitness_increase'</li><li>'Achilles_fitness_decrease'</li><li>'BioCarta_2015'</li><li>'HumanCyc_2015'</li><li>'KEGG_2015'</li><li>'Panther_2015'</li><li>'WikiPathways_2015'</li><li>'ESCAPE'</li><li>'HomoloGene'</li><li>'Disease_Perturbations_from_GEO_down'</li><li>'Disease_Perturbations_from_GEO_up'</li><li>'Drug_Perturbations_from_GEO_down'</li><li>'Genes_Associated_with_NIH_Grants'</li><li>'Drug_Perturbations_from_GEO_up'</li><li>'KEA_2015'</li><li>'Gene_Perturbations_from_GEO_up'</li><li>'Gene_Perturbations_from_GEO_down'</li><li>'ChEA_2015'</li><li>'dbGaP'</li><li>'LINCS_L1000_Chem_Pert_up'</li><li>'LINCS_L1000_Chem_Pert_down'</li><li>'GTEx_Tissue_Expression_Down'</li><li>'GTEx_Tissue_Expression_Up'</li><li>'Ligand_Perturbations_from_GEO_down'</li><li>'Aging_Perturbations_from_GEO_down'</li><li>'Aging_Perturbations_from_GEO_up'</li><li>'Ligand_Perturbations_from_GEO_up'</li><li>'MCF7_Perturbations_from_GEO_down'</li><li>'MCF7_Perturbations_from_GEO_up'</li><li>'Microbe_Perturbations_from_GEO_down'</li><li>'Microbe_Perturbations_from_GEO_up'</li><li>'LINCS_L1000_Ligand_Perturbations_down'</li><li>'LINCS_L1000_Ligand_Perturbations_up'</li><li>'L1000_Kinase_and_GPCR_Perturbations_down'</li><li>'L1000_Kinase_and_GPCR_Perturbations_up'</li><li>'KEGG_2016'</li><li>'WikiPathways_2016'</li><li>'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X'</li><li>'Kinase_Perturbations_from_GEO_down'</li><li>'Kinase_Perturbations_from_GEO_up'</li><li>'BioCarta_2016'</li><li>'HumanCyc_2016'</li><li>'NCI-Nature_2016'</li><li>'Panther_2016'</li><li>'DrugMatrix'</li><li>'ChEA_2016'</li><li>'huMAP'</li><li>'Jensen_TISSUES'</li><li>'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO'</li><li>'Jensen_COMPARTMENTS'</li><li>'Jensen_DISEASES'</li><li>'BioPlex_2017'</li><li>'ARCHS4_Tissues'</li><li>'ARCHS4_Cell-lines'</li><li>'ARCHS4_IDG_Coexp'</li><li>'ARCHS4_Kinases_Coexp'</li><li>'ARCHS4_TFs_Coexp'</li><li>'SysMyo_Muscle_Gene_Sets'</li><li>'miRTarBase_2017'</li><li>'TargetScan_microRNA_2017'</li><li>'Enrichr_Libraries_Most_Popular_Genes'</li><li>'Enrichr_Submissions_TF-Gene_Coocurrence'</li><li>'Data_Acquisition_Method_Most_Popular_Genes'</li><li>'DSigDB'</li><li>'TF_Perturbations_Followed_by_Expression'</li><li>'Chromosome_Location_hg19'</li><li>'Rare_Diseases_AutoRIF_ARCHS4_Predictions'</li><li>'Rare_Diseases_GeneRIF_ARCHS4_Predictions'</li><li>'Rare_Diseases_GeneRIF_Gene_Lists'</li><li>'Rare_Diseases_AutoRIF_Gene_Lists'</li><li>'SubCell_BarCode'</li><li>'GWAS_Catalog_2019'</li><li>'WikiPathways_2019_Human'</li><li>'WikiPathways_2019_Mouse'</li><li>'TRRUST_Transcription_Factors_2019'</li><li>'KEGG_2019_Human'</li><li>'KEGG_2019_Mouse'</li><li>'InterPro_Domains_2019'</li><li>'Pfam_Domains_2019'</li><li>'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019'</li><li>'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019'</li><li>'UK_Biobank_GWAS_v1'</li><li>'BioPlanet_2019'</li><li>'ClinVar_2019'</li><li>'PheWeb_2019'</li><li>'DisGeNET'</li><li>'HMS_LINCS_KinomeScan'</li><li>'CCLE_Proteomics_2020'</li><li>'ProteomicsDB_2020'</li><li>'lncHUB_lncRNA_Co-Expression'</li><li>'Virus-Host_PPI_P-HIPSTer_2020'</li><li>'Elsevier_Pathway_Collection'</li><li>'Table_Mining_of_CRISPR_Studies'</li><li>'COVID-19_Related_Gene_Sets'</li><li>'MSigDB_Hallmark_2020'</li><li>'Enrichr_Users_Contributed_Lists_2020'</li><li>'TG_GATES_2020'</li><li>'Allen_Brain_Atlas_10x_scRNA_2021'</li><li>'Descartes_Cell_Types_and_Tissue_2021'</li><li>'KEGG_2021_Human'</li><li>'WikiPathway_2021_Human'</li><li>'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression'</li><li>'GO_Biological_Process_2021'</li><li>'GO_Cellular_Component_2021'</li><li>'GO_Molecular_Function_2021'</li><li>'MGI_Mammalian_Phenotype_Level_4_2021'</li><li>'CellMarker_Augmented_2021'</li><li>'Orphanet_Augmented_2021'</li><li>'COVID-19_Related_Gene_Sets_2021'</li><li>'PanglaoDB_Augmented_2021'</li><li>'Azimuth_Cell_Types_2021'</li><li>'PhenGenI_Association_2021'</li><li>'GTEx_Aging_Signatures_2021'</li><li>'HDSigDB_Human_2021'</li><li>'HDSigDB_Mouse_2021'</li><li>'HuBMAP_ASCTplusB_augmented_2022'</li><li>'FANTOM6_lncRNA_KD_DEGs'</li><li>'MAGMA_Drugs_and_Diseases'</li><li>'Tabula_Sapiens'</li><li>'ChEA_2022'</li><li>'Diabetes_Perturbations_GEO_2022'</li><li>'LINCS_L1000_Chem_Pert_Consensus_Sigs'</li><li>'LINCS_L1000_CRISPR_KO_Consensus_Sigs'</li><li>'Tabula_Muris'</li><li>'Reactome_2022'</li><li>'SynGO_2022'</li><li>'GlyGen_Glycosylated_Proteins_2022'</li><li>'IDG_Drug_Targets_2022'</li><li>'KOMP2_Mouse_Phenotypes_2022'</li><li>'Metabolomics_Workbench_Metabolites_2022'</li><li>'Proteomics_Drug_Atlas_2023'</li><li>'The_Kinase_Library_2023'</li><li>'GTEx_Tissues_V8_2023'</li><li>'GO_Biological_Process_2023'</li><li>'GO_Cellular_Component_2023'</li><li>'GO_Molecular_Function_2023'</li><li>'PFOCR_Pathways_2023'</li><li>'GWAS_Catalog_2023'</li><li>'GeDiPNet_2023'</li><li>'MAGNET_2023'</li><li>'Azimuth_2023'</li><li>'Rummagene_kinases'</li><li>'Rummagene_signatures'</li><li>'Rummagene_transcription_factors'</li><li>'MoTrPAC_2023'</li><li>'WikiPathway_2023_Human'</li><li>'DepMap_CRISPR_GeneDependency_CellLines_2023'</li><li>'SynGO_2024'</li><li>'CellMarker_2024'</li><li>'DGIdb_Drug_Targets_2024'</li><li>'MGI_Mammalian_Phenotype_Level_4_2024'</li><li>'WikiPathways_2024_Mouse'</li><li>'WikiPathways_2024_Human'</li><li>'The_Kinase_Library_2024'</li><li>'PerturbAtlas'</li><li>'Reactome_Pathways_2024'</li><li>'GO_Biological_Process_2025'</li><li>'GO_Cellular_Component_2025'</li><li>'GO_Molecular_Function_2025'</li><li>'NIBR_DRUGseq_2025_up'</li><li>'NIBR_DRUGseq_2025_down'</li><li>'Jensen_DISEASES_Curated_2025'</li><li>'Jensen_DISEASES_Experimental_2025'</li><li>'COMPARTMENTS_Curated_2025'</li><li>'PerturbAtlas_MouseGenePerturbationSigs'</li><li>'PerturbSeq_ReplogleK562'</li><li>'TISSUES_Experimental_2025'</li><li>'COMPARTMENTS_Experimental_2025'</li><li>'PerturbSeq_ReplogleRPE1'</li><li>'TISSUES_Curated_2025'</li><li>'RummaGEO_DrugPerturbations_2025'</li><li>'RummaGEO_GenePerturbations_2025'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Genome\\_Browser\\_PWMs'\n",
       "\\item 'TRANSFAC\\_and\\_JASPAR\\_PWMs'\n",
       "\\item 'Transcription\\_Factor\\_PPIs'\n",
       "\\item 'ChEA\\_2013'\n",
       "\\item 'Drug\\_Perturbations\\_from\\_GEO\\_2014'\n",
       "\\item 'ENCODE\\_TF\\_ChIP-seq\\_2014'\n",
       "\\item 'BioCarta\\_2013'\n",
       "\\item 'WikiPathways\\_2013'\n",
       "\\item 'Disease\\_Signatures\\_from\\_GEO\\_up\\_2014'\n",
       "\\item 'KEGG\\_2013'\n",
       "\\item 'TF-LOF\\_Expression\\_from\\_GEO'\n",
       "\\item 'TargetScan\\_microRNA'\n",
       "\\item 'PPI\\_Hub\\_Proteins'\n",
       "\\item 'GeneSigDB'\n",
       "\\item 'Chromosome\\_Location'\n",
       "\\item 'Human\\_Gene\\_Atlas'\n",
       "\\item 'Mouse\\_Gene\\_Atlas'\n",
       "\\item 'Human\\_Phenotype\\_Ontology'\n",
       "\\item 'Epigenomics\\_Roadmap\\_HM\\_ChIP-seq'\n",
       "\\item 'KEA\\_2013'\n",
       "\\item 'NURSA\\_Human\\_Endogenous\\_Complexome'\n",
       "\\item 'CORUM'\n",
       "\\item 'SILAC\\_Phosphoproteomics'\n",
       "\\item 'Old\\_CMAP\\_up'\n",
       "\\item 'Old\\_CMAP\\_down'\n",
       "\\item 'OMIM\\_Disease'\n",
       "\\item 'OMIM\\_Expanded'\n",
       "\\item 'VirusMINT'\n",
       "\\item 'MSigDB\\_Computational'\n",
       "\\item 'MSigDB\\_Oncogenic\\_Signatures'\n",
       "\\item 'Disease\\_Signatures\\_from\\_GEO\\_down\\_2014'\n",
       "\\item 'Virus\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'Virus\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Cancer\\_Cell\\_Line\\_Encyclopedia'\n",
       "\\item 'NCI-60\\_Cancer\\_Cell\\_Lines'\n",
       "\\item 'Tissue\\_Protein\\_Expression\\_from\\_ProteomicsDB'\n",
       "\\item 'Tissue\\_Protein\\_Expression\\_from\\_Human\\_Proteome\\_Map'\n",
       "\\item 'HMDB\\_Metabolites'\n",
       "\\item 'Pfam\\_InterPro\\_Domains'\n",
       "\\item 'Allen\\_Brain\\_Atlas\\_up'\n",
       "\\item 'ENCODE\\_TF\\_ChIP-seq\\_2015'\n",
       "\\item 'ENCODE\\_Histone\\_Modifications\\_2015'\n",
       "\\item 'Phosphatase\\_Substrates\\_from\\_DEPOD'\n",
       "\\item 'Allen\\_Brain\\_Atlas\\_down'\n",
       "\\item 'ENCODE\\_Histone\\_Modifications\\_2013'\n",
       "\\item 'Achilles\\_fitness\\_increase'\n",
       "\\item 'Achilles\\_fitness\\_decrease'\n",
       "\\item 'BioCarta\\_2015'\n",
       "\\item 'HumanCyc\\_2015'\n",
       "\\item 'KEGG\\_2015'\n",
       "\\item 'Panther\\_2015'\n",
       "\\item 'WikiPathways\\_2015'\n",
       "\\item 'ESCAPE'\n",
       "\\item 'HomoloGene'\n",
       "\\item 'Disease\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Disease\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'Drug\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Genes\\_Associated\\_with\\_NIH\\_Grants'\n",
       "\\item 'Drug\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'KEA\\_2015'\n",
       "\\item 'Gene\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'Gene\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'ChEA\\_2015'\n",
       "\\item 'dbGaP'\n",
       "\\item 'LINCS\\_L1000\\_Chem\\_Pert\\_up'\n",
       "\\item 'LINCS\\_L1000\\_Chem\\_Pert\\_down'\n",
       "\\item 'GTEx\\_Tissue\\_Expression\\_Down'\n",
       "\\item 'GTEx\\_Tissue\\_Expression\\_Up'\n",
       "\\item 'Ligand\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Aging\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Aging\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'Ligand\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'MCF7\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'MCF7\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'Microbe\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Microbe\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'LINCS\\_L1000\\_Ligand\\_Perturbations\\_down'\n",
       "\\item 'LINCS\\_L1000\\_Ligand\\_Perturbations\\_up'\n",
       "\\item 'L1000\\_Kinase\\_and\\_GPCR\\_Perturbations\\_down'\n",
       "\\item 'L1000\\_Kinase\\_and\\_GPCR\\_Perturbations\\_up'\n",
       "\\item 'KEGG\\_2016'\n",
       "\\item 'WikiPathways\\_2016'\n",
       "\\item 'ENCODE\\_and\\_ChEA\\_Consensus\\_TFs\\_from\\_ChIP-X'\n",
       "\\item 'Kinase\\_Perturbations\\_from\\_GEO\\_down'\n",
       "\\item 'Kinase\\_Perturbations\\_from\\_GEO\\_up'\n",
       "\\item 'BioCarta\\_2016'\n",
       "\\item 'HumanCyc\\_2016'\n",
       "\\item 'NCI-Nature\\_2016'\n",
       "\\item 'Panther\\_2016'\n",
       "\\item 'DrugMatrix'\n",
       "\\item 'ChEA\\_2016'\n",
       "\\item 'huMAP'\n",
       "\\item 'Jensen\\_TISSUES'\n",
       "\\item 'RNA-Seq\\_Disease\\_Gene\\_and\\_Drug\\_Signatures\\_from\\_GEO'\n",
       "\\item 'Jensen\\_COMPARTMENTS'\n",
       "\\item 'Jensen\\_DISEASES'\n",
       "\\item 'BioPlex\\_2017'\n",
       "\\item 'ARCHS4\\_Tissues'\n",
       "\\item 'ARCHS4\\_Cell-lines'\n",
       "\\item 'ARCHS4\\_IDG\\_Coexp'\n",
       "\\item 'ARCHS4\\_Kinases\\_Coexp'\n",
       "\\item 'ARCHS4\\_TFs\\_Coexp'\n",
       "\\item 'SysMyo\\_Muscle\\_Gene\\_Sets'\n",
       "\\item 'miRTarBase\\_2017'\n",
       "\\item 'TargetScan\\_microRNA\\_2017'\n",
       "\\item 'Enrichr\\_Libraries\\_Most\\_Popular\\_Genes'\n",
       "\\item 'Enrichr\\_Submissions\\_TF-Gene\\_Coocurrence'\n",
       "\\item 'Data\\_Acquisition\\_Method\\_Most\\_Popular\\_Genes'\n",
       "\\item 'DSigDB'\n",
       "\\item 'TF\\_Perturbations\\_Followed\\_by\\_Expression'\n",
       "\\item 'Chromosome\\_Location\\_hg19'\n",
       "\\item 'Rare\\_Diseases\\_AutoRIF\\_ARCHS4\\_Predictions'\n",
       "\\item 'Rare\\_Diseases\\_GeneRIF\\_ARCHS4\\_Predictions'\n",
       "\\item 'Rare\\_Diseases\\_GeneRIF\\_Gene\\_Lists'\n",
       "\\item 'Rare\\_Diseases\\_AutoRIF\\_Gene\\_Lists'\n",
       "\\item 'SubCell\\_BarCode'\n",
       "\\item 'GWAS\\_Catalog\\_2019'\n",
       "\\item 'WikiPathways\\_2019\\_Human'\n",
       "\\item 'WikiPathways\\_2019\\_Mouse'\n",
       "\\item 'TRRUST\\_Transcription\\_Factors\\_2019'\n",
       "\\item 'KEGG\\_2019\\_Human'\n",
       "\\item 'KEGG\\_2019\\_Mouse'\n",
       "\\item 'InterPro\\_Domains\\_2019'\n",
       "\\item 'Pfam\\_Domains\\_2019'\n",
       "\\item 'DepMap\\_WG\\_CRISPR\\_Screens\\_Broad\\_CellLines\\_2019'\n",
       "\\item 'DepMap\\_WG\\_CRISPR\\_Screens\\_Sanger\\_CellLines\\_2019'\n",
       "\\item 'UK\\_Biobank\\_GWAS\\_v1'\n",
       "\\item 'BioPlanet\\_2019'\n",
       "\\item 'ClinVar\\_2019'\n",
       "\\item 'PheWeb\\_2019'\n",
       "\\item 'DisGeNET'\n",
       "\\item 'HMS\\_LINCS\\_KinomeScan'\n",
       "\\item 'CCLE\\_Proteomics\\_2020'\n",
       "\\item 'ProteomicsDB\\_2020'\n",
       "\\item 'lncHUB\\_lncRNA\\_Co-Expression'\n",
       "\\item 'Virus-Host\\_PPI\\_P-HIPSTer\\_2020'\n",
       "\\item 'Elsevier\\_Pathway\\_Collection'\n",
       "\\item 'Table\\_Mining\\_of\\_CRISPR\\_Studies'\n",
       "\\item 'COVID-19\\_Related\\_Gene\\_Sets'\n",
       "\\item 'MSigDB\\_Hallmark\\_2020'\n",
       "\\item 'Enrichr\\_Users\\_Contributed\\_Lists\\_2020'\n",
       "\\item 'TG\\_GATES\\_2020'\n",
       "\\item 'Allen\\_Brain\\_Atlas\\_10x\\_scRNA\\_2021'\n",
       "\\item 'Descartes\\_Cell\\_Types\\_and\\_Tissue\\_2021'\n",
       "\\item 'KEGG\\_2021\\_Human'\n",
       "\\item 'WikiPathway\\_2021\\_Human'\n",
       "\\item 'HuBMAP\\_ASCT\\_plus\\_B\\_augmented\\_w\\_RNAseq\\_Coexpression'\n",
       "\\item 'GO\\_Biological\\_Process\\_2021'\n",
       "\\item 'GO\\_Cellular\\_Component\\_2021'\n",
       "\\item 'GO\\_Molecular\\_Function\\_2021'\n",
       "\\item 'MGI\\_Mammalian\\_Phenotype\\_Level\\_4\\_2021'\n",
       "\\item 'CellMarker\\_Augmented\\_2021'\n",
       "\\item 'Orphanet\\_Augmented\\_2021'\n",
       "\\item 'COVID-19\\_Related\\_Gene\\_Sets\\_2021'\n",
       "\\item 'PanglaoDB\\_Augmented\\_2021'\n",
       "\\item 'Azimuth\\_Cell\\_Types\\_2021'\n",
       "\\item 'PhenGenI\\_Association\\_2021'\n",
       "\\item 'GTEx\\_Aging\\_Signatures\\_2021'\n",
       "\\item 'HDSigDB\\_Human\\_2021'\n",
       "\\item 'HDSigDB\\_Mouse\\_2021'\n",
       "\\item 'HuBMAP\\_ASCTplusB\\_augmented\\_2022'\n",
       "\\item 'FANTOM6\\_lncRNA\\_KD\\_DEGs'\n",
       "\\item 'MAGMA\\_Drugs\\_and\\_Diseases'\n",
       "\\item 'Tabula\\_Sapiens'\n",
       "\\item 'ChEA\\_2022'\n",
       "\\item 'Diabetes\\_Perturbations\\_GEO\\_2022'\n",
       "\\item 'LINCS\\_L1000\\_Chem\\_Pert\\_Consensus\\_Sigs'\n",
       "\\item 'LINCS\\_L1000\\_CRISPR\\_KO\\_Consensus\\_Sigs'\n",
       "\\item 'Tabula\\_Muris'\n",
       "\\item 'Reactome\\_2022'\n",
       "\\item 'SynGO\\_2022'\n",
       "\\item 'GlyGen\\_Glycosylated\\_Proteins\\_2022'\n",
       "\\item 'IDG\\_Drug\\_Targets\\_2022'\n",
       "\\item 'KOMP2\\_Mouse\\_Phenotypes\\_2022'\n",
       "\\item 'Metabolomics\\_Workbench\\_Metabolites\\_2022'\n",
       "\\item 'Proteomics\\_Drug\\_Atlas\\_2023'\n",
       "\\item 'The\\_Kinase\\_Library\\_2023'\n",
       "\\item 'GTEx\\_Tissues\\_V8\\_2023'\n",
       "\\item 'GO\\_Biological\\_Process\\_2023'\n",
       "\\item 'GO\\_Cellular\\_Component\\_2023'\n",
       "\\item 'GO\\_Molecular\\_Function\\_2023'\n",
       "\\item 'PFOCR\\_Pathways\\_2023'\n",
       "\\item 'GWAS\\_Catalog\\_2023'\n",
       "\\item 'GeDiPNet\\_2023'\n",
       "\\item 'MAGNET\\_2023'\n",
       "\\item 'Azimuth\\_2023'\n",
       "\\item 'Rummagene\\_kinases'\n",
       "\\item 'Rummagene\\_signatures'\n",
       "\\item 'Rummagene\\_transcription\\_factors'\n",
       "\\item 'MoTrPAC\\_2023'\n",
       "\\item 'WikiPathway\\_2023\\_Human'\n",
       "\\item 'DepMap\\_CRISPR\\_GeneDependency\\_CellLines\\_2023'\n",
       "\\item 'SynGO\\_2024'\n",
       "\\item 'CellMarker\\_2024'\n",
       "\\item 'DGIdb\\_Drug\\_Targets\\_2024'\n",
       "\\item 'MGI\\_Mammalian\\_Phenotype\\_Level\\_4\\_2024'\n",
       "\\item 'WikiPathways\\_2024\\_Mouse'\n",
       "\\item 'WikiPathways\\_2024\\_Human'\n",
       "\\item 'The\\_Kinase\\_Library\\_2024'\n",
       "\\item 'PerturbAtlas'\n",
       "\\item 'Reactome\\_Pathways\\_2024'\n",
       "\\item 'GO\\_Biological\\_Process\\_2025'\n",
       "\\item 'GO\\_Cellular\\_Component\\_2025'\n",
       "\\item 'GO\\_Molecular\\_Function\\_2025'\n",
       "\\item 'NIBR\\_DRUGseq\\_2025\\_up'\n",
       "\\item 'NIBR\\_DRUGseq\\_2025\\_down'\n",
       "\\item 'Jensen\\_DISEASES\\_Curated\\_2025'\n",
       "\\item 'Jensen\\_DISEASES\\_Experimental\\_2025'\n",
       "\\item 'COMPARTMENTS\\_Curated\\_2025'\n",
       "\\item 'PerturbAtlas\\_MouseGenePerturbationSigs'\n",
       "\\item 'PerturbSeq\\_ReplogleK562'\n",
       "\\item 'TISSUES\\_Experimental\\_2025'\n",
       "\\item 'COMPARTMENTS\\_Experimental\\_2025'\n",
       "\\item 'PerturbSeq\\_ReplogleRPE1'\n",
       "\\item 'TISSUES\\_Curated\\_2025'\n",
       "\\item 'RummaGEO\\_DrugPerturbations\\_2025'\n",
       "\\item 'RummaGEO\\_GenePerturbations\\_2025'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Genome_Browser_PWMs'\n",
       "2. 'TRANSFAC_and_JASPAR_PWMs'\n",
       "3. 'Transcription_Factor_PPIs'\n",
       "4. 'ChEA_2013'\n",
       "5. 'Drug_Perturbations_from_GEO_2014'\n",
       "6. 'ENCODE_TF_ChIP-seq_2014'\n",
       "7. 'BioCarta_2013'\n",
       "8. 'WikiPathways_2013'\n",
       "9. 'Disease_Signatures_from_GEO_up_2014'\n",
       "10. 'KEGG_2013'\n",
       "11. 'TF-LOF_Expression_from_GEO'\n",
       "12. 'TargetScan_microRNA'\n",
       "13. 'PPI_Hub_Proteins'\n",
       "14. 'GeneSigDB'\n",
       "15. 'Chromosome_Location'\n",
       "16. 'Human_Gene_Atlas'\n",
       "17. 'Mouse_Gene_Atlas'\n",
       "18. 'Human_Phenotype_Ontology'\n",
       "19. 'Epigenomics_Roadmap_HM_ChIP-seq'\n",
       "20. 'KEA_2013'\n",
       "21. 'NURSA_Human_Endogenous_Complexome'\n",
       "22. 'CORUM'\n",
       "23. 'SILAC_Phosphoproteomics'\n",
       "24. 'Old_CMAP_up'\n",
       "25. 'Old_CMAP_down'\n",
       "26. 'OMIM_Disease'\n",
       "27. 'OMIM_Expanded'\n",
       "28. 'VirusMINT'\n",
       "29. 'MSigDB_Computational'\n",
       "30. 'MSigDB_Oncogenic_Signatures'\n",
       "31. 'Disease_Signatures_from_GEO_down_2014'\n",
       "32. 'Virus_Perturbations_from_GEO_up'\n",
       "33. 'Virus_Perturbations_from_GEO_down'\n",
       "34. 'Cancer_Cell_Line_Encyclopedia'\n",
       "35. 'NCI-60_Cancer_Cell_Lines'\n",
       "36. 'Tissue_Protein_Expression_from_ProteomicsDB'\n",
       "37. 'Tissue_Protein_Expression_from_Human_Proteome_Map'\n",
       "38. 'HMDB_Metabolites'\n",
       "39. 'Pfam_InterPro_Domains'\n",
       "40. 'Allen_Brain_Atlas_up'\n",
       "41. 'ENCODE_TF_ChIP-seq_2015'\n",
       "42. 'ENCODE_Histone_Modifications_2015'\n",
       "43. 'Phosphatase_Substrates_from_DEPOD'\n",
       "44. 'Allen_Brain_Atlas_down'\n",
       "45. 'ENCODE_Histone_Modifications_2013'\n",
       "46. 'Achilles_fitness_increase'\n",
       "47. 'Achilles_fitness_decrease'\n",
       "48. 'BioCarta_2015'\n",
       "49. 'HumanCyc_2015'\n",
       "50. 'KEGG_2015'\n",
       "51. 'Panther_2015'\n",
       "52. 'WikiPathways_2015'\n",
       "53. 'ESCAPE'\n",
       "54. 'HomoloGene'\n",
       "55. 'Disease_Perturbations_from_GEO_down'\n",
       "56. 'Disease_Perturbations_from_GEO_up'\n",
       "57. 'Drug_Perturbations_from_GEO_down'\n",
       "58. 'Genes_Associated_with_NIH_Grants'\n",
       "59. 'Drug_Perturbations_from_GEO_up'\n",
       "60. 'KEA_2015'\n",
       "61. 'Gene_Perturbations_from_GEO_up'\n",
       "62. 'Gene_Perturbations_from_GEO_down'\n",
       "63. 'ChEA_2015'\n",
       "64. 'dbGaP'\n",
       "65. 'LINCS_L1000_Chem_Pert_up'\n",
       "66. 'LINCS_L1000_Chem_Pert_down'\n",
       "67. 'GTEx_Tissue_Expression_Down'\n",
       "68. 'GTEx_Tissue_Expression_Up'\n",
       "69. 'Ligand_Perturbations_from_GEO_down'\n",
       "70. 'Aging_Perturbations_from_GEO_down'\n",
       "71. 'Aging_Perturbations_from_GEO_up'\n",
       "72. 'Ligand_Perturbations_from_GEO_up'\n",
       "73. 'MCF7_Perturbations_from_GEO_down'\n",
       "74. 'MCF7_Perturbations_from_GEO_up'\n",
       "75. 'Microbe_Perturbations_from_GEO_down'\n",
       "76. 'Microbe_Perturbations_from_GEO_up'\n",
       "77. 'LINCS_L1000_Ligand_Perturbations_down'\n",
       "78. 'LINCS_L1000_Ligand_Perturbations_up'\n",
       "79. 'L1000_Kinase_and_GPCR_Perturbations_down'\n",
       "80. 'L1000_Kinase_and_GPCR_Perturbations_up'\n",
       "81. 'KEGG_2016'\n",
       "82. 'WikiPathways_2016'\n",
       "83. 'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X'\n",
       "84. 'Kinase_Perturbations_from_GEO_down'\n",
       "85. 'Kinase_Perturbations_from_GEO_up'\n",
       "86. 'BioCarta_2016'\n",
       "87. 'HumanCyc_2016'\n",
       "88. 'NCI-Nature_2016'\n",
       "89. 'Panther_2016'\n",
       "90. 'DrugMatrix'\n",
       "91. 'ChEA_2016'\n",
       "92. 'huMAP'\n",
       "93. 'Jensen_TISSUES'\n",
       "94. 'RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO'\n",
       "95. 'Jensen_COMPARTMENTS'\n",
       "96. 'Jensen_DISEASES'\n",
       "97. 'BioPlex_2017'\n",
       "98. 'ARCHS4_Tissues'\n",
       "99. 'ARCHS4_Cell-lines'\n",
       "100. 'ARCHS4_IDG_Coexp'\n",
       "101. 'ARCHS4_Kinases_Coexp'\n",
       "102. 'ARCHS4_TFs_Coexp'\n",
       "103. 'SysMyo_Muscle_Gene_Sets'\n",
       "104. 'miRTarBase_2017'\n",
       "105. 'TargetScan_microRNA_2017'\n",
       "106. 'Enrichr_Libraries_Most_Popular_Genes'\n",
       "107. 'Enrichr_Submissions_TF-Gene_Coocurrence'\n",
       "108. 'Data_Acquisition_Method_Most_Popular_Genes'\n",
       "109. 'DSigDB'\n",
       "110. 'TF_Perturbations_Followed_by_Expression'\n",
       "111. 'Chromosome_Location_hg19'\n",
       "112. 'Rare_Diseases_AutoRIF_ARCHS4_Predictions'\n",
       "113. 'Rare_Diseases_GeneRIF_ARCHS4_Predictions'\n",
       "114. 'Rare_Diseases_GeneRIF_Gene_Lists'\n",
       "115. 'Rare_Diseases_AutoRIF_Gene_Lists'\n",
       "116. 'SubCell_BarCode'\n",
       "117. 'GWAS_Catalog_2019'\n",
       "118. 'WikiPathways_2019_Human'\n",
       "119. 'WikiPathways_2019_Mouse'\n",
       "120. 'TRRUST_Transcription_Factors_2019'\n",
       "121. 'KEGG_2019_Human'\n",
       "122. 'KEGG_2019_Mouse'\n",
       "123. 'InterPro_Domains_2019'\n",
       "124. 'Pfam_Domains_2019'\n",
       "125. 'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019'\n",
       "126. 'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019'\n",
       "127. 'UK_Biobank_GWAS_v1'\n",
       "128. 'BioPlanet_2019'\n",
       "129. 'ClinVar_2019'\n",
       "130. 'PheWeb_2019'\n",
       "131. 'DisGeNET'\n",
       "132. 'HMS_LINCS_KinomeScan'\n",
       "133. 'CCLE_Proteomics_2020'\n",
       "134. 'ProteomicsDB_2020'\n",
       "135. 'lncHUB_lncRNA_Co-Expression'\n",
       "136. 'Virus-Host_PPI_P-HIPSTer_2020'\n",
       "137. 'Elsevier_Pathway_Collection'\n",
       "138. 'Table_Mining_of_CRISPR_Studies'\n",
       "139. 'COVID-19_Related_Gene_Sets'\n",
       "140. 'MSigDB_Hallmark_2020'\n",
       "141. 'Enrichr_Users_Contributed_Lists_2020'\n",
       "142. 'TG_GATES_2020'\n",
       "143. 'Allen_Brain_Atlas_10x_scRNA_2021'\n",
       "144. 'Descartes_Cell_Types_and_Tissue_2021'\n",
       "145. 'KEGG_2021_Human'\n",
       "146. 'WikiPathway_2021_Human'\n",
       "147. 'HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression'\n",
       "148. 'GO_Biological_Process_2021'\n",
       "149. 'GO_Cellular_Component_2021'\n",
       "150. 'GO_Molecular_Function_2021'\n",
       "151. 'MGI_Mammalian_Phenotype_Level_4_2021'\n",
       "152. 'CellMarker_Augmented_2021'\n",
       "153. 'Orphanet_Augmented_2021'\n",
       "154. 'COVID-19_Related_Gene_Sets_2021'\n",
       "155. 'PanglaoDB_Augmented_2021'\n",
       "156. 'Azimuth_Cell_Types_2021'\n",
       "157. 'PhenGenI_Association_2021'\n",
       "158. 'GTEx_Aging_Signatures_2021'\n",
       "159. 'HDSigDB_Human_2021'\n",
       "160. 'HDSigDB_Mouse_2021'\n",
       "161. 'HuBMAP_ASCTplusB_augmented_2022'\n",
       "162. 'FANTOM6_lncRNA_KD_DEGs'\n",
       "163. 'MAGMA_Drugs_and_Diseases'\n",
       "164. 'Tabula_Sapiens'\n",
       "165. 'ChEA_2022'\n",
       "166. 'Diabetes_Perturbations_GEO_2022'\n",
       "167. 'LINCS_L1000_Chem_Pert_Consensus_Sigs'\n",
       "168. 'LINCS_L1000_CRISPR_KO_Consensus_Sigs'\n",
       "169. 'Tabula_Muris'\n",
       "170. 'Reactome_2022'\n",
       "171. 'SynGO_2022'\n",
       "172. 'GlyGen_Glycosylated_Proteins_2022'\n",
       "173. 'IDG_Drug_Targets_2022'\n",
       "174. 'KOMP2_Mouse_Phenotypes_2022'\n",
       "175. 'Metabolomics_Workbench_Metabolites_2022'\n",
       "176. 'Proteomics_Drug_Atlas_2023'\n",
       "177. 'The_Kinase_Library_2023'\n",
       "178. 'GTEx_Tissues_V8_2023'\n",
       "179. 'GO_Biological_Process_2023'\n",
       "180. 'GO_Cellular_Component_2023'\n",
       "181. 'GO_Molecular_Function_2023'\n",
       "182. 'PFOCR_Pathways_2023'\n",
       "183. 'GWAS_Catalog_2023'\n",
       "184. 'GeDiPNet_2023'\n",
       "185. 'MAGNET_2023'\n",
       "186. 'Azimuth_2023'\n",
       "187. 'Rummagene_kinases'\n",
       "188. 'Rummagene_signatures'\n",
       "189. 'Rummagene_transcription_factors'\n",
       "190. 'MoTrPAC_2023'\n",
       "191. 'WikiPathway_2023_Human'\n",
       "192. 'DepMap_CRISPR_GeneDependency_CellLines_2023'\n",
       "193. 'SynGO_2024'\n",
       "194. 'CellMarker_2024'\n",
       "195. 'DGIdb_Drug_Targets_2024'\n",
       "196. 'MGI_Mammalian_Phenotype_Level_4_2024'\n",
       "197. 'WikiPathways_2024_Mouse'\n",
       "198. 'WikiPathways_2024_Human'\n",
       "199. 'The_Kinase_Library_2024'\n",
       "200. 'PerturbAtlas'\n",
       "201. 'Reactome_Pathways_2024'\n",
       "202. 'GO_Biological_Process_2025'\n",
       "203. 'GO_Cellular_Component_2025'\n",
       "204. 'GO_Molecular_Function_2025'\n",
       "205. 'NIBR_DRUGseq_2025_up'\n",
       "206. 'NIBR_DRUGseq_2025_down'\n",
       "207. 'Jensen_DISEASES_Curated_2025'\n",
       "208. 'Jensen_DISEASES_Experimental_2025'\n",
       "209. 'COMPARTMENTS_Curated_2025'\n",
       "210. 'PerturbAtlas_MouseGenePerturbationSigs'\n",
       "211. 'PerturbSeq_ReplogleK562'\n",
       "212. 'TISSUES_Experimental_2025'\n",
       "213. 'COMPARTMENTS_Experimental_2025'\n",
       "214. 'PerturbSeq_ReplogleRPE1'\n",
       "215. 'TISSUES_Curated_2025'\n",
       "216. 'RummaGEO_DrugPerturbations_2025'\n",
       "217. 'RummaGEO_GenePerturbations_2025'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"Genome_Browser_PWMs\"                               \n",
       "  [2] \"TRANSFAC_and_JASPAR_PWMs\"                          \n",
       "  [3] \"Transcription_Factor_PPIs\"                         \n",
       "  [4] \"ChEA_2013\"                                         \n",
       "  [5] \"Drug_Perturbations_from_GEO_2014\"                  \n",
       "  [6] \"ENCODE_TF_ChIP-seq_2014\"                           \n",
       "  [7] \"BioCarta_2013\"                                     \n",
       "  [8] \"WikiPathways_2013\"                                 \n",
       "  [9] \"Disease_Signatures_from_GEO_up_2014\"               \n",
       " [10] \"KEGG_2013\"                                         \n",
       " [11] \"TF-LOF_Expression_from_GEO\"                        \n",
       " [12] \"TargetScan_microRNA\"                               \n",
       " [13] \"PPI_Hub_Proteins\"                                  \n",
       " [14] \"GeneSigDB\"                                         \n",
       " [15] \"Chromosome_Location\"                               \n",
       " [16] \"Human_Gene_Atlas\"                                  \n",
       " [17] \"Mouse_Gene_Atlas\"                                  \n",
       " [18] \"Human_Phenotype_Ontology\"                          \n",
       " [19] \"Epigenomics_Roadmap_HM_ChIP-seq\"                   \n",
       " [20] \"KEA_2013\"                                          \n",
       " [21] \"NURSA_Human_Endogenous_Complexome\"                 \n",
       " [22] \"CORUM\"                                             \n",
       " [23] \"SILAC_Phosphoproteomics\"                           \n",
       " [24] \"Old_CMAP_up\"                                       \n",
       " [25] \"Old_CMAP_down\"                                     \n",
       " [26] \"OMIM_Disease\"                                      \n",
       " [27] \"OMIM_Expanded\"                                     \n",
       " [28] \"VirusMINT\"                                         \n",
       " [29] \"MSigDB_Computational\"                              \n",
       " [30] \"MSigDB_Oncogenic_Signatures\"                       \n",
       " [31] \"Disease_Signatures_from_GEO_down_2014\"             \n",
       " [32] \"Virus_Perturbations_from_GEO_up\"                   \n",
       " [33] \"Virus_Perturbations_from_GEO_down\"                 \n",
       " [34] \"Cancer_Cell_Line_Encyclopedia\"                     \n",
       " [35] \"NCI-60_Cancer_Cell_Lines\"                          \n",
       " [36] \"Tissue_Protein_Expression_from_ProteomicsDB\"       \n",
       " [37] \"Tissue_Protein_Expression_from_Human_Proteome_Map\" \n",
       " [38] \"HMDB_Metabolites\"                                  \n",
       " [39] \"Pfam_InterPro_Domains\"                             \n",
       " [40] \"Allen_Brain_Atlas_up\"                              \n",
       " [41] \"ENCODE_TF_ChIP-seq_2015\"                           \n",
       " [42] \"ENCODE_Histone_Modifications_2015\"                 \n",
       " [43] \"Phosphatase_Substrates_from_DEPOD\"                 \n",
       " [44] \"Allen_Brain_Atlas_down\"                            \n",
       " [45] \"ENCODE_Histone_Modifications_2013\"                 \n",
       " [46] \"Achilles_fitness_increase\"                         \n",
       " [47] \"Achilles_fitness_decrease\"                         \n",
       " [48] \"BioCarta_2015\"                                     \n",
       " [49] \"HumanCyc_2015\"                                     \n",
       " [50] \"KEGG_2015\"                                         \n",
       " [51] \"Panther_2015\"                                      \n",
       " [52] \"WikiPathways_2015\"                                 \n",
       " [53] \"ESCAPE\"                                            \n",
       " [54] \"HomoloGene\"                                        \n",
       " [55] \"Disease_Perturbations_from_GEO_down\"               \n",
       " [56] \"Disease_Perturbations_from_GEO_up\"                 \n",
       " [57] \"Drug_Perturbations_from_GEO_down\"                  \n",
       " [58] \"Genes_Associated_with_NIH_Grants\"                  \n",
       " [59] \"Drug_Perturbations_from_GEO_up\"                    \n",
       " [60] \"KEA_2015\"                                          \n",
       " [61] \"Gene_Perturbations_from_GEO_up\"                    \n",
       " [62] \"Gene_Perturbations_from_GEO_down\"                  \n",
       " [63] \"ChEA_2015\"                                         \n",
       " [64] \"dbGaP\"                                             \n",
       " [65] \"LINCS_L1000_Chem_Pert_up\"                          \n",
       " [66] \"LINCS_L1000_Chem_Pert_down\"                        \n",
       " [67] \"GTEx_Tissue_Expression_Down\"                       \n",
       " [68] \"GTEx_Tissue_Expression_Up\"                         \n",
       " [69] \"Ligand_Perturbations_from_GEO_down\"                \n",
       " [70] \"Aging_Perturbations_from_GEO_down\"                 \n",
       " [71] \"Aging_Perturbations_from_GEO_up\"                   \n",
       " [72] \"Ligand_Perturbations_from_GEO_up\"                  \n",
       " [73] \"MCF7_Perturbations_from_GEO_down\"                  \n",
       " [74] \"MCF7_Perturbations_from_GEO_up\"                    \n",
       " [75] \"Microbe_Perturbations_from_GEO_down\"               \n",
       " [76] \"Microbe_Perturbations_from_GEO_up\"                 \n",
       " [77] \"LINCS_L1000_Ligand_Perturbations_down\"             \n",
       " [78] \"LINCS_L1000_Ligand_Perturbations_up\"               \n",
       " [79] \"L1000_Kinase_and_GPCR_Perturbations_down\"          \n",
       " [80] \"L1000_Kinase_and_GPCR_Perturbations_up\"            \n",
       " [81] \"KEGG_2016\"                                         \n",
       " [82] \"WikiPathways_2016\"                                 \n",
       " [83] \"ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X\"         \n",
       " [84] \"Kinase_Perturbations_from_GEO_down\"                \n",
       " [85] \"Kinase_Perturbations_from_GEO_up\"                  \n",
       " [86] \"BioCarta_2016\"                                     \n",
       " [87] \"HumanCyc_2016\"                                     \n",
       " [88] \"NCI-Nature_2016\"                                   \n",
       " [89] \"Panther_2016\"                                      \n",
       " [90] \"DrugMatrix\"                                        \n",
       " [91] \"ChEA_2016\"                                         \n",
       " [92] \"huMAP\"                                             \n",
       " [93] \"Jensen_TISSUES\"                                    \n",
       " [94] \"RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\" \n",
       " [95] \"Jensen_COMPARTMENTS\"                               \n",
       " [96] \"Jensen_DISEASES\"                                   \n",
       " [97] \"BioPlex_2017\"                                      \n",
       " [98] \"ARCHS4_Tissues\"                                    \n",
       " [99] \"ARCHS4_Cell-lines\"                                 \n",
       "[100] \"ARCHS4_IDG_Coexp\"                                  \n",
       "[101] \"ARCHS4_Kinases_Coexp\"                              \n",
       "[102] \"ARCHS4_TFs_Coexp\"                                  \n",
       "[103] \"SysMyo_Muscle_Gene_Sets\"                           \n",
       "[104] \"miRTarBase_2017\"                                   \n",
       "[105] \"TargetScan_microRNA_2017\"                          \n",
       "[106] \"Enrichr_Libraries_Most_Popular_Genes\"              \n",
       "[107] \"Enrichr_Submissions_TF-Gene_Coocurrence\"           \n",
       "[108] \"Data_Acquisition_Method_Most_Popular_Genes\"        \n",
       "[109] \"DSigDB\"                                            \n",
       "[110] \"TF_Perturbations_Followed_by_Expression\"           \n",
       "[111] \"Chromosome_Location_hg19\"                          \n",
       "[112] \"Rare_Diseases_AutoRIF_ARCHS4_Predictions\"          \n",
       "[113] \"Rare_Diseases_GeneRIF_ARCHS4_Predictions\"          \n",
       "[114] \"Rare_Diseases_GeneRIF_Gene_Lists\"                  \n",
       "[115] \"Rare_Diseases_AutoRIF_Gene_Lists\"                  \n",
       "[116] \"SubCell_BarCode\"                                   \n",
       "[117] \"GWAS_Catalog_2019\"                                 \n",
       "[118] \"WikiPathways_2019_Human\"                           \n",
       "[119] \"WikiPathways_2019_Mouse\"                           \n",
       "[120] \"TRRUST_Transcription_Factors_2019\"                 \n",
       "[121] \"KEGG_2019_Human\"                                   \n",
       "[122] \"KEGG_2019_Mouse\"                                   \n",
       "[123] \"InterPro_Domains_2019\"                             \n",
       "[124] \"Pfam_Domains_2019\"                                 \n",
       "[125] \"DepMap_WG_CRISPR_Screens_Broad_CellLines_2019\"     \n",
       "[126] \"DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019\"    \n",
       "[127] \"UK_Biobank_GWAS_v1\"                                \n",
       "[128] \"BioPlanet_2019\"                                    \n",
       "[129] \"ClinVar_2019\"                                      \n",
       "[130] \"PheWeb_2019\"                                       \n",
       "[131] \"DisGeNET\"                                          \n",
       "[132] \"HMS_LINCS_KinomeScan\"                              \n",
       "[133] \"CCLE_Proteomics_2020\"                              \n",
       "[134] \"ProteomicsDB_2020\"                                 \n",
       "[135] \"lncHUB_lncRNA_Co-Expression\"                       \n",
       "[136] \"Virus-Host_PPI_P-HIPSTer_2020\"                     \n",
       "[137] \"Elsevier_Pathway_Collection\"                       \n",
       "[138] \"Table_Mining_of_CRISPR_Studies\"                    \n",
       "[139] \"COVID-19_Related_Gene_Sets\"                        \n",
       "[140] \"MSigDB_Hallmark_2020\"                              \n",
       "[141] \"Enrichr_Users_Contributed_Lists_2020\"              \n",
       "[142] \"TG_GATES_2020\"                                     \n",
       "[143] \"Allen_Brain_Atlas_10x_scRNA_2021\"                  \n",
       "[144] \"Descartes_Cell_Types_and_Tissue_2021\"              \n",
       "[145] \"KEGG_2021_Human\"                                   \n",
       "[146] \"WikiPathway_2021_Human\"                            \n",
       "[147] \"HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression\"\n",
       "[148] \"GO_Biological_Process_2021\"                        \n",
       "[149] \"GO_Cellular_Component_2021\"                        \n",
       "[150] \"GO_Molecular_Function_2021\"                        \n",
       "[151] \"MGI_Mammalian_Phenotype_Level_4_2021\"              \n",
       "[152] \"CellMarker_Augmented_2021\"                         \n",
       "[153] \"Orphanet_Augmented_2021\"                           \n",
       "[154] \"COVID-19_Related_Gene_Sets_2021\"                   \n",
       "[155] \"PanglaoDB_Augmented_2021\"                          \n",
       "[156] \"Azimuth_Cell_Types_2021\"                           \n",
       "[157] \"PhenGenI_Association_2021\"                         \n",
       "[158] \"GTEx_Aging_Signatures_2021\"                        \n",
       "[159] \"HDSigDB_Human_2021\"                                \n",
       "[160] \"HDSigDB_Mouse_2021\"                                \n",
       "[161] \"HuBMAP_ASCTplusB_augmented_2022\"                   \n",
       "[162] \"FANTOM6_lncRNA_KD_DEGs\"                            \n",
       "[163] \"MAGMA_Drugs_and_Diseases\"                          \n",
       "[164] \"Tabula_Sapiens\"                                    \n",
       "[165] \"ChEA_2022\"                                         \n",
       "[166] \"Diabetes_Perturbations_GEO_2022\"                   \n",
       "[167] \"LINCS_L1000_Chem_Pert_Consensus_Sigs\"              \n",
       "[168] \"LINCS_L1000_CRISPR_KO_Consensus_Sigs\"              \n",
       "[169] \"Tabula_Muris\"                                      \n",
       "[170] \"Reactome_2022\"                                     \n",
       "[171] \"SynGO_2022\"                                        \n",
       "[172] \"GlyGen_Glycosylated_Proteins_2022\"                 \n",
       "[173] \"IDG_Drug_Targets_2022\"                             \n",
       "[174] \"KOMP2_Mouse_Phenotypes_2022\"                       \n",
       "[175] \"Metabolomics_Workbench_Metabolites_2022\"           \n",
       "[176] \"Proteomics_Drug_Atlas_2023\"                        \n",
       "[177] \"The_Kinase_Library_2023\"                           \n",
       "[178] \"GTEx_Tissues_V8_2023\"                              \n",
       "[179] \"GO_Biological_Process_2023\"                        \n",
       "[180] \"GO_Cellular_Component_2023\"                        \n",
       "[181] \"GO_Molecular_Function_2023\"                        \n",
       "[182] \"PFOCR_Pathways_2023\"                               \n",
       "[183] \"GWAS_Catalog_2023\"                                 \n",
       "[184] \"GeDiPNet_2023\"                                     \n",
       "[185] \"MAGNET_2023\"                                       \n",
       "[186] \"Azimuth_2023\"                                      \n",
       "[187] \"Rummagene_kinases\"                                 \n",
       "[188] \"Rummagene_signatures\"                              \n",
       "[189] \"Rummagene_transcription_factors\"                   \n",
       "[190] \"MoTrPAC_2023\"                                      \n",
       "[191] \"WikiPathway_2023_Human\"                            \n",
       "[192] \"DepMap_CRISPR_GeneDependency_CellLines_2023\"       \n",
       "[193] \"SynGO_2024\"                                        \n",
       "[194] \"CellMarker_2024\"                                   \n",
       "[195] \"DGIdb_Drug_Targets_2024\"                           \n",
       "[196] \"MGI_Mammalian_Phenotype_Level_4_2024\"              \n",
       "[197] \"WikiPathways_2024_Mouse\"                           \n",
       "[198] \"WikiPathways_2024_Human\"                           \n",
       "[199] \"The_Kinase_Library_2024\"                           \n",
       "[200] \"PerturbAtlas\"                                      \n",
       "[201] \"Reactome_Pathways_2024\"                            \n",
       "[202] \"GO_Biological_Process_2025\"                        \n",
       "[203] \"GO_Cellular_Component_2025\"                        \n",
       "[204] \"GO_Molecular_Function_2025\"                        \n",
       "[205] \"NIBR_DRUGseq_2025_up\"                              \n",
       "[206] \"NIBR_DRUGseq_2025_down\"                            \n",
       "[207] \"Jensen_DISEASES_Curated_2025\"                      \n",
       "[208] \"Jensen_DISEASES_Experimental_2025\"                 \n",
       "[209] \"COMPARTMENTS_Curated_2025\"                         \n",
       "[210] \"PerturbAtlas_MouseGenePerturbationSigs\"            \n",
       "[211] \"PerturbSeq_ReplogleK562\"                           \n",
       "[212] \"TISSUES_Experimental_2025\"                         \n",
       "[213] \"COMPARTMENTS_Experimental_2025\"                    \n",
       "[214] \"PerturbSeq_ReplogleRPE1\"                           \n",
       "[215] \"TISSUES_Curated_2025\"                              \n",
       "[216] \"RummaGEO_DrugPerturbations_2025\"                   \n",
       "[217] \"RummaGEO_GenePerturbations_2025\"                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbs <- listEnrichrDbs()\n",
    "unique(dbs$libraryName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1027cae9-a68f-47c1-a31a-ba56731a57ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.16 (BiocManager 1.30.22), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Installing package(s) 'org.Hs.eg.db'\n",
      "\n",
      "Old packages: 'abind', 'ape', 'BH', 'BiocManager', 'bit', 'bit64', 'bitops',\n",
      "  'bslib', 'cachem', 'caTools', 'colorspace', 'commonmark', 'cowplot',\n",
      "  'crayon', 'crosstalk', 'curl', 'data.table', 'DBI', 'deldir', 'DEoptimR',\n",
      "  'digest', 'docopt', 'dotCall64', 'dqrng', 'farver', 'fastDummies', 'fastICA',\n",
      "  'fastmap', 'fitdistrplus', 'FNN', 'fontawesome', 'fs', 'future',\n",
      "  'future.apply', 'ggplot2', 'ggrepel', 'ggridges', 'globals', 'gplots',\n",
      "  'gtable', 'gtools', 'hdf5r', 'highr', 'htmltools', 'htmlwidgets', 'httpuv',\n",
      "  'igraph', 'kernlab', 'knitr', 'later', 'leiden', 'leidenbase', 'listenv',\n",
      "  'littler', 'locfit', 'mathjaxr', 'MatrixModels', 'matrixStats', 'metap',\n",
      "  'miniUI', 'mixtools', 'multcomp', 'munsell', 'mvtnorm', 'parallelly',\n",
      "  'patchwork', 'pheatmap', 'plotly', 'plotrix', 'polyclip', 'progress',\n",
      "  'progressr', 'promises', 'qlcMatrix', 'quantreg', 'R.oo', 'R.utils', 'RANN',\n",
      "  'rbibutils', 'Rcpp', 'RcppAnnoy', 'RcppArmadillo', 'RcppEigen', 'RcppHNSW',\n",
      "  'RcppParallel', 'RcppTOML', 'RcppZiggurat', 'RCurl', 'Rdpack', 'remotes',\n",
      "  'restfulr', 'reticulate', 'Rfast', 'Rfast2', 'rjson', 'robustbase',\n",
      "  'rprojroot', 'RSpectra', 'RSQLite', 'Rtsne', 'RUnit', 'sandwich', 'sass',\n",
      "  'scales', 'sctransform', 'segmented', 'Seurat', 'SeuratObject', 'shiny',\n",
      "  'slam', 'sp', 'spam', 'SparseM', 'spatstat.data', 'spatstat.explore',\n",
      "  'spatstat.geom', 'spatstat.random', 'spatstat.sparse', 'spatstat.utils',\n",
      "  'tensor', 'TH.data', 'tibble', 'tinytex', 'utf8', 'uwot', 'VGAM', 'viridis',\n",
      "  'WriteXLS', 'xfun', 'XML', 'yaml', 'zoo', 'boot', 'class', 'cluster',\n",
      "  'codetools', 'foreign', 'KernSmooth', 'lattice', 'mgcv', 'nlme', 'nnet',\n",
      "  'rpart', 'spatial', 'survival'\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Install if needed\n",
    "if (!requireNamespace(\"org.Hs.eg.db\", quietly = TRUE)) BiocManager::install(\"org.Hs.eg.db\")\n",
    "if (!requireNamespace(\"AnnotationDbi\", quietly = TRUE)) BiocManager::install(\"AnnotationDbi\")\n",
    "\n",
    "library(org.Hs.eg.db)\n",
    "library(AnnotationDbi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "8a9adee7-970b-41ca-ab20-2fb02d8a1b50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming your dataframe is called top_genes and you want to map the gene column named \"gene\"\n",
    "# Example: top_genes$gene contains gene symbols for one cluster\n",
    "\n",
    "# Convert gene symbols to Entrez IDs\n",
    "gene_symbols <- top_genes$gene\n",
    "entrez_ids <- suppressMessages(mapIds(org.Hs.eg.db, keys = gene_symbols, column = \"ENTREZID\", keytype = \"SYMBOL\", multiVals = \"first\"))\n",
    "\n",
    "# Add Entrez IDs to your data.frame\n",
    "top_genes$entrez <- entrez_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "3c5f1613-3025-4ef0-a64f-d1bfe60b5399",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th><th scope=col>entrez</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>0</td><td>1.780295</td><td>0.383</td><td>0.161</td><td>0</td><td>0</td><td>DCT      </td><td>1638  </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>0</td><td>1.767947</td><td>0.431</td><td>0.175</td><td>0</td><td>0</td><td>AMBN     </td><td>258   </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>0</td><td>1.708512</td><td>0.464</td><td>0.202</td><td>0</td><td>0</td><td>HEY1     </td><td>23462 </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>0</td><td>1.623801</td><td>0.767</td><td>0.311</td><td>0</td><td>0</td><td>LINC01551</td><td>387978</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>0</td><td>1.543925</td><td>0.396</td><td>0.189</td><td>0</td><td>0</td><td>OLFM3    </td><td>118427</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>0</td><td>1.514941</td><td>0.968</td><td>0.510</td><td>0</td><td>0</td><td>SFRP1    </td><td>6422  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 8\n",
       "\\begin{tabular}{r|llllllll}\n",
       "  & p\\_val & avg\\_log2FC & pct.1 & pct.2 & p\\_val\\_adj & cluster & gene & entrez\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 0 & 1.780295 & 0.383 & 0.161 & 0 & 0 & DCT       & 1638  \\\\\n",
       "\t2 & 0 & 1.767947 & 0.431 & 0.175 & 0 & 0 & AMBN      & 258   \\\\\n",
       "\t3 & 0 & 1.708512 & 0.464 & 0.202 & 0 & 0 & HEY1      & 23462 \\\\\n",
       "\t4 & 0 & 1.623801 & 0.767 & 0.311 & 0 & 0 & LINC01551 & 387978\\\\\n",
       "\t5 & 0 & 1.543925 & 0.396 & 0.189 & 0 & 0 & OLFM3     & 118427\\\\\n",
       "\t6 & 0 & 1.514941 & 0.968 & 0.510 & 0 & 0 & SFRP1     & 6422  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 8\n",
       "\n",
       "| <!--/--> | p_val &lt;dbl&gt; | avg_log2FC &lt;dbl&gt; | pct.1 &lt;dbl&gt; | pct.2 &lt;dbl&gt; | p_val_adj &lt;dbl&gt; | cluster &lt;fct&gt; | gene &lt;chr&gt; | entrez &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 0 | 1.780295 | 0.383 | 0.161 | 0 | 0 | DCT       | 1638   |\n",
       "| 2 | 0 | 1.767947 | 0.431 | 0.175 | 0 | 0 | AMBN      | 258    |\n",
       "| 3 | 0 | 1.708512 | 0.464 | 0.202 | 0 | 0 | HEY1      | 23462  |\n",
       "| 4 | 0 | 1.623801 | 0.767 | 0.311 | 0 | 0 | LINC01551 | 387978 |\n",
       "| 5 | 0 | 1.543925 | 0.396 | 0.189 | 0 | 0 | OLFM3     | 118427 |\n",
       "| 6 | 0 | 1.514941 | 0.968 | 0.510 | 0 | 0 | SFRP1     | 6422   |\n",
       "\n"
      ],
      "text/plain": [
       "  p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene      entrez\n",
       "1 0     1.780295   0.383 0.161 0         0       DCT       1638  \n",
       "2 0     1.767947   0.431 0.175 0         0       AMBN      258   \n",
       "3 0     1.708512   0.464 0.202 0         0       HEY1      23462 \n",
       "4 0     1.623801   0.767 0.311 0         0       LINC01551 387978\n",
       "5 0     1.543925   0.396 0.189 0         0       OLFM3     118427\n",
       "6 0     1.514941   0.968 0.510 0         0       SFRP1     6422  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(top_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "56c26c57-ca5e-4052-93c8-06e31e772dbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "97"
      ],
      "text/latex": [
       "97"
      ],
      "text/markdown": [
       "97"
      ],
      "text/plain": [
       "[1] 97"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sum(is.na(top_genes$entrez)) # Number of genes not mapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "3d944e65-3c30-4244-9664-5c76aab67e6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TMEM161B-AS1'</li><li>'C1orf61'</li><li>'WBSCR22'</li><li>'LINC00116'</li><li>'TMEM2'</li><li>'CTGF'</li><li>'TMEM161B-AS1'</li><li>'SEPP1'</li><li>'CYR61'</li><li>'C6orf48'</li><li>'FAM57B'</li><li>'AC004540.4'</li><li>'C7orf55'</li><li>'LINC00969'</li><li>'H2AFZ'</li><li>'HIST1H4C'</li><li>'KIAA0101'</li><li>'TCTEX1D2'</li><li>'AC090498.1'</li><li>'C9orf116'</li><li>'CCDC109B'</li><li>'LINC00969'</li><li>'HIST1H1B'</li><li>'HIST1H1D'</li><li>'H2AFX'</li><li>'RP11-421L21.3'</li><li>'RP11-111M22.3'</li><li>'RP4-665J23.1'</li><li>'HMP19'</li><li>'AC004540.4'</li><li>'C10orf35'</li><li>'FAM57B'</li><li>'KIAA1456'</li><li>'SELM'</li><li>'RP11-247C2.2'</li><li>'ERBB2IP'</li><li>'H2AFZ'</li><li>'CTGF'</li><li>'HIST1H4C'</li><li>'LINC00969'</li><li>'HIST1H1B'</li><li>'SGOL1'</li><li>'SGOL2'</li><li>'HIST1H1D'</li><li>'SEPT10'</li><li>'C9orf116'</li><li>'RP11-111M22.3'</li><li>'CASC5'</li><li>'SGOL2'</li><li>'FAM64A'</li><li>'SGOL1'</li><li>'H2AFX'</li><li>'FAM65B'</li><li>'HMP19'</li><li>'RP4-665J23.1'</li><li>'C10orf35'</li><li>'FAM57B'</li><li>'SEPT3'</li><li>'HIST1H3B'</li><li>'HIST1H1B'</li><li>'HIST1H3G'</li><li>'HIST1H2AG'</li><li>'HIST1H1A'</li><li>'HIST2H2AC'</li><li>'HIST1H1D'</li><li>'HIST1H1E'</li><li>'HIST1H1C'</li><li>'KIAA0101'</li><li>'HIST1H4C'</li><li>'FAM64A'</li><li>'H2AFX'</li><li>'SGOL1'</li><li>'ZNF724P'</li><li>'C21orf59'</li><li>'SEPT6'</li><li>'TMEM2'</li><li>'C9orf3'</li><li>'ST5'</li><li>'PLA2G16'</li><li>'CTGF'</li><li>'CYR61'</li><li>'RP11-445F12.1'</li><li>'AC004540.4'</li><li>'FAM57B'</li><li>'LINC00969'</li><li>'FAM212A'</li><li>'CTD-3252C9.4'</li><li>'AC091801.1'</li><li>'RP11-490M8.1'</li><li>'RP11-538P18.2'</li><li>'AC007405.6'</li><li>'RP11-356J5.12'</li><li>'RP11-395G23.3'</li><li>'KB-1562D12.1'</li><li>'PVRL4'</li><li>'RP11-532F12.5'</li><li>'LINC00152'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TMEM161B-AS1'\n",
       "\\item 'C1orf61'\n",
       "\\item 'WBSCR22'\n",
       "\\item 'LINC00116'\n",
       "\\item 'TMEM2'\n",
       "\\item 'CTGF'\n",
       "\\item 'TMEM161B-AS1'\n",
       "\\item 'SEPP1'\n",
       "\\item 'CYR61'\n",
       "\\item 'C6orf48'\n",
       "\\item 'FAM57B'\n",
       "\\item 'AC004540.4'\n",
       "\\item 'C7orf55'\n",
       "\\item 'LINC00969'\n",
       "\\item 'H2AFZ'\n",
       "\\item 'HIST1H4C'\n",
       "\\item 'KIAA0101'\n",
       "\\item 'TCTEX1D2'\n",
       "\\item 'AC090498.1'\n",
       "\\item 'C9orf116'\n",
       "\\item 'CCDC109B'\n",
       "\\item 'LINC00969'\n",
       "\\item 'HIST1H1B'\n",
       "\\item 'HIST1H1D'\n",
       "\\item 'H2AFX'\n",
       "\\item 'RP11-421L21.3'\n",
       "\\item 'RP11-111M22.3'\n",
       "\\item 'RP4-665J23.1'\n",
       "\\item 'HMP19'\n",
       "\\item 'AC004540.4'\n",
       "\\item 'C10orf35'\n",
       "\\item 'FAM57B'\n",
       "\\item 'KIAA1456'\n",
       "\\item 'SELM'\n",
       "\\item 'RP11-247C2.2'\n",
       "\\item 'ERBB2IP'\n",
       "\\item 'H2AFZ'\n",
       "\\item 'CTGF'\n",
       "\\item 'HIST1H4C'\n",
       "\\item 'LINC00969'\n",
       "\\item 'HIST1H1B'\n",
       "\\item 'SGOL1'\n",
       "\\item 'SGOL2'\n",
       "\\item 'HIST1H1D'\n",
       "\\item 'SEPT10'\n",
       "\\item 'C9orf116'\n",
       "\\item 'RP11-111M22.3'\n",
       "\\item 'CASC5'\n",
       "\\item 'SGOL2'\n",
       "\\item 'FAM64A'\n",
       "\\item 'SGOL1'\n",
       "\\item 'H2AFX'\n",
       "\\item 'FAM65B'\n",
       "\\item 'HMP19'\n",
       "\\item 'RP4-665J23.1'\n",
       "\\item 'C10orf35'\n",
       "\\item 'FAM57B'\n",
       "\\item 'SEPT3'\n",
       "\\item 'HIST1H3B'\n",
       "\\item 'HIST1H1B'\n",
       "\\item 'HIST1H3G'\n",
       "\\item 'HIST1H2AG'\n",
       "\\item 'HIST1H1A'\n",
       "\\item 'HIST2H2AC'\n",
       "\\item 'HIST1H1D'\n",
       "\\item 'HIST1H1E'\n",
       "\\item 'HIST1H1C'\n",
       "\\item 'KIAA0101'\n",
       "\\item 'HIST1H4C'\n",
       "\\item 'FAM64A'\n",
       "\\item 'H2AFX'\n",
       "\\item 'SGOL1'\n",
       "\\item 'ZNF724P'\n",
       "\\item 'C21orf59'\n",
       "\\item 'SEPT6'\n",
       "\\item 'TMEM2'\n",
       "\\item 'C9orf3'\n",
       "\\item 'ST5'\n",
       "\\item 'PLA2G16'\n",
       "\\item 'CTGF'\n",
       "\\item 'CYR61'\n",
       "\\item 'RP11-445F12.1'\n",
       "\\item 'AC004540.4'\n",
       "\\item 'FAM57B'\n",
       "\\item 'LINC00969'\n",
       "\\item 'FAM212A'\n",
       "\\item 'CTD-3252C9.4'\n",
       "\\item 'AC091801.1'\n",
       "\\item 'RP11-490M8.1'\n",
       "\\item 'RP11-538P18.2'\n",
       "\\item 'AC007405.6'\n",
       "\\item 'RP11-356J5.12'\n",
       "\\item 'RP11-395G23.3'\n",
       "\\item 'KB-1562D12.1'\n",
       "\\item 'PVRL4'\n",
       "\\item 'RP11-532F12.5'\n",
       "\\item 'LINC00152'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TMEM161B-AS1'\n",
       "2. 'C1orf61'\n",
       "3. 'WBSCR22'\n",
       "4. 'LINC00116'\n",
       "5. 'TMEM2'\n",
       "6. 'CTGF'\n",
       "7. 'TMEM161B-AS1'\n",
       "8. 'SEPP1'\n",
       "9. 'CYR61'\n",
       "10. 'C6orf48'\n",
       "11. 'FAM57B'\n",
       "12. 'AC004540.4'\n",
       "13. 'C7orf55'\n",
       "14. 'LINC00969'\n",
       "15. 'H2AFZ'\n",
       "16. 'HIST1H4C'\n",
       "17. 'KIAA0101'\n",
       "18. 'TCTEX1D2'\n",
       "19. 'AC090498.1'\n",
       "20. 'C9orf116'\n",
       "21. 'CCDC109B'\n",
       "22. 'LINC00969'\n",
       "23. 'HIST1H1B'\n",
       "24. 'HIST1H1D'\n",
       "25. 'H2AFX'\n",
       "26. 'RP11-421L21.3'\n",
       "27. 'RP11-111M22.3'\n",
       "28. 'RP4-665J23.1'\n",
       "29. 'HMP19'\n",
       "30. 'AC004540.4'\n",
       "31. 'C10orf35'\n",
       "32. 'FAM57B'\n",
       "33. 'KIAA1456'\n",
       "34. 'SELM'\n",
       "35. 'RP11-247C2.2'\n",
       "36. 'ERBB2IP'\n",
       "37. 'H2AFZ'\n",
       "38. 'CTGF'\n",
       "39. 'HIST1H4C'\n",
       "40. 'LINC00969'\n",
       "41. 'HIST1H1B'\n",
       "42. 'SGOL1'\n",
       "43. 'SGOL2'\n",
       "44. 'HIST1H1D'\n",
       "45. 'SEPT10'\n",
       "46. 'C9orf116'\n",
       "47. 'RP11-111M22.3'\n",
       "48. 'CASC5'\n",
       "49. 'SGOL2'\n",
       "50. 'FAM64A'\n",
       "51. 'SGOL1'\n",
       "52. 'H2AFX'\n",
       "53. 'FAM65B'\n",
       "54. 'HMP19'\n",
       "55. 'RP4-665J23.1'\n",
       "56. 'C10orf35'\n",
       "57. 'FAM57B'\n",
       "58. 'SEPT3'\n",
       "59. 'HIST1H3B'\n",
       "60. 'HIST1H1B'\n",
       "61. 'HIST1H3G'\n",
       "62. 'HIST1H2AG'\n",
       "63. 'HIST1H1A'\n",
       "64. 'HIST2H2AC'\n",
       "65. 'HIST1H1D'\n",
       "66. 'HIST1H1E'\n",
       "67. 'HIST1H1C'\n",
       "68. 'KIAA0101'\n",
       "69. 'HIST1H4C'\n",
       "70. 'FAM64A'\n",
       "71. 'H2AFX'\n",
       "72. 'SGOL1'\n",
       "73. 'ZNF724P'\n",
       "74. 'C21orf59'\n",
       "75. 'SEPT6'\n",
       "76. 'TMEM2'\n",
       "77. 'C9orf3'\n",
       "78. 'ST5'\n",
       "79. 'PLA2G16'\n",
       "80. 'CTGF'\n",
       "81. 'CYR61'\n",
       "82. 'RP11-445F12.1'\n",
       "83. 'AC004540.4'\n",
       "84. 'FAM57B'\n",
       "85. 'LINC00969'\n",
       "86. 'FAM212A'\n",
       "87. 'CTD-3252C9.4'\n",
       "88. 'AC091801.1'\n",
       "89. 'RP11-490M8.1'\n",
       "90. 'RP11-538P18.2'\n",
       "91. 'AC007405.6'\n",
       "92. 'RP11-356J5.12'\n",
       "93. 'RP11-395G23.3'\n",
       "94. 'KB-1562D12.1'\n",
       "95. 'PVRL4'\n",
       "96. 'RP11-532F12.5'\n",
       "97. 'LINC00152'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"TMEM161B-AS1\"  \"C1orf61\"       \"WBSCR22\"       \"LINC00116\"    \n",
       " [5] \"TMEM2\"         \"CTGF\"          \"TMEM161B-AS1\"  \"SEPP1\"        \n",
       " [9] \"CYR61\"         \"C6orf48\"       \"FAM57B\"        \"AC004540.4\"   \n",
       "[13] \"C7orf55\"       \"LINC00969\"     \"H2AFZ\"         \"HIST1H4C\"     \n",
       "[17] \"KIAA0101\"      \"TCTEX1D2\"      \"AC090498.1\"    \"C9orf116\"     \n",
       "[21] \"CCDC109B\"      \"LINC00969\"     \"HIST1H1B\"      \"HIST1H1D\"     \n",
       "[25] \"H2AFX\"         \"RP11-421L21.3\" \"RP11-111M22.3\" \"RP4-665J23.1\" \n",
       "[29] \"HMP19\"         \"AC004540.4\"    \"C10orf35\"      \"FAM57B\"       \n",
       "[33] \"KIAA1456\"      \"SELM\"          \"RP11-247C2.2\"  \"ERBB2IP\"      \n",
       "[37] \"H2AFZ\"         \"CTGF\"          \"HIST1H4C\"      \"LINC00969\"    \n",
       "[41] \"HIST1H1B\"      \"SGOL1\"         \"SGOL2\"         \"HIST1H1D\"     \n",
       "[45] \"SEPT10\"        \"C9orf116\"      \"RP11-111M22.3\" \"CASC5\"        \n",
       "[49] \"SGOL2\"         \"FAM64A\"        \"SGOL1\"         \"H2AFX\"        \n",
       "[53] \"FAM65B\"        \"HMP19\"         \"RP4-665J23.1\"  \"C10orf35\"     \n",
       "[57] \"FAM57B\"        \"SEPT3\"         \"HIST1H3B\"      \"HIST1H1B\"     \n",
       "[61] \"HIST1H3G\"      \"HIST1H2AG\"     \"HIST1H1A\"      \"HIST2H2AC\"    \n",
       "[65] \"HIST1H1D\"      \"HIST1H1E\"      \"HIST1H1C\"      \"KIAA0101\"     \n",
       "[69] \"HIST1H4C\"      \"FAM64A\"        \"H2AFX\"         \"SGOL1\"        \n",
       "[73] \"ZNF724P\"       \"C21orf59\"      \"SEPT6\"         \"TMEM2\"        \n",
       "[77] \"C9orf3\"        \"ST5\"           \"PLA2G16\"       \"CTGF\"         \n",
       "[81] \"CYR61\"         \"RP11-445F12.1\" \"AC004540.4\"    \"FAM57B\"       \n",
       "[85] \"LINC00969\"     \"FAM212A\"       \"CTD-3252C9.4\"  \"AC091801.1\"   \n",
       "[89] \"RP11-490M8.1\"  \"RP11-538P18.2\" \"AC007405.6\"    \"RP11-356J5.12\"\n",
       "[93] \"RP11-395G23.3\" \"KB-1562D12.1\"  \"PVRL4\"         \"RP11-532F12.5\"\n",
       "[97] \"LINC00152\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>p_val</th><th scope=col>avg_log2FC</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>cluster</th><th scope=col>gene</th><th scope=col>entrez</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " p\\_val & avg\\_log2FC & pct.1 & pct.2 & p\\_val\\_adj & cluster & gene & entrez\\\\\n",
       " <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <fct> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 8\n",
       "\n",
       "| p_val &lt;dbl&gt; | avg_log2FC &lt;dbl&gt; | pct.1 &lt;dbl&gt; | pct.2 &lt;dbl&gt; | p_val_adj &lt;dbl&gt; | cluster &lt;fct&gt; | gene &lt;chr&gt; | entrez &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene entrez"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unmapped_genes <- top_genes[is.na(top_genes$entrez), \"gene\"]\n",
    "unmapped_genes\n",
    "top_genes[is.na(top_genes$entrez) & top_genes$gene %in% genes_of_interest, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "07976acc-d96f-48a5-be4b-35b89588020b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows with NA Entrez IDs\n",
    "top_genes <- top_genes[!is.na(top_genes$entrez), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c15ab23f-a7a2-4ed2-b720-2c5b1828a4d1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\"><tr><td>capture.output {utils}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Send Output to a Character String or File</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Evaluates its arguments with the output being returned as a character\n",
       "string or sent to a file.  Related to <code>sink</code> similarly to how\n",
       "<code>with</code> is related to <code>attach</code>.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>capture.output(..., file = NULL, append = FALSE,\n",
       "               type = c(\"output\", \"message\"), split = FALSE)\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr valign=\"top\"><td><code>...</code></td>\n",
       "<td>\n",
       "<p>Expressions to be evaluated.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>file</code></td>\n",
       "<td>\n",
       "<p>A file name or a connection, or <code>NULL</code> to return\n",
       "the output as a character vector.  If the connection is not open,\n",
       "it will be opened initially and closed on exit.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>append</code></td>\n",
       "<td>\n",
       "<p>logical.  If <code>file</code> a file name or unopened\n",
       "connection, append or overwrite?</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>type, split</code></td>\n",
       "<td>\n",
       "<p>are passed to <code>sink()</code>, see there.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>It works via <code>sink(&lt;file connection&gt;)</code> and hence the <span style=\"font-family: Courier New, Courier; color: #666666;\"><b>R</b></span> code\n",
       "in <code>dots</code> must <em>not</em> interfere with the connection (e.g., by\n",
       "calling <code>closeAllConnections()</code>).\n",
       "</p>\n",
       "<p>An attempt is made to write output as far as possible to <code>file</code>\n",
       "if there is an error in evaluating the expressions, but for\n",
       "<code>file = NULL</code> all output will be lost.\n",
       "</p>\n",
       "<p>Messages sent to <code>stderr()</code> (including those from\n",
       "<code>message</code>, <code>warning</code> and <code>stop</code>)\n",
       "are captured by <code>type = \"message\"</code>.  Note that this can be\n",
       "&ldquo;unsafe&rdquo; and should only be used with care.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A character string (if <code>file = NULL</code>), or invisible <code>NULL</code>.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       " <p><code>sink</code>, <code>textConnection</code> </p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'>require(stats)\n",
       "glmout &lt;- capture.output(summary(glm(case ~ spontaneous+induced,\n",
       "                                     data = infert, family = binomial())))\n",
       "glmout[1:5]\n",
       "capture.output(1+1, 2+2)\n",
       "capture.output({1+1; 2+2})\n",
       "\n",
       "## Not run: ## on Unix-alike with a2ps available\n",
       "op &lt;- options(useFancyQuotes=FALSE)\n",
       "pdf &lt;- pipe(\"a2ps -o - | ps2pdf - tempout.pdf\", \"w\")\n",
       "capture.output(example(glm), file = pdf)\n",
       "close(pdf); options(op) ; system(\"evince tempout.pdf &amp;\")\n",
       "\n",
       "## End(Not run)\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>utils</em> version 4.2.0 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{capture.output}{Send Output to a Character String or File}{capture.output}\n",
       "\\keyword{utilities}{capture.output}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Evaluates its arguments with the output being returned as a character\n",
       "string or sent to a file.  Related to \\code{\\LinkA{sink}{sink}} similarly to how\n",
       "\\code{\\LinkA{with}{with}} is related to \\code{\\LinkA{attach}{attach}}.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "capture.output(..., file = NULL, append = FALSE,\n",
       "               type = c(\"output\", \"message\"), split = FALSE)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{...}] Expressions to be evaluated.\n",
       "\\item[\\code{file}] A file name or a \\LinkA{connection}{connection}, or \\code{NULL} to return\n",
       "the output as a character vector.  If the connection is not open,\n",
       "it will be opened initially and closed on exit.\n",
       "\\item[\\code{append}] logical.  If \\code{file} a file name or unopened\n",
       "connection, append or overwrite?\n",
       "\\item[\\code{type, split}] are passed to \\code{\\LinkA{sink}{sink}()}, see there.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "It works via \\code{\\LinkA{sink}{sink}(<file connection>)} and hence the \\R{} code\n",
       "in \\code{dots} must \\emph{not} interfere with the connection (e.g., by\n",
       "calling \\code{\\LinkA{closeAllConnections}{closeAllConnections}()}).\n",
       "\n",
       "An attempt is made to write output as far as possible to \\code{file}\n",
       "if there is an error in evaluating the expressions, but for\n",
       "\\code{file = NULL} all output will be lost.\n",
       "\n",
       "Messages sent to \\code{\\LinkA{stderr}{stderr}()} (including those from\n",
       "\\code{\\LinkA{message}{message}}, \\code{\\LinkA{warning}{warning}} and \\code{\\LinkA{stop}{stop}})\n",
       "are captured by \\code{type = \"message\"}.  Note that this can be\n",
       "``unsafe'' and should only be used with care.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A character string (if \\code{file = NULL}), or invisible \\code{NULL}.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       " \\code{\\LinkA{sink}{sink}}, \\code{\\LinkA{textConnection}{textConnection}} \n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "require(stats)\n",
       "glmout <- capture.output(summary(glm(case ~ spontaneous+induced,\n",
       "                                     data = infert, family = binomial())))\n",
       "glmout[1:5]\n",
       "capture.output(1+1, 2+2)\n",
       "capture.output({1+1; 2+2})\n",
       "\n",
       "## Not run: ## on Unix-alike with a2ps available\n",
       "op <- options(useFancyQuotes=FALSE)\n",
       "pdf <- pipe(\"a2ps -o - | ps2pdf - tempout.pdf\", \"w\")\n",
       "capture.output(example(glm), file = pdf)\n",
       "close(pdf); options(op) ; system(\"evince tempout.pdf &\")\n",
       "\n",
       "## End(Not run)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "capture.output              package:utils              R Documentation\n",
       "\n",
       "_\bS_\be_\bn_\bd _\bO_\bu_\bt_\bp_\bu_\bt _\bt_\bo _\ba _\bC_\bh_\ba_\br_\ba_\bc_\bt_\be_\br _\bS_\bt_\br_\bi_\bn_\bg _\bo_\br _\bF_\bi_\bl_\be\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Evaluates its arguments with the output being returned as a\n",
       "     character string or sent to a file.  Related to ‘sink’ similarly\n",
       "     to how ‘with’ is related to ‘attach’.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     capture.output(..., file = NULL, append = FALSE,\n",
       "                    type = c(\"output\", \"message\"), split = FALSE)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "     ...: Expressions to be evaluated.\n",
       "\n",
       "    file: A file name or a connection, or ‘NULL’ to return the output\n",
       "          as a character vector.  If the connection is not open, it\n",
       "          will be opened initially and closed on exit.\n",
       "\n",
       "  append: logical.  If ‘file’ a file name or unopened connection,\n",
       "          append or overwrite?\n",
       "\n",
       "type, split: are passed to ‘sink()’, see there.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     It works via ‘sink(<file connection>)’ and hence the R code in\n",
       "     ‘dots’ must _not_ interfere with the connection (e.g., by calling\n",
       "     ‘closeAllConnections()’).\n",
       "\n",
       "     An attempt is made to write output as far as possible to ‘file’ if\n",
       "     there is an error in evaluating the expressions, but for ‘file =\n",
       "     NULL’ all output will be lost.\n",
       "\n",
       "     Messages sent to ‘stderr()’ (including those from ‘message’,\n",
       "     ‘warning’ and ‘stop’) are captured by ‘type = \"message\"’.  Note\n",
       "     that this can be “unsafe” and should only be used with care.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A character string (if ‘file = NULL’), or invisible ‘NULL’.\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘sink’, ‘textConnection’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     require(stats)\n",
       "     glmout <- capture.output(summary(glm(case ~ spontaneous+induced,\n",
       "                                          data = infert, family = binomial())))\n",
       "     glmout[1:5]\n",
       "     capture.output(1+1, 2+2)\n",
       "     capture.output({1+1; 2+2})\n",
       "     \n",
       "     ## Not run:\n",
       "     ## on Unix-alike with a2ps available\n",
       "     op <- options(useFancyQuotes=FALSE)\n",
       "     pdf <- pipe(\"a2ps -o - | ps2pdf - tempout.pdf\", \"w\")\n",
       "     capture.output(example(glm), file = pdf)\n",
       "     close(pdf); options(op) ; system(\"evince tempout.pdf &\")\n",
       "     ## End(Not run)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?capture.output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "6b19616e-f08e-4578-bc5e-f88cf1e45abb",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "12.911 sec elapsed\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>$Allen_Brain_Atlas_10x_scRNA_2021</strong> = <table class=\"dataframe\">\n",
       "<caption>A data.frame: 250 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Mouse 372 SMC up                </td><td>4/73 </td><td>0.0004046182</td><td>0.08264024</td><td>0</td><td>0</td><td>12.636407</td><td> 98.722771</td><td>CCND2;COL4A2;FILIP1;SYNE2                      </td></tr>\n",
       "\t<tr><td>Human Inh L2 PAX6 FREM2 up      </td><td>7/328</td><td>0.0009632692</td><td>0.08264024</td><td>0</td><td>0</td><td> 4.853016</td><td> 33.705059</td><td>EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN</td></tr>\n",
       "\t<tr><td>Human Inh L1 PAX6 MIR101-1 up   </td><td>3/44 </td><td>0.0011940188</td><td>0.08264024</td><td>0</td><td>0</td><td>15.798515</td><td>106.330809</td><td>EFNB2;TRIM24;CNTLN                             </td></tr>\n",
       "\t<tr><td>Mouse 358 Astro up              </td><td>4/100</td><td>0.0013222439</td><td>0.08264024</td><td>0</td><td>0</td><td> 9.070055</td><td> 60.120180</td><td>TTYH1;ID4;SOX9;SAT1                            </td></tr>\n",
       "\t<tr><td>Mouse 356 Astro up              </td><td>3/72 </td><td>0.0048859972</td><td>0.23162675</td><td>0</td><td>0</td><td> 9.374291</td><td> 49.884183</td><td>TTYH1;SOX9;SAT1                                </td></tr>\n",
       "\t<tr><td>Mouse 334 L6b/CT ENT up         </td><td>2/26 </td><td>0.0067368642</td><td>0.23162675</td><td>0</td><td>0</td><td>17.814516</td><td> 89.075444</td><td>URI1;RAB8B                                     </td></tr>\n",
       "\t<tr><td>Mouse 31 Sncg up                </td><td>8/606</td><td>0.0082013131</td><td>0.23162675</td><td>0</td><td>0</td><td> 2.968823</td><td> 14.260627</td><td>SOX2;RSL1D1;URI1;GAR1;GPATCH4;MPDZ;EIF3D;GNL3  </td></tr>\n",
       "\t<tr><td>Mouse 343 L6b CTX up            </td><td>3/87 </td><td>0.0082513789</td><td>0.23162675</td><td>0</td><td>0</td><td> 7.694488</td><td> 36.913342</td><td>EMX1;URI1;RAB8B                                </td></tr>\n",
       "\t<tr><td>Mouse 268 ProS up               </td><td>2/29 </td><td>0.0083385630</td><td>0.23162675</td><td>0</td><td>0</td><td>15.832736</td><td> 75.789160</td><td>EFNB2;RAB8B                                    </td></tr>\n",
       "\t<tr><td>Human Astro L1 FGFR3 SERPINI2 up</td><td>3/101</td><td>0.0123712614</td><td>0.27676560</td><td>0</td><td>0</td><td> 6.590617</td><td> 28.948487</td><td>RFX4;CACHD1;ZFP36L1                            </td></tr>\n",
       "\t<tr><td>Human Inh L1-5 VIP CD27-AS1 up  </td><td>2/37 </td><td>0.0133465555</td><td>0.27676560</td><td>0</td><td>0</td><td>12.208909</td><td> 52.699720</td><td>FOXG1;RAB8B                                    </td></tr>\n",
       "\t<tr><td>Mouse 165 L2/3 IT ENTl up       </td><td>2/38 </td><td>0.0140449013</td><td>0.27676560</td><td>0</td><td>0</td><td>11.869176</td><td> 50.627919</td><td>EFNB2;PKIB                                     </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 SST KLHL1 up     </td><td>2/40 </td><td>0.0154880314</td><td>0.27676560</td><td>0</td><td>0</td><td>11.243350</td><td> 46.858773</td><td>ARX;RAB8B                                      </td></tr>\n",
       "\t<tr><td>Mouse 262 L6 Car3 up            </td><td>4/201</td><td>0.0154988734</td><td>0.27676560</td><td>0</td><td>0</td><td> 4.397389</td><td> 18.323869</td><td>MYO10;B3GAT2;GAR1;PAWR                         </td></tr>\n",
       "\t<tr><td>Human Endo L2-5 NOSTRIN SRGN up </td><td>2/43 </td><td>0.0177664394</td><td>0.29610732</td><td>0</td><td>0</td><td>10.419093</td><td> 41.993569</td><td>YES1;SYNE2                                     </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 PVALB FAM150B up </td><td>2/48 </td><td>0.0218566319</td><td>0.31041640</td><td>0</td><td>0</td><td> 9.284245</td><td> 35.495998</td><td>FILIP1;CREB5                                   </td></tr>\n",
       "\t<tr><td>Mouse 30 Sncg up                </td><td>8/731</td><td>0.0229697283</td><td>0.31041640</td><td>0</td><td>0</td><td> 2.439643</td><td>  9.206184</td><td>SOX2;RSL1D1;URI1;LIX1;GAR1;GPATCH4;EIF3D;GNL3  </td></tr>\n",
       "\t<tr><td>Mouse 105 Pvalb up              </td><td>2/50 </td><td>0.0235916467</td><td>0.31041640</td><td>0</td><td>0</td><td> 8.896505</td><td> 33.333983</td><td>URI1;YES1                                      </td></tr>\n",
       "\t<tr><td>Mouse 181 L2/3 IT CTX up        </td><td>2/50 </td><td>0.0235916467</td><td>0.31041640</td><td>0</td><td>0</td><td> 8.896505</td><td> 33.333983</td><td>MYO10;GAR1                                     </td></tr>\n",
       "\t<tr><td>Mouse 279 CA1 down              </td><td>1/6  </td><td>0.0281669971</td><td>0.35208746</td><td>0</td><td>0</td><td>42.340426</td><td>151.138565</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Mouse 283 CA1 down              </td><td>1/7  </td><td>0.0327845741</td><td>0.35635407</td><td>0</td><td>0</td><td>35.281915</td><td>120.586429</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Mouse 285 CA1 down              </td><td>1/7  </td><td>0.0327845741</td><td>0.35635407</td><td>0</td><td>0</td><td>35.281915</td><td>120.586429</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Mouse 287 CA1 down              </td><td>1/7  </td><td>0.0327845741</td><td>0.35635407</td><td>0</td><td>0</td><td>35.281915</td><td>120.586429</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Astro L1-6 FGFR3 PLCG1 up </td><td>2/64 </td><td>0.0372052475</td><td>0.37459555</td><td>0</td><td>0</td><td> 6.882761</td><td> 22.653269</td><td>TTYH1;RFX4                                     </td></tr>\n",
       "\t<tr><td>Mouse 333 L6b/CT ENT up         </td><td>2/66 </td><td>0.0393463853</td><td>0.37459555</td><td>0</td><td>0</td><td> 6.667003</td><td> 21.570095</td><td>URI1;RAB8B                                     </td></tr>\n",
       "\t<tr><td>Mouse 34 Sncg up                </td><td>2/68 </td><td>0.0415333663</td><td>0.37459555</td><td>0</td><td>0</td><td> 6.464321</td><td> 20.564673</td><td>URI1;LIX1                                      </td></tr>\n",
       "\t<tr><td>Mouse 286 CA1 down              </td><td>1/9  </td><td>0.0419547020</td><td>0.37459555</td><td>0</td><td>0</td><td>26.458777</td><td> 83.905140</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Mouse 288 CA1 down              </td><td>1/9  </td><td>0.0419547020</td><td>0.37459555</td><td>0</td><td>0</td><td>26.458777</td><td> 83.905140</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Mouse 271 ProS down             </td><td>1/10 </td><td>0.0465074540</td><td>0.37506011</td><td>0</td><td>0</td><td>23.517730</td><td> 72.155753</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Mouse 278 CA1-ProS down         </td><td>1/10 </td><td>0.0465074540</td><td>0.37506011</td><td>0</td><td>0</td><td>23.517730</td><td> 72.155753</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>Human Astro L1-6 FGFR3 PLCG1 down  </td><td>3/880 </td><td>0.7943071</td><td>0.8960408</td><td>0</td><td>0</td><td>0.7075009</td><td>0.162926961</td><td>URI1;LIMCH1;MPDZ                                     </td></tr>\n",
       "\t<tr><td>Mouse 356 Astro down               </td><td>2/617 </td><td>0.7956842</td><td>0.8960408</td><td>0</td><td>0</td><td>0.6745345</td><td>0.154166813</td><td>FAM171B;ZFP36L1                                      </td></tr>\n",
       "\t<tr><td>Human Exc L6 FEZF2 PROKR2 up       </td><td>2/640 </td><td>0.8122398</td><td>0.9059726</td><td>0</td><td>0</td><td>0.6494421</td><td>0.135057771</td><td>RSL1D1;RPL3                                          </td></tr>\n",
       "\t<tr><td>Mouse 361 Oligo down               </td><td>3/913 </td><td>0.8145518</td><td>0.9059726</td><td>0</td><td>0</td><td>0.6806617</td><td>0.139615446</td><td>RPL3;FAM171B;RAB11FIP2                               </td></tr>\n",
       "\t<tr><td>Mouse 364 Oligo down               </td><td>2/650 </td><td>0.8190620</td><td>0.9059726</td><td>0</td><td>0</td><td>0.6390880</td><td>0.127559120</td><td>RPL3;NCALD                                           </td></tr>\n",
       "\t<tr><td>Mouse 378 Micro down               </td><td>1/365 </td><td>0.8269063</td><td>0.9059726</td><td>0</td><td>0</td><td>0.5711071</td><td>0.108546852</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Mouse 360 OPC down                 </td><td>3/937 </td><td>0.8282101</td><td>0.9059726</td><td>0</td><td>0</td><td>0.6623336</td><td>0.124842228</td><td>RPL3;FAM171B;RAB11FIP2                               </td></tr>\n",
       "\t<tr><td>Mouse 295 CA3 up                   </td><td>2/667 </td><td>0.8301576</td><td>0.9059726</td><td>0</td><td>0</td><td>0.6222007</td><td>0.115816226</td><td>EFNB2;MPDZ                                           </td></tr>\n",
       "\t<tr><td>Mouse 121 DG down                  </td><td>1/372 </td><td>0.8326850</td><td>0.9059726</td><td>0</td><td>0</td><td>0.5601308</td><td>0.102559839</td><td>RAB11FIP2                                            </td></tr>\n",
       "\t<tr><td>Mouse 122 DG down                  </td><td>1/373 </td><td>0.8334948</td><td>0.9059726</td><td>0</td><td>0</td><td>0.5585964</td><td>0.101735923</td><td>RAB11FIP2                                            </td></tr>\n",
       "\t<tr><td>Mouse 285 CA1 up                   </td><td>1/409 </td><td>0.8602019</td><td>0.9267974</td><td>0</td><td>0</td><td>0.5083698</td><td>0.076554464</td><td>RAB8B                                                </td></tr>\n",
       "\t<tr><td>Mouse 293 CA3 up                   </td><td>1/417 </td><td>0.8655349</td><td>0.9267974</td><td>0</td><td>0</td><td>0.4983889</td><td>0.071971160</td><td>PLAGL1                                               </td></tr>\n",
       "\t<tr><td>Mouse 371 Peri down                </td><td>1/419 </td><td>0.8668363</td><td>0.9267974</td><td>0</td><td>0</td><td>0.4959534</td><td>0.070874264</td><td>NCALD                                                </td></tr>\n",
       "\t<tr><td>Mouse 379 PVM down                 </td><td>1/420 </td><td>0.8674824</td><td>0.9267974</td><td>0</td><td>0</td><td>0.4947443</td><td>0.070332888</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Human Exc L3-5 FEZF2 ASGR2 up      </td><td>2/759 </td><td>0.8803019</td><td>0.9354647</td><td>0</td><td>0</td><td>0.5439695</td><td>0.069350857</td><td>INTU;RAB8B                                           </td></tr>\n",
       "\t<tr><td>Mouse 96 Sst up                    </td><td>8/2354</td><td>0.8841243</td><td>0.9354647</td><td>0</td><td>0</td><td>0.6882441</td><td>0.084762478</td><td>URI1;TRIM24;RPS27L;GPATCH4;FGFBP3;PAICS;ADAMTS6;CNTLN</td></tr>\n",
       "\t<tr><td>Human Oligo L3-6 OPALIN ENPP6 down </td><td>2/774 </td><td>0.8870642</td><td>0.9354647</td><td>0</td><td>0</td><td>0.5329823</td><td>0.063871512</td><td>FAM171B;TRMT11                                       </td></tr>\n",
       "\t<tr><td>Mouse 372 SMC down                 </td><td>1/466 </td><td>0.8940751</td><td>0.9354647</td><td>0</td><td>0</td><td>0.4447495</td><td>0.049796621</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Mouse 362 Oligo down               </td><td>2/791 </td><td>0.8943043</td><td>0.9354647</td><td>0</td><td>0</td><td>0.5210352</td><td>0.058204424</td><td>RPL3;FAM171B                                         </td></tr>\n",
       "\t<tr><td>Mouse 374 VLMC down                </td><td>1/513 </td><td>0.9157885</td><td>0.9539464</td><td>0</td><td>0</td><td>0.4029463</td><td>0.035447105</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Mouse 375 VLMC down                </td><td>1/535 </td><td>0.9243769</td><td>0.9588972</td><td>0</td><td>0</td><td>0.3859072</td><td>0.030345983</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Human Oligo L2-6 OPALIN FTH1P3 down</td><td>2/912 </td><td>0.9346953</td><td>0.9655944</td><td>0</td><td>0</td><td>0.4488952</td><td>0.030315972</td><td>URI1;MPDZ                                            </td></tr>\n",
       "\t<tr><td>Mouse 370 Endo down                </td><td>1/607 </td><td>0.9468633</td><td>0.9675483</td><td>0</td><td>0</td><td>0.3387929</td><td>0.018498280</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Mouse 369 Endo down                </td><td>1/611 </td><td>0.9478969</td><td>0.9675483</td><td>0</td><td>0</td><td>0.3365016</td><td>0.018006041</td><td>RPL3                                                 </td></tr>\n",
       "\t<tr><td>Human Inh L3-5 PVALB ISG20 up      </td><td>7/2402</td><td>0.9481973</td><td>0.9675483</td><td>0</td><td>0</td><td>0.5815620</td><td>0.030934839</td><td>ZNF397;PKIB;NOC3L;ABCE1;PAICS;CNTLN;EIF3B            </td></tr>\n",
       "\t<tr><td>Mouse 110 Pvalb up                 </td><td>1/631 </td><td>0.9527739</td><td>0.9682662</td><td>0</td><td>0</td><td>0.3254813</td><td>0.015746004</td><td>PAICS                                                </td></tr>\n",
       "\t<tr><td>Human Micro L1-6 TYROBP CD74 down  </td><td>1/704 </td><td>0.9670376</td><td>0.9781560</td><td>0</td><td>0</td><td>0.2905784</td><td>0.009739584</td><td>URI1                                                 </td></tr>\n",
       "\t<tr><td>Human Exc L5 THEMIS LINC01116 up   </td><td>6/2315</td><td>0.9703307</td><td>0.9781560</td><td>0</td><td>0</td><td>0.5137493</td><td>0.015473257</td><td>DPP7;DCT;INTU;NOC3L;ABCE1;EIF3B                      </td></tr>\n",
       "\t<tr><td>Mouse 1 CR down                    </td><td>1/769 </td><td>0.9760958</td><td>0.9800158</td><td>0</td><td>0</td><td>0.2650848</td><td>0.006413612</td><td>FAM171B                                              </td></tr>\n",
       "\t<tr><td>Mouse 359 OPC down                 </td><td>1/841 </td><td>0.9832754</td><td>0.9832754</td><td>0</td><td>0</td><td>0.2414514</td><td>0.004072323</td><td>FAM171B                                              </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\textbf{\\$Allen\\_Brain\\_Atlas\\_10x\\_scRNA\\_2021} = A data.frame: 250 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t Mouse 372 SMC up                 & 4/73  & 0.0004046182 & 0.08264024 & 0 & 0 & 12.636407 &  98.722771 & CCND2;COL4A2;FILIP1;SYNE2                      \\\\\n",
       "\t Human Inh L2 PAX6 FREM2 up       & 7/328 & 0.0009632692 & 0.08264024 & 0 & 0 &  4.853016 &  33.705059 & EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN\\\\\n",
       "\t Human Inh L1 PAX6 MIR101-1 up    & 3/44  & 0.0011940188 & 0.08264024 & 0 & 0 & 15.798515 & 106.330809 & EFNB2;TRIM24;CNTLN                             \\\\\n",
       "\t Mouse 358 Astro up               & 4/100 & 0.0013222439 & 0.08264024 & 0 & 0 &  9.070055 &  60.120180 & TTYH1;ID4;SOX9;SAT1                            \\\\\n",
       "\t Mouse 356 Astro up               & 3/72  & 0.0048859972 & 0.23162675 & 0 & 0 &  9.374291 &  49.884183 & TTYH1;SOX9;SAT1                                \\\\\n",
       "\t Mouse 334 L6b/CT ENT up          & 2/26  & 0.0067368642 & 0.23162675 & 0 & 0 & 17.814516 &  89.075444 & URI1;RAB8B                                     \\\\\n",
       "\t Mouse 31 Sncg up                 & 8/606 & 0.0082013131 & 0.23162675 & 0 & 0 &  2.968823 &  14.260627 & SOX2;RSL1D1;URI1;GAR1;GPATCH4;MPDZ;EIF3D;GNL3  \\\\\n",
       "\t Mouse 343 L6b CTX up             & 3/87  & 0.0082513789 & 0.23162675 & 0 & 0 &  7.694488 &  36.913342 & EMX1;URI1;RAB8B                                \\\\\n",
       "\t Mouse 268 ProS up                & 2/29  & 0.0083385630 & 0.23162675 & 0 & 0 & 15.832736 &  75.789160 & EFNB2;RAB8B                                    \\\\\n",
       "\t Human Astro L1 FGFR3 SERPINI2 up & 3/101 & 0.0123712614 & 0.27676560 & 0 & 0 &  6.590617 &  28.948487 & RFX4;CACHD1;ZFP36L1                            \\\\\n",
       "\t Human Inh L1-5 VIP CD27-AS1 up   & 2/37  & 0.0133465555 & 0.27676560 & 0 & 0 & 12.208909 &  52.699720 & FOXG1;RAB8B                                    \\\\\n",
       "\t Mouse 165 L2/3 IT ENTl up        & 2/38  & 0.0140449013 & 0.27676560 & 0 & 0 & 11.869176 &  50.627919 & EFNB2;PKIB                                     \\\\\n",
       "\t Human Inh L5-6 SST KLHL1 up      & 2/40  & 0.0154880314 & 0.27676560 & 0 & 0 & 11.243350 &  46.858773 & ARX;RAB8B                                      \\\\\n",
       "\t Mouse 262 L6 Car3 up             & 4/201 & 0.0154988734 & 0.27676560 & 0 & 0 &  4.397389 &  18.323869 & MYO10;B3GAT2;GAR1;PAWR                         \\\\\n",
       "\t Human Endo L2-5 NOSTRIN SRGN up  & 2/43  & 0.0177664394 & 0.29610732 & 0 & 0 & 10.419093 &  41.993569 & YES1;SYNE2                                     \\\\\n",
       "\t Human Inh L5-6 PVALB FAM150B up  & 2/48  & 0.0218566319 & 0.31041640 & 0 & 0 &  9.284245 &  35.495998 & FILIP1;CREB5                                   \\\\\n",
       "\t Mouse 30 Sncg up                 & 8/731 & 0.0229697283 & 0.31041640 & 0 & 0 &  2.439643 &   9.206184 & SOX2;RSL1D1;URI1;LIX1;GAR1;GPATCH4;EIF3D;GNL3  \\\\\n",
       "\t Mouse 105 Pvalb up               & 2/50  & 0.0235916467 & 0.31041640 & 0 & 0 &  8.896505 &  33.333983 & URI1;YES1                                      \\\\\n",
       "\t Mouse 181 L2/3 IT CTX up         & 2/50  & 0.0235916467 & 0.31041640 & 0 & 0 &  8.896505 &  33.333983 & MYO10;GAR1                                     \\\\\n",
       "\t Mouse 279 CA1 down               & 1/6   & 0.0281669971 & 0.35208746 & 0 & 0 & 42.340426 & 151.138565 & NCALD                                          \\\\\n",
       "\t Mouse 283 CA1 down               & 1/7   & 0.0327845741 & 0.35635407 & 0 & 0 & 35.281915 & 120.586429 & NCALD                                          \\\\\n",
       "\t Mouse 285 CA1 down               & 1/7   & 0.0327845741 & 0.35635407 & 0 & 0 & 35.281915 & 120.586429 & NCALD                                          \\\\\n",
       "\t Mouse 287 CA1 down               & 1/7   & 0.0327845741 & 0.35635407 & 0 & 0 & 35.281915 & 120.586429 & NCALD                                          \\\\\n",
       "\t Human Astro L1-6 FGFR3 PLCG1 up  & 2/64  & 0.0372052475 & 0.37459555 & 0 & 0 &  6.882761 &  22.653269 & TTYH1;RFX4                                     \\\\\n",
       "\t Mouse 333 L6b/CT ENT up          & 2/66  & 0.0393463853 & 0.37459555 & 0 & 0 &  6.667003 &  21.570095 & URI1;RAB8B                                     \\\\\n",
       "\t Mouse 34 Sncg up                 & 2/68  & 0.0415333663 & 0.37459555 & 0 & 0 &  6.464321 &  20.564673 & URI1;LIX1                                      \\\\\n",
       "\t Mouse 286 CA1 down               & 1/9   & 0.0419547020 & 0.37459555 & 0 & 0 & 26.458777 &  83.905140 & NCALD                                          \\\\\n",
       "\t Mouse 288 CA1 down               & 1/9   & 0.0419547020 & 0.37459555 & 0 & 0 & 26.458777 &  83.905140 & NCALD                                          \\\\\n",
       "\t Mouse 271 ProS down              & 1/10  & 0.0465074540 & 0.37506011 & 0 & 0 & 23.517730 &  72.155753 & NCALD                                          \\\\\n",
       "\t Mouse 278 CA1-ProS down          & 1/10  & 0.0465074540 & 0.37506011 & 0 & 0 & 23.517730 &  72.155753 & NCALD                                          \\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t Human Astro L1-6 FGFR3 PLCG1 down   & 3/880  & 0.7943071 & 0.8960408 & 0 & 0 & 0.7075009 & 0.162926961 & URI1;LIMCH1;MPDZ                                     \\\\\n",
       "\t Mouse 356 Astro down                & 2/617  & 0.7956842 & 0.8960408 & 0 & 0 & 0.6745345 & 0.154166813 & FAM171B;ZFP36L1                                      \\\\\n",
       "\t Human Exc L6 FEZF2 PROKR2 up        & 2/640  & 0.8122398 & 0.9059726 & 0 & 0 & 0.6494421 & 0.135057771 & RSL1D1;RPL3                                          \\\\\n",
       "\t Mouse 361 Oligo down                & 3/913  & 0.8145518 & 0.9059726 & 0 & 0 & 0.6806617 & 0.139615446 & RPL3;FAM171B;RAB11FIP2                               \\\\\n",
       "\t Mouse 364 Oligo down                & 2/650  & 0.8190620 & 0.9059726 & 0 & 0 & 0.6390880 & 0.127559120 & RPL3;NCALD                                           \\\\\n",
       "\t Mouse 378 Micro down                & 1/365  & 0.8269063 & 0.9059726 & 0 & 0 & 0.5711071 & 0.108546852 & RPL3                                                 \\\\\n",
       "\t Mouse 360 OPC down                  & 3/937  & 0.8282101 & 0.9059726 & 0 & 0 & 0.6623336 & 0.124842228 & RPL3;FAM171B;RAB11FIP2                               \\\\\n",
       "\t Mouse 295 CA3 up                    & 2/667  & 0.8301576 & 0.9059726 & 0 & 0 & 0.6222007 & 0.115816226 & EFNB2;MPDZ                                           \\\\\n",
       "\t Mouse 121 DG down                   & 1/372  & 0.8326850 & 0.9059726 & 0 & 0 & 0.5601308 & 0.102559839 & RAB11FIP2                                            \\\\\n",
       "\t Mouse 122 DG down                   & 1/373  & 0.8334948 & 0.9059726 & 0 & 0 & 0.5585964 & 0.101735923 & RAB11FIP2                                            \\\\\n",
       "\t Mouse 285 CA1 up                    & 1/409  & 0.8602019 & 0.9267974 & 0 & 0 & 0.5083698 & 0.076554464 & RAB8B                                                \\\\\n",
       "\t Mouse 293 CA3 up                    & 1/417  & 0.8655349 & 0.9267974 & 0 & 0 & 0.4983889 & 0.071971160 & PLAGL1                                               \\\\\n",
       "\t Mouse 371 Peri down                 & 1/419  & 0.8668363 & 0.9267974 & 0 & 0 & 0.4959534 & 0.070874264 & NCALD                                                \\\\\n",
       "\t Mouse 379 PVM down                  & 1/420  & 0.8674824 & 0.9267974 & 0 & 0 & 0.4947443 & 0.070332888 & RPL3                                                 \\\\\n",
       "\t Human Exc L3-5 FEZF2 ASGR2 up       & 2/759  & 0.8803019 & 0.9354647 & 0 & 0 & 0.5439695 & 0.069350857 & INTU;RAB8B                                           \\\\\n",
       "\t Mouse 96 Sst up                     & 8/2354 & 0.8841243 & 0.9354647 & 0 & 0 & 0.6882441 & 0.084762478 & URI1;TRIM24;RPS27L;GPATCH4;FGFBP3;PAICS;ADAMTS6;CNTLN\\\\\n",
       "\t Human Oligo L3-6 OPALIN ENPP6 down  & 2/774  & 0.8870642 & 0.9354647 & 0 & 0 & 0.5329823 & 0.063871512 & FAM171B;TRMT11                                       \\\\\n",
       "\t Mouse 372 SMC down                  & 1/466  & 0.8940751 & 0.9354647 & 0 & 0 & 0.4447495 & 0.049796621 & RPL3                                                 \\\\\n",
       "\t Mouse 362 Oligo down                & 2/791  & 0.8943043 & 0.9354647 & 0 & 0 & 0.5210352 & 0.058204424 & RPL3;FAM171B                                         \\\\\n",
       "\t Mouse 374 VLMC down                 & 1/513  & 0.9157885 & 0.9539464 & 0 & 0 & 0.4029463 & 0.035447105 & RPL3                                                 \\\\\n",
       "\t Mouse 375 VLMC down                 & 1/535  & 0.9243769 & 0.9588972 & 0 & 0 & 0.3859072 & 0.030345983 & RPL3                                                 \\\\\n",
       "\t Human Oligo L2-6 OPALIN FTH1P3 down & 2/912  & 0.9346953 & 0.9655944 & 0 & 0 & 0.4488952 & 0.030315972 & URI1;MPDZ                                            \\\\\n",
       "\t Mouse 370 Endo down                 & 1/607  & 0.9468633 & 0.9675483 & 0 & 0 & 0.3387929 & 0.018498280 & RPL3                                                 \\\\\n",
       "\t Mouse 369 Endo down                 & 1/611  & 0.9478969 & 0.9675483 & 0 & 0 & 0.3365016 & 0.018006041 & RPL3                                                 \\\\\n",
       "\t Human Inh L3-5 PVALB ISG20 up       & 7/2402 & 0.9481973 & 0.9675483 & 0 & 0 & 0.5815620 & 0.030934839 & ZNF397;PKIB;NOC3L;ABCE1;PAICS;CNTLN;EIF3B            \\\\\n",
       "\t Mouse 110 Pvalb up                  & 1/631  & 0.9527739 & 0.9682662 & 0 & 0 & 0.3254813 & 0.015746004 & PAICS                                                \\\\\n",
       "\t Human Micro L1-6 TYROBP CD74 down   & 1/704  & 0.9670376 & 0.9781560 & 0 & 0 & 0.2905784 & 0.009739584 & URI1                                                 \\\\\n",
       "\t Human Exc L5 THEMIS LINC01116 up    & 6/2315 & 0.9703307 & 0.9781560 & 0 & 0 & 0.5137493 & 0.015473257 & DPP7;DCT;INTU;NOC3L;ABCE1;EIF3B                      \\\\\n",
       "\t Mouse 1 CR down                     & 1/769  & 0.9760958 & 0.9800158 & 0 & 0 & 0.2650848 & 0.006413612 & FAM171B                                              \\\\\n",
       "\t Mouse 359 OPC down                  & 1/841  & 0.9832754 & 0.9832754 & 0 & 0 & 0.2414514 & 0.004072323 & FAM171B                                              \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "**$Allen_Brain_Atlas_10x_scRNA_2021** = \n",
       "A data.frame: 250 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| Mouse 372 SMC up                 | 4/73  | 0.0004046182 | 0.08264024 | 0 | 0 | 12.636407 |  98.722771 | CCND2;COL4A2;FILIP1;SYNE2                       |\n",
       "| Human Inh L2 PAX6 FREM2 up       | 7/328 | 0.0009632692 | 0.08264024 | 0 | 0 |  4.853016 |  33.705059 | EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN |\n",
       "| Human Inh L1 PAX6 MIR101-1 up    | 3/44  | 0.0011940188 | 0.08264024 | 0 | 0 | 15.798515 | 106.330809 | EFNB2;TRIM24;CNTLN                              |\n",
       "| Mouse 358 Astro up               | 4/100 | 0.0013222439 | 0.08264024 | 0 | 0 |  9.070055 |  60.120180 | TTYH1;ID4;SOX9;SAT1                             |\n",
       "| Mouse 356 Astro up               | 3/72  | 0.0048859972 | 0.23162675 | 0 | 0 |  9.374291 |  49.884183 | TTYH1;SOX9;SAT1                                 |\n",
       "| Mouse 334 L6b/CT ENT up          | 2/26  | 0.0067368642 | 0.23162675 | 0 | 0 | 17.814516 |  89.075444 | URI1;RAB8B                                      |\n",
       "| Mouse 31 Sncg up                 | 8/606 | 0.0082013131 | 0.23162675 | 0 | 0 |  2.968823 |  14.260627 | SOX2;RSL1D1;URI1;GAR1;GPATCH4;MPDZ;EIF3D;GNL3   |\n",
       "| Mouse 343 L6b CTX up             | 3/87  | 0.0082513789 | 0.23162675 | 0 | 0 |  7.694488 |  36.913342 | EMX1;URI1;RAB8B                                 |\n",
       "| Mouse 268 ProS up                | 2/29  | 0.0083385630 | 0.23162675 | 0 | 0 | 15.832736 |  75.789160 | EFNB2;RAB8B                                     |\n",
       "| Human Astro L1 FGFR3 SERPINI2 up | 3/101 | 0.0123712614 | 0.27676560 | 0 | 0 |  6.590617 |  28.948487 | RFX4;CACHD1;ZFP36L1                             |\n",
       "| Human Inh L1-5 VIP CD27-AS1 up   | 2/37  | 0.0133465555 | 0.27676560 | 0 | 0 | 12.208909 |  52.699720 | FOXG1;RAB8B                                     |\n",
       "| Mouse 165 L2/3 IT ENTl up        | 2/38  | 0.0140449013 | 0.27676560 | 0 | 0 | 11.869176 |  50.627919 | EFNB2;PKIB                                      |\n",
       "| Human Inh L5-6 SST KLHL1 up      | 2/40  | 0.0154880314 | 0.27676560 | 0 | 0 | 11.243350 |  46.858773 | ARX;RAB8B                                       |\n",
       "| Mouse 262 L6 Car3 up             | 4/201 | 0.0154988734 | 0.27676560 | 0 | 0 |  4.397389 |  18.323869 | MYO10;B3GAT2;GAR1;PAWR                          |\n",
       "| Human Endo L2-5 NOSTRIN SRGN up  | 2/43  | 0.0177664394 | 0.29610732 | 0 | 0 | 10.419093 |  41.993569 | YES1;SYNE2                                      |\n",
       "| Human Inh L5-6 PVALB FAM150B up  | 2/48  | 0.0218566319 | 0.31041640 | 0 | 0 |  9.284245 |  35.495998 | FILIP1;CREB5                                    |\n",
       "| Mouse 30 Sncg up                 | 8/731 | 0.0229697283 | 0.31041640 | 0 | 0 |  2.439643 |   9.206184 | SOX2;RSL1D1;URI1;LIX1;GAR1;GPATCH4;EIF3D;GNL3   |\n",
       "| Mouse 105 Pvalb up               | 2/50  | 0.0235916467 | 0.31041640 | 0 | 0 |  8.896505 |  33.333983 | URI1;YES1                                       |\n",
       "| Mouse 181 L2/3 IT CTX up         | 2/50  | 0.0235916467 | 0.31041640 | 0 | 0 |  8.896505 |  33.333983 | MYO10;GAR1                                      |\n",
       "| Mouse 279 CA1 down               | 1/6   | 0.0281669971 | 0.35208746 | 0 | 0 | 42.340426 | 151.138565 | NCALD                                           |\n",
       "| Mouse 283 CA1 down               | 1/7   | 0.0327845741 | 0.35635407 | 0 | 0 | 35.281915 | 120.586429 | NCALD                                           |\n",
       "| Mouse 285 CA1 down               | 1/7   | 0.0327845741 | 0.35635407 | 0 | 0 | 35.281915 | 120.586429 | NCALD                                           |\n",
       "| Mouse 287 CA1 down               | 1/7   | 0.0327845741 | 0.35635407 | 0 | 0 | 35.281915 | 120.586429 | NCALD                                           |\n",
       "| Human Astro L1-6 FGFR3 PLCG1 up  | 2/64  | 0.0372052475 | 0.37459555 | 0 | 0 |  6.882761 |  22.653269 | TTYH1;RFX4                                      |\n",
       "| Mouse 333 L6b/CT ENT up          | 2/66  | 0.0393463853 | 0.37459555 | 0 | 0 |  6.667003 |  21.570095 | URI1;RAB8B                                      |\n",
       "| Mouse 34 Sncg up                 | 2/68  | 0.0415333663 | 0.37459555 | 0 | 0 |  6.464321 |  20.564673 | URI1;LIX1                                       |\n",
       "| Mouse 286 CA1 down               | 1/9   | 0.0419547020 | 0.37459555 | 0 | 0 | 26.458777 |  83.905140 | NCALD                                           |\n",
       "| Mouse 288 CA1 down               | 1/9   | 0.0419547020 | 0.37459555 | 0 | 0 | 26.458777 |  83.905140 | NCALD                                           |\n",
       "| Mouse 271 ProS down              | 1/10  | 0.0465074540 | 0.37506011 | 0 | 0 | 23.517730 |  72.155753 | NCALD                                           |\n",
       "| Mouse 278 CA1-ProS down          | 1/10  | 0.0465074540 | 0.37506011 | 0 | 0 | 23.517730 |  72.155753 | NCALD                                           |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Human Astro L1-6 FGFR3 PLCG1 down   | 3/880  | 0.7943071 | 0.8960408 | 0 | 0 | 0.7075009 | 0.162926961 | URI1;LIMCH1;MPDZ                                      |\n",
       "| Mouse 356 Astro down                | 2/617  | 0.7956842 | 0.8960408 | 0 | 0 | 0.6745345 | 0.154166813 | FAM171B;ZFP36L1                                       |\n",
       "| Human Exc L6 FEZF2 PROKR2 up        | 2/640  | 0.8122398 | 0.9059726 | 0 | 0 | 0.6494421 | 0.135057771 | RSL1D1;RPL3                                           |\n",
       "| Mouse 361 Oligo down                | 3/913  | 0.8145518 | 0.9059726 | 0 | 0 | 0.6806617 | 0.139615446 | RPL3;FAM171B;RAB11FIP2                                |\n",
       "| Mouse 364 Oligo down                | 2/650  | 0.8190620 | 0.9059726 | 0 | 0 | 0.6390880 | 0.127559120 | RPL3;NCALD                                            |\n",
       "| Mouse 378 Micro down                | 1/365  | 0.8269063 | 0.9059726 | 0 | 0 | 0.5711071 | 0.108546852 | RPL3                                                  |\n",
       "| Mouse 360 OPC down                  | 3/937  | 0.8282101 | 0.9059726 | 0 | 0 | 0.6623336 | 0.124842228 | RPL3;FAM171B;RAB11FIP2                                |\n",
       "| Mouse 295 CA3 up                    | 2/667  | 0.8301576 | 0.9059726 | 0 | 0 | 0.6222007 | 0.115816226 | EFNB2;MPDZ                                            |\n",
       "| Mouse 121 DG down                   | 1/372  | 0.8326850 | 0.9059726 | 0 | 0 | 0.5601308 | 0.102559839 | RAB11FIP2                                             |\n",
       "| Mouse 122 DG down                   | 1/373  | 0.8334948 | 0.9059726 | 0 | 0 | 0.5585964 | 0.101735923 | RAB11FIP2                                             |\n",
       "| Mouse 285 CA1 up                    | 1/409  | 0.8602019 | 0.9267974 | 0 | 0 | 0.5083698 | 0.076554464 | RAB8B                                                 |\n",
       "| Mouse 293 CA3 up                    | 1/417  | 0.8655349 | 0.9267974 | 0 | 0 | 0.4983889 | 0.071971160 | PLAGL1                                                |\n",
       "| Mouse 371 Peri down                 | 1/419  | 0.8668363 | 0.9267974 | 0 | 0 | 0.4959534 | 0.070874264 | NCALD                                                 |\n",
       "| Mouse 379 PVM down                  | 1/420  | 0.8674824 | 0.9267974 | 0 | 0 | 0.4947443 | 0.070332888 | RPL3                                                  |\n",
       "| Human Exc L3-5 FEZF2 ASGR2 up       | 2/759  | 0.8803019 | 0.9354647 | 0 | 0 | 0.5439695 | 0.069350857 | INTU;RAB8B                                            |\n",
       "| Mouse 96 Sst up                     | 8/2354 | 0.8841243 | 0.9354647 | 0 | 0 | 0.6882441 | 0.084762478 | URI1;TRIM24;RPS27L;GPATCH4;FGFBP3;PAICS;ADAMTS6;CNTLN |\n",
       "| Human Oligo L3-6 OPALIN ENPP6 down  | 2/774  | 0.8870642 | 0.9354647 | 0 | 0 | 0.5329823 | 0.063871512 | FAM171B;TRMT11                                        |\n",
       "| Mouse 372 SMC down                  | 1/466  | 0.8940751 | 0.9354647 | 0 | 0 | 0.4447495 | 0.049796621 | RPL3                                                  |\n",
       "| Mouse 362 Oligo down                | 2/791  | 0.8943043 | 0.9354647 | 0 | 0 | 0.5210352 | 0.058204424 | RPL3;FAM171B                                          |\n",
       "| Mouse 374 VLMC down                 | 1/513  | 0.9157885 | 0.9539464 | 0 | 0 | 0.4029463 | 0.035447105 | RPL3                                                  |\n",
       "| Mouse 375 VLMC down                 | 1/535  | 0.9243769 | 0.9588972 | 0 | 0 | 0.3859072 | 0.030345983 | RPL3                                                  |\n",
       "| Human Oligo L2-6 OPALIN FTH1P3 down | 2/912  | 0.9346953 | 0.9655944 | 0 | 0 | 0.4488952 | 0.030315972 | URI1;MPDZ                                             |\n",
       "| Mouse 370 Endo down                 | 1/607  | 0.9468633 | 0.9675483 | 0 | 0 | 0.3387929 | 0.018498280 | RPL3                                                  |\n",
       "| Mouse 369 Endo down                 | 1/611  | 0.9478969 | 0.9675483 | 0 | 0 | 0.3365016 | 0.018006041 | RPL3                                                  |\n",
       "| Human Inh L3-5 PVALB ISG20 up       | 7/2402 | 0.9481973 | 0.9675483 | 0 | 0 | 0.5815620 | 0.030934839 | ZNF397;PKIB;NOC3L;ABCE1;PAICS;CNTLN;EIF3B             |\n",
       "| Mouse 110 Pvalb up                  | 1/631  | 0.9527739 | 0.9682662 | 0 | 0 | 0.3254813 | 0.015746004 | PAICS                                                 |\n",
       "| Human Micro L1-6 TYROBP CD74 down   | 1/704  | 0.9670376 | 0.9781560 | 0 | 0 | 0.2905784 | 0.009739584 | URI1                                                  |\n",
       "| Human Exc L5 THEMIS LINC01116 up    | 6/2315 | 0.9703307 | 0.9781560 | 0 | 0 | 0.5137493 | 0.015473257 | DPP7;DCT;INTU;NOC3L;ABCE1;EIF3B                       |\n",
       "| Mouse 1 CR down                     | 1/769  | 0.9760958 | 0.9800158 | 0 | 0 | 0.2650848 | 0.006413612 | FAM171B                                               |\n",
       "| Mouse 359 OPC down                  | 1/841  | 0.9832754 | 0.9832754 | 0 | 0 | 0.2414514 | 0.004072323 | FAM171B                                               |\n",
       "\n"
      ],
      "text/plain": [
       "$Allen_Brain_Atlas_10x_scRNA_2021\n",
       "                                   Term Overlap      P.value Adjusted.P.value\n",
       "1                      Mouse 372 SMC up    4/73 0.0004046182       0.08264024\n",
       "2            Human Inh L2 PAX6 FREM2 up   7/328 0.0009632692       0.08264024\n",
       "3         Human Inh L1 PAX6 MIR101-1 up    3/44 0.0011940188       0.08264024\n",
       "4                    Mouse 358 Astro up   4/100 0.0013222439       0.08264024\n",
       "5                    Mouse 356 Astro up    3/72 0.0048859972       0.23162675\n",
       "6               Mouse 334 L6b/CT ENT up    2/26 0.0067368642       0.23162675\n",
       "7                      Mouse 31 Sncg up   8/606 0.0082013131       0.23162675\n",
       "8                  Mouse 343 L6b CTX up    3/87 0.0082513789       0.23162675\n",
       "9                     Mouse 268 ProS up    2/29 0.0083385630       0.23162675\n",
       "10     Human Astro L1 FGFR3 SERPINI2 up   3/101 0.0123712614       0.27676560\n",
       "11       Human Inh L1-5 VIP CD27-AS1 up    2/37 0.0133465555       0.27676560\n",
       "12            Mouse 165 L2/3 IT ENTl up    2/38 0.0140449013       0.27676560\n",
       "13          Human Inh L5-6 SST KLHL1 up    2/40 0.0154880314       0.27676560\n",
       "14                 Mouse 262 L6 Car3 up   4/201 0.0154988734       0.27676560\n",
       "15      Human Endo L2-5 NOSTRIN SRGN up    2/43 0.0177664394       0.29610732\n",
       "16      Human Inh L5-6 PVALB FAM150B up    2/48 0.0218566319       0.31041640\n",
       "17                     Mouse 30 Sncg up   8/731 0.0229697283       0.31041640\n",
       "18                   Mouse 105 Pvalb up    2/50 0.0235916467       0.31041640\n",
       "19             Mouse 181 L2/3 IT CTX up    2/50 0.0235916467       0.31041640\n",
       "20                   Mouse 279 CA1 down     1/6 0.0281669971       0.35208746\n",
       "21                   Mouse 283 CA1 down     1/7 0.0327845741       0.35635407\n",
       "22                   Mouse 285 CA1 down     1/7 0.0327845741       0.35635407\n",
       "23                   Mouse 287 CA1 down     1/7 0.0327845741       0.35635407\n",
       "24      Human Astro L1-6 FGFR3 PLCG1 up    2/64 0.0372052475       0.37459555\n",
       "25              Mouse 333 L6b/CT ENT up    2/66 0.0393463853       0.37459555\n",
       "26                     Mouse 34 Sncg up    2/68 0.0415333663       0.37459555\n",
       "27                   Mouse 286 CA1 down     1/9 0.0419547020       0.37459555\n",
       "28                   Mouse 288 CA1 down     1/9 0.0419547020       0.37459555\n",
       "29                  Mouse 271 ProS down    1/10 0.0465074540       0.37506011\n",
       "30              Mouse 278 CA1-ProS down    1/10 0.0465074540       0.37506011\n",
       "31                   Mouse 282 CA1 down    1/10 0.0465074540       0.37506011\n",
       "32              Mouse 139 L2 IT ENTl up    2/74 0.0483590340       0.37523538\n",
       "33         Human Exc L6 FEZF2 KLK7 down    1/11 0.0510387981       0.37523538\n",
       "34              Mouse 276 CA1-ProS down    1/11 0.0510387981       0.37523538\n",
       "35       Human Exc L3-5 RORB LAMA4 down    1/12 0.0555488338       0.37523538\n",
       "36                      Mouse 50 Vip up    1/12 0.0555488338       0.37523538\n",
       "37     Human Exc L2-3 LINC00507 DSG3 up   3/184 0.0573841939       0.37523538\n",
       "38   Human Exc L2-3 LINC00507 DSG3 down    1/13 0.0600376601       0.37523538\n",
       "39                      Mouse 40 Vip up    1/13 0.0600376601       0.37523538\n",
       "40                  Mouse 272 ProS down    1/13 0.0600376601       0.37523538\n",
       "41                     Mouse 7 Lamp5 up    1/14 0.0645053757       0.38396057\n",
       "42                   Mouse 284 CA1 down    1/14 0.0645053757       0.38396057\n",
       "43                      Mouse 41 Vip up    1/15 0.0689520787       0.39915662\n",
       "44         Human Exc L6 FEZF2 POGK down    1/16 0.0733778669       0.39915662\n",
       "45                   Mouse 281 CA1 down    1/16 0.0733778669       0.39915662\n",
       "46          Mouse 169 L2/3 IT ENTl down    2/94 0.0736823899       0.39915662\n",
       "47              Mouse 137 L2 IT ENTl up    2/95 0.0750414448       0.39915662\n",
       "48      Human Exc L3-5 FEZF2 ASGR2 down    1/18 0.0821670875       0.42492730\n",
       "49                    Mouse 271 ProS up   2/103 0.0861946207       0.42492730\n",
       "50           Human Inh L1 LAMP5 NMBR up    1/19 0.0865307131       0.42492730\n",
       "51     Human Inh L3-6 PAX6 LINC01497 up   2/106 0.0904990018       0.42492730\n",
       "52     Human Exc L5 FEZF2 NREP-AS1 down    1/20 0.0908738104       0.42492730\n",
       "53                 Mouse 22 Ndnf HPF up   2/108 0.0934033899       0.42492730\n",
       "54  Human Exc L3-5 FEZF2 LINC01107 down    1/21 0.0951964750       0.42492730\n",
       "55             Mouse 308 L5 NP CTX down    1/22 0.0994988018       0.42492730\n",
       "56          Human Inh L5 PVALB LRIG3 up   3/234 0.1004061115       0.42492730\n",
       "57         Human Inh L5-6 VIP COL4A3 up   2/115 0.1037759572       0.42492730\n",
       "58                   Mouse 289 CA1 down    1/24 0.1080428207       0.42492730\n",
       "59                Mouse 3 Lamp5 Lhx6 up    1/24 0.1080428207       0.42492730\n",
       "60          Human Inh L2-5 VIP SOX11 up   2/119 0.1098390280       0.42492730\n",
       "61    Human Inh L5-6 PVALB ZFPM2-AS1 up   2/120 0.1113693366       0.42492730\n",
       "62   Human Inh L3-6 PAX6 LINC01497 down    1/25 0.1122847006       0.42492730\n",
       "63                   Mouse 280 CA1 down    1/25 0.1122847006       0.42492730\n",
       "64        Human Inh L5-6 SST DNAJC14 up   5/543 0.1164224001       0.42492730\n",
       "65       Human Exc L5-6 FEZF2 CFTR down    1/26 0.1165066190       0.42492730\n",
       "66   Human Exc L5-6 FEZF2 IFNG-AS1 down    1/26 0.1165066190       0.42492730\n",
       "67                Mouse 63 Sst Chodl up   4/402 0.1245535839       0.42492730\n",
       "68  Human Exc L5 FEZF2 RNF144A-AS1 down    1/28 0.1248909420       0.42492730\n",
       "69                      Mouse 85 Sst up    1/28 0.1248909420       0.42492730\n",
       "70                Mouse 62 Sst Chodl up   2/129 0.1253856925       0.42492730\n",
       "71                      Mouse 68 Sst up    1/29 0.1290535312       0.42492730\n",
       "72                 Mouse 118 Pvalb down   2/132 0.1301487202       0.42492730\n",
       "73    Human Exc L2 LINC00507 ATP7B down    1/30 0.1331965279       0.42492730\n",
       "74                     Mouse 8 Lamp5 up    1/30 0.1331965279       0.42492730\n",
       "75     Human Exc L5-6 THEMIS SMYD1 down    1/31 0.1373200234       0.42492730\n",
       "76      Human Inh L1 PAX6 MIR101-1 down    1/31 0.1373200234       0.42492730\n",
       "77                     Mouse 35 Sncg up   2/138 0.1397980942       0.42492730\n",
       "78                    Mouse 369 Endo up   2/138 0.1397980942       0.42492730\n",
       "79                   Mouse 357 Astro up    1/32 0.1414241085       0.42492730\n",
       "80                   Mouse 25 Sncg down    1/32 0.1414241085       0.42492730\n",
       "81             Mouse 307 L5 NP CTX down    1/32 0.1414241085       0.42492730\n",
       "82                     Mouse 264 SUB up   2/141 0.1446802717       0.42492730\n",
       "83          Human Exc L3-5 RORB LNX2 up    1/33 0.1455088735       0.42492730\n",
       "84               Mouse 149 L2 IT PAR up    1/33 0.1455088735       0.42492730\n",
       "85               Human Inh L6 SST TH up   3/279 0.1472417428       0.42492730\n",
       "86              Mouse 335 L6b/CT ENT up    1/34 0.1495744084       0.42492730\n",
       "87                     Mouse 359 OPC up    1/34 0.1495744084       0.42492730\n",
       "88               Mouse 262 L6 Car3 down    1/34 0.1495744084       0.42492730\n",
       "89             Mouse 190 L4/5 IT CTX up    1/35 0.1536208029       0.42672445\n",
       "90                   Mouse 290 CA1 down    1/35 0.1536208029       0.42672445\n",
       "91                     Mouse 279 CA1 up    1/36 0.1576481460       0.43309930\n",
       "92     Human Inh L3-6 VIP ZIM2-AS1 down    1/37 0.1616565266       0.43591061\n",
       "93              Mouse 337 L6b/CT ENT up    1/38 0.1656460329       0.43591061\n",
       "94      Human Inh L3-6 VIP UG0898H09 up    1/38 0.1656460329       0.43591061\n",
       "95                Mouse 60 Sst Chodl up    1/38 0.1656460329       0.43591061\n",
       "96               Mouse 24 Ndnf HPF down    1/39 0.1696167528       0.44171029\n",
       "97          Human Inh L1-6 LAMP5 CA1 up    1/40 0.1735687740       0.44464968\n",
       "98                Mouse 61 Sst Chodl up   2/160 0.1763492568       0.44464968\n",
       "99             Mouse 119 Pvalb Vipr2 up    1/41 0.1775021834       0.44464968\n",
       "100         Human Inh L1-6 SST NPY down    1/42 0.1814170679       0.44464968\n",
       "101            Mouse 231 L6 IT CTX down    1/42 0.1814170679       0.44464968\n",
       "102               Mouse 276 CA1-ProS up    1/42 0.1814170679       0.44464968\n",
       "103                    Mouse 36 Sncg up    1/43 0.1853135138       0.44687832\n",
       "104           Mouse 166 L2/3 IT ENTl up    1/44 0.1891916069       0.44687832\n",
       "105      Human Astro L1-6 FGFR3 AQP1 up    1/45 0.1930514329       0.44687832\n",
       "106         Human Inh L1-2 VIP HTR3A up    1/45 0.1930514329       0.44687832\n",
       "107               Mouse 4 Lamp5 Lhx6 up    1/45 0.1930514329       0.44687832\n",
       "108                    Mouse 19 Pax6 up    1/45 0.1930514329       0.44687832\n",
       "109                   Mouse 375 VLMC up    1/46 0.1968930767       0.44802818\n",
       "110                  Mouse 365 Oligo up   3/323 0.1985310789       0.44802818\n",
       "111                    Mouse 9 Lamp5 up    1/47 0.2007166233       0.44802818\n",
       "112            Mouse 306 L5 NP CTX down    1/47 0.2007166233       0.44802818\n",
       "113                Mouse 260 L6 Car3 up   2/177 0.2055172707       0.45284731\n",
       "114                     Mouse 92 Sst up    1/49 0.2083097614       0.45284731\n",
       "115              Mouse 303 L5 NP CTX up    1/49 0.2083097614       0.45284731\n",
       "116   Human Inh L5-6 PVALB SST CRHR2 up    1/50 0.2120795206       0.45316137\n",
       "117           Mouse 161 L2/3 IT ENTl up    1/50 0.2120795206       0.45316137\n",
       "118            Mouse 309 L5 NP CTX down    1/51 0.2158315174       0.45727016\n",
       "119    Human OPC L1-6 PDGFRA COL20A1 up    1/52 0.2195658348       0.46127276\n",
       "120                Mouse 258 L6 Car3 up    1/54 0.2269817607       0.47287867\n",
       "121                Mouse 23 Ndnf HPF up    1/55 0.2306635328       0.47657755\n",
       "122            Mouse 303 L5 NP CTX down    1/58 0.2416050601       0.49509234\n",
       "123                       Mouse 1 CR up    1/59 0.2452179080       0.49841038\n",
       "124        Human Exc L5 FEZF2 CSN1S1 up 11/1834 0.2526858757       0.50944733\n",
       "125                    Mouse 373 SMC up    1/63 0.2594997861       0.51488053\n",
       "126                    Mouse 289 CA1 up    1/63 0.2594997861       0.51488053\n",
       "127                    Mouse 26 Sncg up   2/211 0.2650611409       0.52177390\n",
       "128                    Mouse 291 CA3 up    1/67 0.2735142155       0.53420745\n",
       "129      Human Inh L1-6 VIP SLC7A6OS up    1/70 0.2838525092       0.54452817\n",
       "130      Human Inh L5-6 LAMP5 CRABP1 up    1/70 0.2838525092       0.54452817\n",
       "131                   Mouse 370 Endo up    1/71 0.2872661523       0.54452817\n",
       "132                  Mouse 378 Micro up    1/72 0.2906636941       0.54452817\n",
       "133                   Mouse 10 Lamp5 up    1/73 0.2940452098       0.54452817\n",
       "134                   Mouse 272 ProS up    1/73 0.2940452098       0.54452817\n",
       "135               Mouse 278 CA1-ProS up    1/73 0.2940452098       0.54452817\n",
       "136                  Mouse 361 Oligo up    1/74 0.2974107742       0.54671098\n",
       "137        Human Inh L5-6 SST FBN2 down    1/76 0.3040943463       0.55290018\n",
       "138        Human Inh L3-5 SST CDH3 down    1/77 0.3074125019       0.55290018\n",
       "139                    Mouse 296 CA3 up    1/77 0.3074125019       0.55290018\n",
       "140                     Mouse 89 Sst up    1/80 0.3172733272       0.56655951\n",
       "141   Human Oligo L2-6 OPALIN FTH1P3 up    1/81 0.3205292976       0.56831436\n",
       "142                Mouse 366 Oligo down   5/795 0.3264996909       0.57166995\n",
       "143                    Mouse 379 PVM up    1/83 0.3269952141       0.57166995\n",
       "144           Mouse 234 L3 RSP-ACA down   4/610 0.3293677765       0.57181906\n",
       "145      Human Exc L5-6 THEMIS SMYD1 up   3/428 0.3323478295       0.57301350\n",
       "146                    Mouse 281 CA1 up    1/87 0.3397449481       0.58058707\n",
       "147              Mouse 148 L2 IT PAR up    1/88 0.3428948565       0.58058707\n",
       "148                Mouse 347 L6b CTX up    1/89 0.3460298948       0.58058707\n",
       "149                    Mouse 37 Sncg up    1/89 0.3460298948       0.58058707\n",
       "150                   Mouse 120 DG down   3/442 0.3505738162       0.58428969\n",
       "151          Mouse 181 L2/3 IT CTX down    1/93 0.3584227326       0.59271350\n",
       "152                    Mouse 299 CA3 up  6/1028 0.3628556831       0.59271350\n",
       "153                    Mouse 360 OPC up    1/95 0.3645317221       0.59271350\n",
       "154                Mouse 368 Oligo down   4/643 0.3651115151       0.59271350\n",
       "155                  Mouse 377 Micro up    1/98 0.3735874448       0.59869783\n",
       "156             Mouse 138 L2 IT ENTl up    1/98 0.3735874448       0.59869783\n",
       "157                     Mouse 123 DG up   1/101 0.3825154583       0.60910105\n",
       "158   Human Oligo L2-6 OPALIN MAP6D1 up   1/102 0.3854633916       0.60991043\n",
       "159                  Mouse 373 SMC down   3/489 0.4112245202       0.64657943\n",
       "160   Human Endo L2-5 NOSTRIN SRGN down   4/703 0.4298035406       0.67156803\n",
       "161                    Mouse 284 CA1 up   2/309 0.4329204897       0.67223679\n",
       "162             Mouse 6 Lamp5 Lhx6 down   2/313 0.4393992382       0.67808524\n",
       "163               Mouse 267 SUB-ProS up   1/124 0.4469072266       0.68544053\n",
       "164          Human Inh L1-2 VIP WNT4 up   3/523 0.4540720570       0.69218301\n",
       "165                Mouse 349 L6b CTX up   1/129 0.4600006578       0.69697069\n",
       "166 Human VLMC L1-5 PDGFRA COLEC12 down   4/742 0.4710257353       0.70937611\n",
       "167                    Mouse 29 Sncg up   5/956 0.4783840670       0.71117251\n",
       "168                    Mouse 294 CA3 up  7/1373 0.4799780639       0.71117251\n",
       "169                    Mouse 283 CA1 up   1/138 0.4828005414       0.71117251\n",
       "170                     Mouse 71 Sst up   2/341 0.4835973036       0.71117251\n",
       "171  Human Astro L1 FGFR3 SERPINI2 down   5/976 0.4967354944       0.72213363\n",
       "172                    Mouse 298 CA3 up   3/558 0.4968279397       0.72213363\n",
       "173                    Mouse 38 Sncg up   1/150 0.5117248768       0.73948682\n",
       "174                Mouse 20 Ndnf HPF up   1/157 0.5278520754       0.75407439\n",
       "175                    Mouse 27 Sncg up   1/157 0.5278520754       0.75407439\n",
       "176    Human Oligo L3-6 OPALIN ENPP6 up   1/171 0.5585418048       0.78595295\n",
       "177          Human Exc L6 FEZF2 KLK7 up   1/175 0.5669418477       0.78595295\n",
       "178                Mouse 345 L6b CTX up   3/622 0.5704779818       0.78595295\n",
       "179        Human Exc L5-6 FEZF2 CFTR up   1/177 0.5710823641       0.78595295\n",
       "180          Mouse 132 L2/3 IT PPP down   1/177 0.5710823641       0.78595295\n",
       "181           Human Inh L1-6 SST NPY up   1/178 0.5731379085       0.78595295\n",
       "182                Mouse 365 Oligo down   3/625 0.5737686874       0.78595295\n",
       "183                Mouse 261 L6 Car3 up   2/406 0.5775764149       0.78595295\n",
       "184                     Mouse 98 Sst up   1/182 0.5812630644       0.78595295\n",
       "185      Human Inh L5-6 PVALB KCNIP2 up   2/409 0.5816051820       0.78595295\n",
       "186                  Mouse 109 Pvalb up   1/186 0.5892351577       0.79198274\n",
       "187                 Mouse 376 VLMC down   2/422 0.5987412848       0.79581766\n",
       "188          Mouse 129 L2/3 IT APr down   1/191 0.5989893720       0.79581766\n",
       "189                Mouse 357 Astro down   3/651 0.6016381532       0.79581766\n",
       "190                Mouse 358 Astro down   3/656 0.6068616699       0.79850220\n",
       "191               Mouse 6 Lamp5 Lhx6 up   1/198 0.6122613333       0.80138918\n",
       "192                   Mouse 270 ProS up   3/673 0.6242868157       0.81228205\n",
       "193                    Mouse 282 CA1 up   2/446 0.6290010555       0.81228205\n",
       "194                    Mouse 300 CA3 up   4/907 0.6303308696       0.81228205\n",
       "195                  Mouse 363 Oligo up   1/212 0.6375150047       0.81732693\n",
       "196                Mouse 367 Oligo down   3/698 0.6489589837       0.82775381\n",
       "197    Human Astro L1-6 FGFR3 AQP1 down   3/707 0.6575602593       0.83446733\n",
       "198                     Mouse 97 Sst up  5/1184 0.6695201668       0.84270421\n",
       "199      Human Exc L5-6 FEZF2 SH2D1B up   1/232 0.6707925535       0.84270421\n",
       "200                  Mouse 111 Pvalb up   2/486 0.6754986105       0.84437326\n",
       "201                    Mouse 292 CA3 up   3/731 0.6797659983       0.84548010\n",
       "202              Mouse 238 L5 PT CTX up   1/240 0.6832405165       0.84559470\n",
       "203   Human Exc L3-5 FEZF2 LINC01107 up  7/1687 0.6996524511       0.86164095\n",
       "204       Human Inh L5-6 SST C4orf26 up   2/519 0.7102489884       0.87040317\n",
       "205          Human Exc L6 FEZF2 POGK up   1/264 0.7178573213       0.87543576\n",
       "206                Mouse 363 Oligo down   3/784 0.7250444448       0.87990831\n",
       "207                Mouse 21 Ndnf HPF up   1/277 0.7350166464       0.88770126\n",
       "208              Mouse 229 L6 IT CTX up   1/281 0.7400856262       0.88860364\n",
       "209             Mouse 236 L3 RSP-ACA up   2/555 0.7446012466       0.88860364\n",
       "210                    Mouse 287 CA1 up   2/558 0.7473026102       0.88860364\n",
       "211                  Mouse 367 Oligo up   1/291 0.7523421479       0.88860364\n",
       "212      Human Exc L5 FEZF2 NREP-AS1 up   1/292 0.7535358845       0.88860364\n",
       "213                  Mouse 355 V3d down   2/576 0.7630070723       0.89554821\n",
       "214                    Mouse 28 Sncg up   1/304 0.7674242110       0.89604077\n",
       "215  Human Oligo L5-6 OPALIN LDLRAP1 up   1/308 0.7718793032       0.89604077\n",
       "216          Human Exc L6 FEZF2 PDYN up  4/1103 0.7756775106       0.89604077\n",
       "217                  Mouse 364 Oligo up   1/318 0.7826509850       0.89604077\n",
       "218  Human OPC L1-6 PDGFRA COL20A1 down  4/1117 0.7841173190       0.89604077\n",
       "219         Human Exc L3-5 RORB RPRM up   1/327 0.7919143059       0.89604077\n",
       "220                    Mouse 355 V3d up   1/327 0.7919143059       0.89604077\n",
       "221   Human Astro L1-6 FGFR3 PLCG1 down   3/880 0.7943070536       0.89604077\n",
       "222                Mouse 356 Astro down   2/617 0.7956842047       0.89604077\n",
       "223        Human Exc L6 FEZF2 PROKR2 up   2/640 0.8122397971       0.90597265\n",
       "224                Mouse 361 Oligo down   3/913 0.8145518244       0.90597265\n",
       "225                Mouse 364 Oligo down   2/650 0.8190619599       0.90597265\n",
       "226                Mouse 378 Micro down   1/365 0.8269062752       0.90597265\n",
       "227                  Mouse 360 OPC down   3/937 0.8282100724       0.90597265\n",
       "228                    Mouse 295 CA3 up   2/667 0.8301576384       0.90597265\n",
       "229                   Mouse 121 DG down   1/372 0.8326850360       0.90597265\n",
       "230                   Mouse 122 DG down   1/373 0.8334948364       0.90597265\n",
       "231                    Mouse 285 CA1 up   1/409 0.8602019152       0.92679742\n",
       "232                    Mouse 293 CA3 up   1/417 0.8655348613       0.92679742\n",
       "233                 Mouse 371 Peri down   1/419 0.8668363332       0.92679742\n",
       "234                  Mouse 379 PVM down   1/420 0.8674823870       0.92679742\n",
       "235       Human Exc L3-5 FEZF2 ASGR2 up   2/759 0.8803019284       0.93546473\n",
       "236                     Mouse 96 Sst up  8/2354 0.8841243336       0.93546473\n",
       "237  Human Oligo L3-6 OPALIN ENPP6 down   2/774 0.8870641722       0.93546473\n",
       "238                  Mouse 372 SMC down   1/466 0.8940750581       0.93546473\n",
       "239                Mouse 362 Oligo down   2/791 0.8943042861       0.93546473\n",
       "240                 Mouse 374 VLMC down   1/513 0.9157885371       0.95394639\n",
       "241                 Mouse 375 VLMC down   1/535 0.9243768603       0.95889716\n",
       "242 Human Oligo L2-6 OPALIN FTH1P3 down   2/912 0.9346953489       0.96559437\n",
       "243                 Mouse 370 Endo down   1/607 0.9468632956       0.96754826\n",
       "244                 Mouse 369 Endo down   1/611 0.9478969059       0.96754826\n",
       "245       Human Inh L3-5 PVALB ISG20 up  7/2402 0.9481972976       0.96754826\n",
       "246                  Mouse 110 Pvalb up   1/631 0.9527739470       0.96826621\n",
       "247   Human Micro L1-6 TYROBP CD74 down   1/704 0.9670375785       0.97815598\n",
       "248    Human Exc L5 THEMIS LINC01116 up  6/2315 0.9703307361       0.97815598\n",
       "249                     Mouse 1 CR down   1/769 0.9760957742       0.98001584\n",
       "250                  Mouse 359 OPC down   1/841 0.9832754179       0.98327542\n",
       "    Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score\n",
       "1             0                    0 12.6364071   9.872277e+01\n",
       "2             0                    0  4.8530161   3.370506e+01\n",
       "3             0                    0 15.7985154   1.063308e+02\n",
       "4             0                    0  9.0700549   6.012018e+01\n",
       "5             0                    0  9.3742911   4.988418e+01\n",
       "6             0                    0 17.8145161   8.907544e+01\n",
       "7             0                    0  2.9688233   1.426063e+01\n",
       "8             0                    0  7.6944876   3.691334e+01\n",
       "9             0                    0 15.8327360   7.578916e+01\n",
       "10            0                    0  6.5906167   2.894849e+01\n",
       "11            0                    0 12.2089094   5.269972e+01\n",
       "12            0                    0 11.8691756   5.062792e+01\n",
       "13            0                    0 11.2433503   4.685877e+01\n",
       "14            0                    0  4.3973894   1.832387e+01\n",
       "15            0                    0 10.4190926   4.199357e+01\n",
       "16            0                    0  9.2842450   3.549600e+01\n",
       "17            0                    0  2.4396432   9.206184e+00\n",
       "18            0                    0  8.8965054   3.333398e+01\n",
       "19            0                    0  8.8965054   3.333398e+01\n",
       "20            0                    0 42.3404255   1.511386e+02\n",
       "21            0                    0 35.2819149   1.205864e+02\n",
       "22            0                    0 35.2819149   1.205864e+02\n",
       "23            0                    0 35.2819149   1.205864e+02\n",
       "24            0                    0  6.8827610   2.265327e+01\n",
       "25            0                    0  6.6670027   2.157009e+01\n",
       "26            0                    0  6.4643206   2.056467e+01\n",
       "27            0                    0 26.4587766   8.390514e+01\n",
       "28            0                    0 26.4587766   8.390514e+01\n",
       "29            0                    0 23.5177305   7.215575e+01\n",
       "30            0                    0 23.5177305   7.215575e+01\n",
       "31            0                    0 23.5177305   7.215575e+01\n",
       "32            0                    0  5.9238351   1.794390e+01\n",
       "33            0                    0 21.1648936   6.296914e+01\n",
       "34            0                    0 21.1648936   6.296914e+01\n",
       "35            0                    0 19.2398453   5.561263e+01\n",
       "36            0                    0 19.2398453   5.561263e+01\n",
       "37            0                    0  3.5534470   1.015570e+01\n",
       "38            0                    0 17.6356383   4.960523e+01\n",
       "39            0                    0 17.6356383   4.960523e+01\n",
       "40            0                    0 17.6356383   4.960523e+01\n",
       "41            0                    0 16.2782324   4.461874e+01\n",
       "42            0                    0 16.2782324   4.461874e+01\n",
       "43            0                    0 15.1147416   4.042201e+01\n",
       "44            0                    0 14.1063830   3.684775e+01\n",
       "45            0                    0 14.1063830   3.684775e+01\n",
       "46            0                    0  4.6313698   1.207857e+01\n",
       "47            0                    0  4.5813389   1.186436e+01\n",
       "48            0                    0 12.4455569   3.110145e+01\n",
       "49            0                    0  4.2167572   1.033589e+01\n",
       "50            0                    0 11.7535461   2.876393e+01\n",
       "51            0                    0  4.0944996   9.836693e+00\n",
       "52            0                    0 11.1343785   2.670340e+01\n",
       "53            0                    0  4.0168391   9.523233e+00\n",
       "54            0                    0 10.5771277   2.487542e+01\n",
       "55            0                    0 10.0729483   2.324443e+01\n",
       "56            0                    0  2.7772445   6.383586e+00\n",
       "57            0                    0  3.7666762   8.533484e+00\n",
       "58            0                    0  9.1961147   2.046345e+01\n",
       "59            0                    0  9.1961147   2.046345e+01\n",
       "60            0                    0  3.6371657   8.033551e+00\n",
       "61            0                    0  3.6061600   7.915172e+00\n",
       "62            0                    0  8.8125000   1.927045e+01\n",
       "63            0                    0  8.8125000   1.927045e+01\n",
       "64            0                    0  1.9998967   4.300839e+00\n",
       "65            0                    0  8.4595745   1.818645e+01\n",
       "66            0                    0  8.4595745   1.818645e+01\n",
       "67            0                    0  2.1543984   4.487653e+00\n",
       "68            0                    0  7.8321513   1.629334e+01\n",
       "69            0                    0  7.8321513   1.629334e+01\n",
       "70            0                    0  3.3490814   6.953901e+00\n",
       "71            0                    0  7.5520517   1.546304e+01\n",
       "72            0                    0  3.2712986   6.670431e+00\n",
       "73            0                    0  7.2912693   1.469869e+01\n",
       "74            0                    0  7.2912693   1.469869e+01\n",
       "75            0                    0  7.0478723   1.399314e+01\n",
       "76            0                    0  7.0478723   1.399314e+01\n",
       "77            0                    0  3.1260278   6.150635e+00\n",
       "78            0                    0  3.1260278   6.150635e+00\n",
       "79            0                    0  6.8201784   1.334021e+01\n",
       "80            0                    0  6.8201784   1.334021e+01\n",
       "81            0                    0  6.8201784   1.334021e+01\n",
       "82            0                    0  3.0580955   5.911999e+00\n",
       "83            0                    0  6.6067154   1.273456e+01\n",
       "84            0                    0  6.6067154   1.273456e+01\n",
       "85            0                    0  2.3191163   4.442684e+00\n",
       "86            0                    0  6.4061896   1.217151e+01\n",
       "87            0                    0  6.4061896   1.217151e+01\n",
       "88            0                    0  6.4061896   1.217151e+01\n",
       "89            0                    0  6.2174593   1.164697e+01\n",
       "90            0                    0  6.2174593   1.164697e+01\n",
       "91            0                    0  6.0395137   1.115734e+01\n",
       "92            0                    0  5.8714539   1.069944e+01\n",
       "93            0                    0  5.7124784   1.027048e+01\n",
       "94            0                    0  5.7124784   1.027048e+01\n",
       "95            0                    0  5.7124784   1.027048e+01\n",
       "96            0                    0  5.5618701   9.867947e+00\n",
       "97            0                    0  5.4189853   9.489626e+00\n",
       "98            0                    0  2.6877637   4.664046e+00\n",
       "99            0                    0  5.2832447   9.133527e+00\n",
       "100           0                    0  5.1541256   8.797869e+00\n",
       "101           0                    0  5.1541256   8.797869e+00\n",
       "102           0                    0  5.1541256   8.797869e+00\n",
       "103           0                    0  5.0311550   8.481049e+00\n",
       "104           0                    0  4.9139040   8.181626e+00\n",
       "105           0                    0  4.8019826   7.898294e+00\n",
       "106           0                    0  4.8019826   7.898294e+00\n",
       "107           0                    0  4.8019826   7.898294e+00\n",
       "108           0                    0  4.8019826   7.898294e+00\n",
       "109           0                    0  4.6950355   7.629876e+00\n",
       "110           0                    0  1.9957541   3.226754e+00\n",
       "111           0                    0  4.5927382   7.375300e+00\n",
       "112           0                    0  4.5927382   7.375300e+00\n",
       "113           0                    0  2.4245776   3.836228e+00\n",
       "114           0                    0  4.4009309   6.903868e+00\n",
       "115           0                    0  4.4009309   6.903868e+00\n",
       "116           0                    0  4.3108988   6.685316e+00\n",
       "117           0                    0  4.3108988   6.685316e+00\n",
       "118           0                    0  4.2244681   6.477196e+00\n",
       "119           0                    0  4.1414268   6.278830e+00\n",
       "120           0                    0  3.9847451   5.908921e+00\n",
       "121           0                    0  3.9107565   5.736279e+00\n",
       "122           0                    0  3.7043673   5.261872e+00\n",
       "123           0                    0  3.6403155   5.116857e+00\n",
       "124           0                    0  1.2988925   1.786767e+00\n",
       "125           0                    0  3.4047701   4.593033e+00\n",
       "126           0                    0  3.4047701   4.593033e+00\n",
       "127           0                    0  2.0266502   2.690976e+00\n",
       "128           0                    0  3.1977756   4.145602e+00\n",
       "129           0                    0  3.0582794   3.851293e+00\n",
       "130           0                    0  3.0582794   3.851293e+00\n",
       "131           0                    0  3.0144377   3.760047e+00\n",
       "132           0                    0  2.9718310   3.671960e+00\n",
       "133           0                    0  2.9304078   3.586883e+00\n",
       "134           0                    0  2.9304078   3.586883e+00\n",
       "135           0                    0  2.9304078   3.586883e+00\n",
       "136           0                    0  2.8901195   3.504677e+00\n",
       "137           0                    0  2.8127660   3.348365e+00\n",
       "138           0                    0  2.7756159   3.274019e+00\n",
       "139           0                    0  2.7756159   3.274019e+00\n",
       "140           0                    0  2.6698088   3.064918e+00\n",
       "141           0                    0  2.6363032   2.999537e+00\n",
       "142           0                    0  1.3442335   1.504636e+00\n",
       "143           0                    0  2.5717436   2.874720e+00\n",
       "144           0                    0  1.3998477   1.554643e+00\n",
       "145           0                    0  1.4946292   1.646443e+00\n",
       "146           0                    0  2.4516329   2.646685e+00\n",
       "147           0                    0  2.4233309   2.593767e+00\n",
       "148           0                    0  2.3956721   2.542359e+00\n",
       "149           0                    0  2.3956721   2.542359e+00\n",
       "150           0                    0  1.4459245   1.515595e+00\n",
       "151           0                    0  2.2910500   2.350714e+00\n",
       "152           0                    0  1.2456079   1.262735e+00\n",
       "153           0                    0  2.2420779   2.262574e+00\n",
       "154           0                    0  1.3252850   1.335294e+00\n",
       "155           0                    0  2.1724062   2.138958e+00\n",
       "156           0                    0  2.1724062   2.138958e+00\n",
       "157           0                    0  2.1069149   2.024716e+00\n",
       "158           0                    0  2.0859490   1.988554e+00\n",
       "159           0                    0  1.3029388   1.157812e+00\n",
       "160           0                    0  1.2077536   1.019860e+00\n",
       "161           0                    0  1.3728416   1.149345e+00\n",
       "162           0                    0  1.3549079   1.114204e+00\n",
       "163           0                    0  1.7109497   1.378006e+00\n",
       "164           0                    0  1.2156145   9.597269e-01\n",
       "165           0                    0  1.6437001   1.276378e+00\n",
       "166           0                    0  1.1416064   8.594498e-01\n",
       "167           0                    0  1.1072555   8.164253e-01\n",
       "168           0                    0  1.0795704   7.924207e-01\n",
       "169           0                    0  1.5350210   1.117728e+00\n",
       "170           0                    0  1.2412218   9.017510e-01\n",
       "171           0                    0  1.0833047   7.579857e-01\n",
       "172           0                    0  1.1368978   7.952731e-01\n",
       "173           0                    0  1.4105383   9.450158e-01\n",
       "174           0                    0  1.3467676   8.605026e-01\n",
       "175           0                    0  1.3467676   8.605026e-01\n",
       "176           0                    0  1.2349812   7.192849e-01\n",
       "177           0                    0  1.2063463   6.845998e-01\n",
       "178           0                    0  1.0159795   5.702497e-01\n",
       "179           0                    0  1.1925169   6.680740e-01\n",
       "180           0                    0  1.1925169   6.680740e-01\n",
       "181           0                    0  1.1857194   6.600057e-01\n",
       "182           0                    0  1.0109220   5.615964e-01\n",
       "183           0                    0  1.0380603   5.698064e-01\n",
       "184           0                    0  1.1592806   6.289698e-01\n",
       "185           0                    0  1.0302502   5.583579e-01\n",
       "186           0                    0  1.1339850   5.997986e-01\n",
       "187           0                    0  0.9976959   5.117438e-01\n",
       "188           0                    0  1.1038634   5.657426e-01\n",
       "189           0                    0  0.9690519   4.923744e-01\n",
       "190           0                    0  0.9613822   4.801666e-01\n",
       "191           0                    0  1.0642618   5.221227e-01\n",
       "192           0                    0  0.9361616   4.410682e-01\n",
       "193           0                    0  0.9426039   4.370122e-01\n",
       "194           0                    0  0.9249754   4.268858e-01\n",
       "195           0                    0  0.9929414   4.469998e-01\n",
       "196           0                    0  0.9013137   3.897152e-01\n",
       "197           0                    0  0.8893744   3.728425e-01\n",
       "198           0                    0  0.8823862   3.540080e-01\n",
       "199           0                    0  0.9060514   3.617821e-01\n",
       "200           0                    0  0.8629254   3.385293e-01\n",
       "201           0                    0  0.8589793   3.315717e-01\n",
       "202           0                    0  0.8753672   3.334347e-01\n",
       "203           0                    0  0.8629261   3.082127e-01\n",
       "204           0                    0  0.8064724   2.759262e-01\n",
       "205           0                    0  0.7945150   2.633694e-01\n",
       "206           0                    0  0.7984746   2.567274e-01\n",
       "207           0                    0  0.7565911   2.329258e-01\n",
       "208           0                    0  0.7456307   2.244269e-01\n",
       "209           0                    0  0.7525715   2.219382e-01\n",
       "210           0                    0  0.7483948   2.179962e-01\n",
       "211           0                    0  0.7195525   2.047588e-01\n",
       "212           0                    0  0.7170432   2.029079e-01\n",
       "213           0                    0  0.7242516   1.959014e-01\n",
       "214           0                    0  0.6882241   1.821836e-01\n",
       "215           0                    0  0.6791184   1.758422e-01\n",
       "216           0                    0  0.7521723   1.910656e-01\n",
       "217           0                    0  0.6573596   1.610981e-01\n",
       "218           0                    0  0.7421581   1.804904e-01\n",
       "219           0                    0  0.6389179   1.490609e-01\n",
       "220           0                    0  0.6389179   1.490609e-01\n",
       "221           0                    0  0.7075009   1.629270e-01\n",
       "222           0                    0  0.6745345   1.541668e-01\n",
       "223           0                    0  0.6494421   1.350578e-01\n",
       "224           0                    0  0.6806617   1.396154e-01\n",
       "225           0                    0  0.6390880   1.275591e-01\n",
       "226           0                    0  0.5711071   1.085469e-01\n",
       "227           0                    0  0.6623336   1.248422e-01\n",
       "228           0                    0  0.6222007   1.158162e-01\n",
       "229           0                    0  0.5601308   1.025598e-01\n",
       "230           0                    0  0.5585964   1.017359e-01\n",
       "231           0                    0  0.5083698   7.655446e-02\n",
       "232           0                    0  0.4983889   7.197116e-02\n",
       "233           0                    0  0.4959534   7.087426e-02\n",
       "234           0                    0  0.4947443   7.033289e-02\n",
       "235           0                    0  0.5439695   6.935086e-02\n",
       "236           0                    0  0.6882441   8.476248e-02\n",
       "237           0                    0  0.5329823   6.387151e-02\n",
       "238           0                    0  0.4447495   4.979662e-02\n",
       "239           0                    0  0.5210352   5.820442e-02\n",
       "240           0                    0  0.4029463   3.544710e-02\n",
       "241           0                    0  0.3859072   3.034598e-02\n",
       "242           0                    0  0.4488952   3.031597e-02\n",
       "243           0                    0  0.3387929   1.849828e-02\n",
       "244           0                    0  0.3365016   1.800604e-02\n",
       "245           0                    0  0.5815620   3.093484e-02\n",
       "246           0                    0  0.3254813   1.574600e-02\n",
       "247           0                    0  0.2905784   9.739584e-03\n",
       "248           0                    0  0.5137493   1.547326e-02\n",
       "249           0                    0  0.2650848   6.413612e-03\n",
       "250           0                    0  0.2414514   4.072323e-03\n",
       "                                                                       Genes\n",
       "1                                                  CCND2;COL4A2;FILIP1;SYNE2\n",
       "2                            EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN\n",
       "3                                                         EFNB2;TRIM24;CNTLN\n",
       "4                                                        TTYH1;ID4;SOX9;SAT1\n",
       "5                                                            TTYH1;SOX9;SAT1\n",
       "6                                                                 URI1;RAB8B\n",
       "7                              SOX2;RSL1D1;URI1;GAR1;GPATCH4;MPDZ;EIF3D;GNL3\n",
       "8                                                            EMX1;URI1;RAB8B\n",
       "9                                                                EFNB2;RAB8B\n",
       "10                                                       RFX4;CACHD1;ZFP36L1\n",
       "11                                                               FOXG1;RAB8B\n",
       "12                                                                EFNB2;PKIB\n",
       "13                                                                 ARX;RAB8B\n",
       "14                                                    MYO10;B3GAT2;GAR1;PAWR\n",
       "15                                                                YES1;SYNE2\n",
       "16                                                              FILIP1;CREB5\n",
       "17                             SOX2;RSL1D1;URI1;LIX1;GAR1;GPATCH4;EIF3D;GNL3\n",
       "18                                                                 URI1;YES1\n",
       "19                                                                MYO10;GAR1\n",
       "20                                                                     NCALD\n",
       "21                                                                     NCALD\n",
       "22                                                                     NCALD\n",
       "23                                                                     NCALD\n",
       "24                                                                TTYH1;RFX4\n",
       "25                                                                URI1;RAB8B\n",
       "26                                                                 URI1;LIX1\n",
       "27                                                                     NCALD\n",
       "28                                                                     NCALD\n",
       "29                                                                     NCALD\n",
       "30                                                                     NCALD\n",
       "31                                                                     NCALD\n",
       "32                                                                EFNB2;PKIB\n",
       "33                                                                       QKI\n",
       "34                                                                     NCALD\n",
       "35                                                                       QKI\n",
       "36                                                                     CCND2\n",
       "37                                                     EFNB2;LINC01551;FOXG1\n",
       "38                                                                     NCALD\n",
       "39                                                                      INTU\n",
       "40                                                                     NCALD\n",
       "41                                                                      MPDZ\n",
       "42                                                                     NCALD\n",
       "43                                                                      INTU\n",
       "44                                                                       QKI\n",
       "45                                                                     NCALD\n",
       "46                                                                 URI1;RPL3\n",
       "47                                                                EFNB2;PKIB\n",
       "48                                                                     NCALD\n",
       "49                                                               EFNB2;RAB8B\n",
       "50                                                                    TRIM24\n",
       "51                                                              TRIM24;RAB8B\n",
       "52                                                                       QKI\n",
       "53                                                                 URI1;MPDZ\n",
       "54                                                                     NCALD\n",
       "55                                                                   FAM171B\n",
       "56                                                          PKIB;NOC3L;CNTLN\n",
       "57                                                                RPL3;RAB8B\n",
       "58                                                                     NCALD\n",
       "59                                                                      MPDZ\n",
       "60                                                                INTU;RAB8B\n",
       "61                                                                 ARX;RAB8B\n",
       "62                                                                    LIMCH1\n",
       "63                                                                     NCALD\n",
       "64                                         LINC01551;RPL3;RPS27L;ABCE1;RAB8B\n",
       "65                                                                     NCALD\n",
       "66                                                                     NCALD\n",
       "67                                                   SOX2;IQCB1;HEY1;ADAMTS6\n",
       "68                                                                       QKI\n",
       "69                                                                     MASP1\n",
       "70                                                                SOX2;SFRP1\n",
       "71                                                                     MASP1\n",
       "72                                                         FAM171B;RAB11FIP2\n",
       "73                                                                     NCALD\n",
       "74                                                                      MPDZ\n",
       "75                                                                     NCALD\n",
       "76                                                                    LIMCH1\n",
       "77                                                                 URI1;LIX1\n",
       "78                                                              COL4A2;MYO10\n",
       "79                                                                      SOX9\n",
       "80                                                                     NCALD\n",
       "81                                                                   FAM171B\n",
       "82                                                               EFNB2;RAB8B\n",
       "83                                                                     EFNB2\n",
       "84                                                                     EFNB2\n",
       "85                                                        EFNB2;FOXG1;RPS27L\n",
       "86                                                                      URI1\n",
       "87                                                                      VCAN\n",
       "88                                                                      RPL3\n",
       "89                                                                      PKIB\n",
       "90                                                                     NCALD\n",
       "91                                                                     RAB8B\n",
       "92                                                                     NCALD\n",
       "93                                                                      URI1\n",
       "94                                                                     RAB8B\n",
       "95                                                                   ADAMTS6\n",
       "96                                                                     NCALD\n",
       "97                                                                    COL4A2\n",
       "98                                                                SOX2;SFRP1\n",
       "99                                                                      YES1\n",
       "100                                                                    NCALD\n",
       "101                                                                     RPL3\n",
       "102                                                                    EFNB2\n",
       "103                                                                     INTU\n",
       "104                                                                     PKIB\n",
       "105                                                                     RFX4\n",
       "106                                                                     INTU\n",
       "107                                                                     MPDZ\n",
       "108                                                                     MPDZ\n",
       "109                                                                   COL4A2\n",
       "110                                                         TTYH1;SAT1;CREB5\n",
       "111                                                                     MPDZ\n",
       "112                                                                  FAM171B\n",
       "113                                                             B3GAT2;RAB8B\n",
       "114                                                                     SOX2\n",
       "115                                                                    MYO10\n",
       "116                                                                     HEY1\n",
       "117                                                                     LIX1\n",
       "118                                                                  FAM171B\n",
       "119                                                                     VCAN\n",
       "120                                                                   B3GAT2\n",
       "121                                                                     MPDZ\n",
       "122                                                                     RPL3\n",
       "123                                                                     EMX2\n",
       "124 ZNF397;EFNB2;LINC01551;INTU;FOXG1;TRIM24;CACHD1;RPS27L;FEZF2;RAB8B;EIF3B\n",
       "125                                                                    CCND2\n",
       "126                                                                    RAB8B\n",
       "127                                                                URI1;LIX1\n",
       "128                                                                    CCND2\n",
       "129                                                                LINC01551\n",
       "130                                                                   COL4A2\n",
       "131                                                                   COL4A2\n",
       "132                                                                     SAT1\n",
       "133                                                                     MPDZ\n",
       "134                                                                    EFNB2\n",
       "135                                                                    RAB8B\n",
       "136                                                                     VCAN\n",
       "137                                                                  FAM171B\n",
       "138                                                                  FAM171B\n",
       "139                                                                    EFNB2\n",
       "140                                                                    MASP1\n",
       "141                                                                    CREB5\n",
       "142                                        URI1;RPL3;FAM171B;RAB11FIP2;NCALD\n",
       "143                                                                     SAT1\n",
       "144                                             RPL3;FAM171B;RAB11FIP2;NCALD\n",
       "145                                                  LINC01551;B3GAT2;CACHD1\n",
       "146                                                                    RAB8B\n",
       "147                                                                    EFNB2\n",
       "148                                                                     EMX1\n",
       "149                                                                     URI1\n",
       "150                                                       IQCB1;URI1;FAM171B\n",
       "151                                                                     RPL3\n",
       "152                                      RSL1D1;CCND2;GPC1;CACHD1;ABCE1;MPDZ\n",
       "153                                                                     VCAN\n",
       "154                                             RPL3;FAM171B;RAB11FIP2;NCALD\n",
       "155                                                                     SAT1\n",
       "156                                                                     PKIB\n",
       "157                                                                    RAB8B\n",
       "158                                                                     RPL3\n",
       "159                                                     RPL3;RAB11FIP2;NCALD\n",
       "160                                                   URI1;LIMCH1;MPDZ;NCALD\n",
       "161                                                            TRIM24;TRMT11\n",
       "162                                                          RAB11FIP2;NCALD\n",
       "163                                                                    EFNB2\n",
       "164                                                        INTU;GTPBP6;RAB8B\n",
       "165                                                                     PKIB\n",
       "166                                                   URI1;LIMCH1;MPDZ;NCALD\n",
       "167                                            URI1;LIX1;GPATCH4;PAICS;EIF3D\n",
       "168                                RSL1D1;COL4A2;MYO10;LIX1;GPC1;CACHD1;MPDZ\n",
       "169                                                                    RAB8B\n",
       "170                                                               URI1;MASP1\n",
       "171                                            URI1;LIMCH1;GPC1;FAM171B;MPDZ\n",
       "172                                                        CCND2;GPC1;CACHD1\n",
       "173                                                                     URI1\n",
       "174                                                                     URI1\n",
       "175                                                                     URI1\n",
       "176                                                                    CREB5\n",
       "177                                                                   FILIP1\n",
       "178                                                      IQCB1;EFNB2;GPATCH4\n",
       "179                                                                     GNL3\n",
       "180                                                                    IQCB1\n",
       "181                                                                   TRIM24\n",
       "182                                                       RPL3;FAM171B;NCALD\n",
       "183                                                              EFNB2;MYO10\n",
       "184                                                                     URI1\n",
       "185                                                                ARX;RAB8B\n",
       "186                                                                     YES1\n",
       "187                                                           RPL3;RAB11FIP2\n",
       "188                                                                    IQCB1\n",
       "189                                                        URI1;RPL3;FAM171B\n",
       "190                                                       RPL3;FAM171B;NCALD\n",
       "191                                                                   TRMT11\n",
       "192                                                       EFNB2;RSL1D1;RAB8B\n",
       "193                                                             RSL1D1;RAB8B\n",
       "194                                                  MYO10;GPC1;CACHD1;EIF3B\n",
       "195                                                                     SOX2\n",
       "196                                                       RPL3;FAM171B;NCALD\n",
       "197                                                      LIMCH1;FAM171B;MPDZ\n",
       "198                                            IQCB1;YES1;TRMT11;PAICS;CNTLN\n",
       "199                                                                     RPL3\n",
       "200                                                                URI1;YES1\n",
       "201                                                       RSL1D1;PLAGL1;MPDZ\n",
       "202                                                                     EMX1\n",
       "203                                    EMX1;DCT;INTU;PPAT;FOXG1;RPS27L;RAB8B\n",
       "204                                                               RSL1D1;ARX\n",
       "205                                                                    RAB8B\n",
       "206                                                       RPL3;FAM171B;NCALD\n",
       "207                                                                     URI1\n",
       "208                                                                    EIF3B\n",
       "209                                                                EMX1;PKIB\n",
       "210                                                             RSL1D1;RAB8B\n",
       "211                                                                    TTYH1\n",
       "212                                                                    ABCE1\n",
       "213                                                          RAB11FIP2;NCALD\n",
       "214                                                                  GPATCH4\n",
       "215                                                                  TMEM123\n",
       "216                                                 ZNF397;RSL1D1;PKIB;FOXG1\n",
       "217                                                                    TTYH1\n",
       "218                                                 URI1;LIMCH1;FAM171B;MPDZ\n",
       "219                                                                   RSL1D1\n",
       "220                                                                   RPS27L\n",
       "221                                                         URI1;LIMCH1;MPDZ\n",
       "222                                                          FAM171B;ZFP36L1\n",
       "223                                                              RSL1D1;RPL3\n",
       "224                                                   RPL3;FAM171B;RAB11FIP2\n",
       "225                                                               RPL3;NCALD\n",
       "226                                                                     RPL3\n",
       "227                                                   RPL3;FAM171B;RAB11FIP2\n",
       "228                                                               EFNB2;MPDZ\n",
       "229                                                                RAB11FIP2\n",
       "230                                                                RAB11FIP2\n",
       "231                                                                    RAB8B\n",
       "232                                                                   PLAGL1\n",
       "233                                                                    NCALD\n",
       "234                                                                     RPL3\n",
       "235                                                               INTU;RAB8B\n",
       "236                    URI1;TRIM24;RPS27L;GPATCH4;FGFBP3;PAICS;ADAMTS6;CNTLN\n",
       "237                                                           FAM171B;TRMT11\n",
       "238                                                                     RPL3\n",
       "239                                                             RPL3;FAM171B\n",
       "240                                                                     RPL3\n",
       "241                                                                     RPL3\n",
       "242                                                                URI1;MPDZ\n",
       "243                                                                     RPL3\n",
       "244                                                                     RPL3\n",
       "245                                ZNF397;PKIB;NOC3L;ABCE1;PAICS;CNTLN;EIF3B\n",
       "246                                                                    PAICS\n",
       "247                                                                     URI1\n",
       "248                                          DPP7;DCT;INTU;NOC3L;ABCE1;EIF3B\n",
       "249                                                                  FAM171B\n",
       "250                                                                  FAM171B\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"a\"\n"
     ]
    }
   ],
   "source": [
    "tic()\n",
    "\n",
    "genes_cluster <- top_genes %>% filter(cluster == 0) %>% pull(gene)\n",
    "# Optionally, use Entrez IDs instead:\n",
    "# genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "\n",
    "# Perform enrichment analysis\n",
    "enriched <- suppressMessages(suppressWarnings((enrichr(genes_cluster, databases = c(\n",
    "        \"Allen_Brain_Atlas_10x_scRNA_2021\"#, \n",
    "        # \"PanglaoDB_Augmented_2021\", \n",
    "        # \"CellMarker_Augmented_2021/2024\", \n",
    "        # \"Azimuth_Cell_Types_2021/2023\", \n",
    "        # \"Tabula_Sapiens/Tabula_Muris\", \n",
    "        # \"Descartes_Cell_Types_and_Tissue_2021\", \n",
    "        # \"HuBMAP_ASCT_plus_B_*\", \n",
    "        # \"Allen_Brain_Atlas_up/down\"\n",
    ")))))\n",
    "\n",
    "toc()\n",
    "#class(enriched)\n",
    "head(enriched)\n",
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "dfadab98-02ef-4e75-9d38-c15f00c38ff6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Mouse 372 SMC up             </td><td>4/73 </td><td>0.0004046182</td><td>0.08264024</td><td>0</td><td>0</td><td>12.636407</td><td> 98.72277</td><td>CCND2;COL4A2;FILIP1;SYNE2                      </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Human Inh L2 PAX6 FREM2 up   </td><td>7/328</td><td>0.0009632692</td><td>0.08264024</td><td>0</td><td>0</td><td> 4.853016</td><td> 33.70506</td><td>EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Human Inh L1 PAX6 MIR101-1 up</td><td>3/44 </td><td>0.0011940188</td><td>0.08264024</td><td>0</td><td>0</td><td>15.798515</td><td>106.33081</td><td>EFNB2;TRIM24;CNTLN                             </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Mouse 358 Astro up           </td><td>4/100</td><td>0.0013222439</td><td>0.08264024</td><td>0</td><td>0</td><td> 9.070055</td><td> 60.12018</td><td>TTYH1;ID4;SOX9;SAT1                            </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Mouse 356 Astro up           </td><td>3/72 </td><td>0.0048859972</td><td>0.23162675</td><td>0</td><td>0</td><td> 9.374291</td><td> 49.88418</td><td>TTYH1;SOX9;SAT1                                </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Mouse 334 L6b/CT ENT up      </td><td>2/26 </td><td>0.0067368642</td><td>0.23162675</td><td>0</td><td>0</td><td>17.814516</td><td> 89.07544</td><td>URI1;RAB8B                                     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 9\n",
       "\\begin{tabular}{r|lllllllll}\n",
       "  & Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       "  & <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & Mouse 372 SMC up              & 4/73  & 0.0004046182 & 0.08264024 & 0 & 0 & 12.636407 &  98.72277 & CCND2;COL4A2;FILIP1;SYNE2                      \\\\\n",
       "\t2 & Human Inh L2 PAX6 FREM2 up    & 7/328 & 0.0009632692 & 0.08264024 & 0 & 0 &  4.853016 &  33.70506 & EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN\\\\\n",
       "\t3 & Human Inh L1 PAX6 MIR101-1 up & 3/44  & 0.0011940188 & 0.08264024 & 0 & 0 & 15.798515 & 106.33081 & EFNB2;TRIM24;CNTLN                             \\\\\n",
       "\t4 & Mouse 358 Astro up            & 4/100 & 0.0013222439 & 0.08264024 & 0 & 0 &  9.070055 &  60.12018 & TTYH1;ID4;SOX9;SAT1                            \\\\\n",
       "\t5 & Mouse 356 Astro up            & 3/72  & 0.0048859972 & 0.23162675 & 0 & 0 &  9.374291 &  49.88418 & TTYH1;SOX9;SAT1                                \\\\\n",
       "\t6 & Mouse 334 L6b/CT ENT up       & 2/26  & 0.0067368642 & 0.23162675 & 0 & 0 & 17.814516 &  89.07544 & URI1;RAB8B                                     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 9\n",
       "\n",
       "| <!--/--> | Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Mouse 372 SMC up              | 4/73  | 0.0004046182 | 0.08264024 | 0 | 0 | 12.636407 |  98.72277 | CCND2;COL4A2;FILIP1;SYNE2                       |\n",
       "| 2 | Human Inh L2 PAX6 FREM2 up    | 7/328 | 0.0009632692 | 0.08264024 | 0 | 0 |  4.853016 |  33.70506 | EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN |\n",
       "| 3 | Human Inh L1 PAX6 MIR101-1 up | 3/44  | 0.0011940188 | 0.08264024 | 0 | 0 | 15.798515 | 106.33081 | EFNB2;TRIM24;CNTLN                              |\n",
       "| 4 | Mouse 358 Astro up            | 4/100 | 0.0013222439 | 0.08264024 | 0 | 0 |  9.070055 |  60.12018 | TTYH1;ID4;SOX9;SAT1                             |\n",
       "| 5 | Mouse 356 Astro up            | 3/72  | 0.0048859972 | 0.23162675 | 0 | 0 |  9.374291 |  49.88418 | TTYH1;SOX9;SAT1                                 |\n",
       "| 6 | Mouse 334 L6b/CT ENT up       | 2/26  | 0.0067368642 | 0.23162675 | 0 | 0 | 17.814516 |  89.07544 | URI1;RAB8B                                      |\n",
       "\n"
      ],
      "text/plain": [
       "  Term                          Overlap P.value      Adjusted.P.value\n",
       "1 Mouse 372 SMC up              4/73    0.0004046182 0.08264024      \n",
       "2 Human Inh L2 PAX6 FREM2 up    7/328   0.0009632692 0.08264024      \n",
       "3 Human Inh L1 PAX6 MIR101-1 up 3/44    0.0011940188 0.08264024      \n",
       "4 Mouse 358 Astro up            4/100   0.0013222439 0.08264024      \n",
       "5 Mouse 356 Astro up            3/72    0.0048859972 0.23162675      \n",
       "6 Mouse 334 L6b/CT ENT up       2/26    0.0067368642 0.23162675      \n",
       "  Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score\n",
       "1 0           0                    12.636407   98.72277     \n",
       "2 0           0                     4.853016   33.70506     \n",
       "3 0           0                    15.798515  106.33081     \n",
       "4 0           0                     9.070055   60.12018     \n",
       "5 0           0                     9.374291   49.88418     \n",
       "6 0           0                    17.814516   89.07544     \n",
       "  Genes                                          \n",
       "1 CCND2;COL4A2;FILIP1;SYNE2                      \n",
       "2 EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN\n",
       "3 EFNB2;TRIM24;CNTLN                             \n",
       "4 TTYH1;ID4;SOX9;SAT1                            \n",
       "5 TTYH1;SOX9;SAT1                                \n",
       "6 URI1;RAB8B                                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(enriched$Allen_Brain_Atlas_10x_scRNA_2021)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e9d49acd-a9e6-420f-b46f-6f8dc8adf634",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "250"
      ],
      "text/latex": [
       "250"
      ],
      "text/markdown": [
       "250"
      ],
      "text/plain": [
       "[1] 250"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th><th scope=col>cluster</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Mouse 372 SMC up             </td><td>4/73 </td><td>0.0004046182</td><td>0.08264024</td><td>0</td><td>0</td><td>12.636407</td><td> 98.72277</td><td>CCND2;COL4A2;FILIP1;SYNE2                      </td><td>0</td></tr>\n",
       "\t<tr><td>Human Inh L2 PAX6 FREM2 up   </td><td>7/328</td><td>0.0009632692</td><td>0.08264024</td><td>0</td><td>0</td><td> 4.853016</td><td> 33.70506</td><td>EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN</td><td>0</td></tr>\n",
       "\t<tr><td>Human Inh L1 PAX6 MIR101-1 up</td><td>3/44 </td><td>0.0011940188</td><td>0.08264024</td><td>0</td><td>0</td><td>15.798515</td><td>106.33081</td><td>EFNB2;TRIM24;CNTLN                             </td><td>0</td></tr>\n",
       "\t<tr><td>Mouse 358 Astro up           </td><td>4/100</td><td>0.0013222439</td><td>0.08264024</td><td>0</td><td>0</td><td> 9.070055</td><td> 60.12018</td><td>TTYH1;ID4;SOX9;SAT1                            </td><td>0</td></tr>\n",
       "\t<tr><td>Mouse 356 Astro up           </td><td>3/72 </td><td>0.0048859972</td><td>0.23162675</td><td>0</td><td>0</td><td> 9.374291</td><td> 49.88418</td><td>TTYH1;SOX9;SAT1                                </td><td>0</td></tr>\n",
       "\t<tr><td>Mouse 334 L6b/CT ENT up      </td><td>2/26 </td><td>0.0067368642</td><td>0.23162675</td><td>0</td><td>0</td><td>17.814516</td><td> 89.07544</td><td>URI1;RAB8B                                     </td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes & cluster\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Mouse 372 SMC up              & 4/73  & 0.0004046182 & 0.08264024 & 0 & 0 & 12.636407 &  98.72277 & CCND2;COL4A2;FILIP1;SYNE2                       & 0\\\\\n",
       "\t Human Inh L2 PAX6 FREM2 up    & 7/328 & 0.0009632692 & 0.08264024 & 0 & 0 &  4.853016 &  33.70506 & EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN & 0\\\\\n",
       "\t Human Inh L1 PAX6 MIR101-1 up & 3/44  & 0.0011940188 & 0.08264024 & 0 & 0 & 15.798515 & 106.33081 & EFNB2;TRIM24;CNTLN                              & 0\\\\\n",
       "\t Mouse 358 Astro up            & 4/100 & 0.0013222439 & 0.08264024 & 0 & 0 &  9.070055 &  60.12018 & TTYH1;ID4;SOX9;SAT1                             & 0\\\\\n",
       "\t Mouse 356 Astro up            & 3/72  & 0.0048859972 & 0.23162675 & 0 & 0 &  9.374291 &  49.88418 & TTYH1;SOX9;SAT1                                 & 0\\\\\n",
       "\t Mouse 334 L6b/CT ENT up       & 2/26  & 0.0067368642 & 0.23162675 & 0 & 0 & 17.814516 &  89.07544 & URI1;RAB8B                                      & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 10\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; | cluster &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| Mouse 372 SMC up              | 4/73  | 0.0004046182 | 0.08264024 | 0 | 0 | 12.636407 |  98.72277 | CCND2;COL4A2;FILIP1;SYNE2                       | 0 |\n",
       "| Human Inh L2 PAX6 FREM2 up    | 7/328 | 0.0009632692 | 0.08264024 | 0 | 0 |  4.853016 |  33.70506 | EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN | 0 |\n",
       "| Human Inh L1 PAX6 MIR101-1 up | 3/44  | 0.0011940188 | 0.08264024 | 0 | 0 | 15.798515 | 106.33081 | EFNB2;TRIM24;CNTLN                              | 0 |\n",
       "| Mouse 358 Astro up            | 4/100 | 0.0013222439 | 0.08264024 | 0 | 0 |  9.070055 |  60.12018 | TTYH1;ID4;SOX9;SAT1                             | 0 |\n",
       "| Mouse 356 Astro up            | 3/72  | 0.0048859972 | 0.23162675 | 0 | 0 |  9.374291 |  49.88418 | TTYH1;SOX9;SAT1                                 | 0 |\n",
       "| Mouse 334 L6b/CT ENT up       | 2/26  | 0.0067368642 | 0.23162675 | 0 | 0 | 17.814516 |  89.07544 | URI1;RAB8B                                      | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Term                          Overlap P.value      Adjusted.P.value\n",
       "1 Mouse 372 SMC up              4/73    0.0004046182 0.08264024      \n",
       "2 Human Inh L2 PAX6 FREM2 up    7/328   0.0009632692 0.08264024      \n",
       "3 Human Inh L1 PAX6 MIR101-1 up 3/44    0.0011940188 0.08264024      \n",
       "4 Mouse 358 Astro up            4/100   0.0013222439 0.08264024      \n",
       "5 Mouse 356 Astro up            3/72    0.0048859972 0.23162675      \n",
       "6 Mouse 334 L6b/CT ENT up       2/26    0.0067368642 0.23162675      \n",
       "  Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score\n",
       "1 0           0                    12.636407   98.72277     \n",
       "2 0           0                     4.853016   33.70506     \n",
       "3 0           0                    15.798515  106.33081     \n",
       "4 0           0                     9.070055   60.12018     \n",
       "5 0           0                     9.374291   49.88418     \n",
       "6 0           0                    17.814516   89.07544     \n",
       "  Genes                                           cluster\n",
       "1 CCND2;COL4A2;FILIP1;SYNE2                       0      \n",
       "2 EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN 0      \n",
       "3 EFNB2;TRIM24;CNTLN                              0      \n",
       "4 TTYH1;ID4;SOX9;SAT1                             0      \n",
       "5 TTYH1;SOX9;SAT1                                 0      \n",
       "6 URI1;RAB8B                                      0      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allen <- enriched$Allen_Brain_Atlas_10x_scRNA_2021 %>% as.data.table()\n",
    "allen$cluster <- 0\n",
    "nrow(allen)\n",
    "head(allen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "9a2d682a-5b3d-4dc6-83de-06cacbfd769f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2"
      ],
      "text/latex": [
       "2"
      ],
      "text/markdown": [
       "2"
      ],
      "text/plain": [
       "[1] 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 2 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th><th scope=col>cluster</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Human Inh L2 PAX6 FREM2 up   </td><td>7/328</td><td>0.0009632692</td><td>0.08264024</td><td>0</td><td>0</td><td> 4.853016</td><td> 33.70506</td><td>EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN</td><td>0</td></tr>\n",
       "\t<tr><td>Human Inh L1 PAX6 MIR101-1 up</td><td>3/44 </td><td>0.0011940188</td><td>0.08264024</td><td>0</td><td>0</td><td>15.798515</td><td>106.33081</td><td>EFNB2;TRIM24;CNTLN                             </td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 2 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes & cluster\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t Human Inh L2 PAX6 FREM2 up    & 7/328 & 0.0009632692 & 0.08264024 & 0 & 0 &  4.853016 &  33.70506 & EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN & 0\\\\\n",
       "\t Human Inh L1 PAX6 MIR101-1 up & 3/44  & 0.0011940188 & 0.08264024 & 0 & 0 & 15.798515 & 106.33081 & EFNB2;TRIM24;CNTLN                              & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 2 × 10\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; | cluster &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| Human Inh L2 PAX6 FREM2 up    | 7/328 | 0.0009632692 | 0.08264024 | 0 | 0 |  4.853016 |  33.70506 | EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN | 0 |\n",
       "| Human Inh L1 PAX6 MIR101-1 up | 3/44  | 0.0011940188 | 0.08264024 | 0 | 0 | 15.798515 | 106.33081 | EFNB2;TRIM24;CNTLN                              | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "  Term                          Overlap P.value      Adjusted.P.value\n",
       "1 Human Inh L2 PAX6 FREM2 up    7/328   0.0009632692 0.08264024      \n",
       "2 Human Inh L1 PAX6 MIR101-1 up 3/44    0.0011940188 0.08264024      \n",
       "  Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score\n",
       "1 0           0                     4.853016   33.70506     \n",
       "2 0           0                    15.798515  106.33081     \n",
       "  Genes                                           cluster\n",
       "1 EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN 0      \n",
       "2 EFNB2;TRIM24;CNTLN                              0      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allel_human <- allen[grepl(\"Human\", allen$Term, ignore.case = TRUE) & allen$Adjusted.P.value < 10e-2]\n",
    "nrow(allel_human)\n",
    "head(allel_human)\n",
    "#head(allel_human$Term, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b0c546b1-2894-4032-9940-e98e0aa17ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation.enrichR <- function (\n",
    "    markers = cluster_markers,\n",
    "    top_n = 50,\n",
    "    database = \"Allen_Brain_Atlas_10x_scRNA_2021\",\n",
    "    output = NULL #or paste0(\"Results/Annotations/EnrichR_output_cluster\", cl, \".txt\")\n",
    ") {\n",
    "    tic(\"EnrichR\")\n",
    "    top_genes <- cluster_markers %>% group_by(cluster) %>% top_n(n = top_n, wt = avg_log2FC) %>% as.data.frame()\n",
    "\n",
    "    # if (!requireNamespace(\"org.Hs.eg.db\", quietly = TRUE)) BiocManager::install(\"org.Hs.eg.db\")\n",
    "    # if (!requireNamespace(\"AnnotationDbi\", quietly = TRUE)) BiocManager::install(\"AnnotationDbi\")\n",
    "    # library(org.Hs.eg.db)\n",
    "    # library(AnnotationDbi)\n",
    "    # Convert gene symbols to Entrez IDs\n",
    "    gene_symbols <- top_genes$gene\n",
    "    entrez_ids <- suppressMessages(\n",
    "        mapIds(org.Hs.eg.db, keys = gene_symbols, column = \"ENTREZID\", keytype = \"SYMBOL\", multiVals = \"first\")\n",
    "    )\n",
    "    # Add Entrez IDs to your data.frame\n",
    "    top_genes$entrez <- entrez_ids\n",
    "    # Remove rows with NA Entrez IDs\n",
    "    top_genes <- top_genes[!is.na(top_genes$entrez), ]\n",
    "\n",
    "    annotation_list <- list()\n",
    "    \n",
    "    for (cl in unique(top_genes$cluster)) {\n",
    "        genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "        # Optionally, use Entrez IDs instead:\n",
    "        # genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "        \n",
    "        # Perform enrichment analysis\n",
    "        enriched <- enrichr(genes_cluster, databases = \"Allen_Brain_Atlas_10x_scRNA_2021\")\n",
    "\n",
    "        annotation <- enriched$Allen_Brain_Atlas_10x_scRNA_2021 %>% as.data.table()\n",
    "        annotation$cluster <- cl\n",
    "\n",
    "        annotation_list[[cl]] <- allen[grepl(\"Human\", allen$Term, ignore.case = TRUE) & allen$Adjusted.P.value < 5e-2]\n",
    "    }\n",
    "    \n",
    "    toc()\n",
    "\n",
    "    return(annotation_list)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "82c8e34e-fcb3-4fc0-bb3b-67e9737c64ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = cluster_markers\n",
    "top_n = 200\n",
    "database = \"Allen_Brain_Atlas_10x_scRNA_2021\"\n",
    "output = NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "bb9d78a1-b097-48bf-8341-b08d44488678",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes <- cluster_markers %>% group_by(cluster) %>% top_n(n = top_n, wt = avg_log2FC) %>% as.data.frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "ecc2dd1c-6399-4f0c-8dd7-eeb015bf626e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_symbols <- top_genes$gene\n",
    "entrez_ids <- suppressMessages(\n",
    "    mapIds(org.Hs.eg.db, keys = gene_symbols, column = \"ENTREZID\", keytype = \"SYMBOL\", multiVals = \"first\")\n",
    ")\n",
    "# Add Entrez IDs to your data.frame\n",
    "top_genes$entrez <- entrez_ids\n",
    "# Remove rows with NA Entrez IDs\n",
    "top_genes <- top_genes[!is.na(top_genes$entrez), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "fefb0c0d-e2a4-421c-ba7e-617c50340116",
   "metadata": {},
   "outputs": [],
   "source": [
    "a <- unique(top_genes$cluster)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "09ca6778-ddac-44f8-9d67-4200c27176d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "2"
      ],
      "text/latex": [
       "2"
      ],
      "text/markdown": [
       "2"
      ],
      "text/plain": [
       "[1] 2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "as.numeric(a)+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7634ed79-f1de-4ff0-9c2e-cc1fc4a60e66",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_list <- list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "d764c07d-b8b6-4534-b20a-e5f08b9bcce3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n"
     ]
    }
   ],
   "source": [
    "for (cl in unique(top_genes$cluster)) {\n",
    "    genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "    # Optionally, use Entrez IDs instead:\n",
    "    # genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "    \n",
    "    # Perform enrichment analysis\n",
    "    enriched <- enrichr(genes_cluster, databases = \"Allen_Brain_Atlas_10x_scRNA_2021\")\n",
    "\n",
    "    annotation <- enriched$Allen_Brain_Atlas_10x_scRNA_2021 %>% as.data.table()\n",
    "    annotation$cluster <- cl\n",
    "\n",
    "    annotation_list[[cl]] <- allen[grepl(\"Human\", allen$Term, ignore.case = TRUE) & allen$Adjusted.P.value < 50e-2]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "714028e0-8258-483a-b1b8-ba927d8f8ae1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n"
     ]
    }
   ],
   "source": [
    "cl <- 0\n",
    "genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "# Optionally, use Entrez IDs instead:\n",
    "# genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "\n",
    "# Perform enrichment analysis\n",
    "enriched <- enrichr(genes_cluster, databases = \"Allen_Brain_Atlas_10x_scRNA_2021\")\n",
    "\n",
    "annotation <- enriched$Allen_Brain_Atlas_10x_scRNA_2021 %>% as.data.table()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "3afe1b75-b6c0-4d80-bcf1-baf81bc68660",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>$Allen_Brain_Atlas_10x_scRNA_2021</strong> = <table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\textbf{\\$Allen\\_Brain\\_Atlas\\_10x\\_scRNA\\_2021} = A data.frame: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "**$Allen_Brain_Atlas_10x_scRNA_2021** = \n",
       "A data.frame: 0 × 9\n",
       "\n",
       "| Term &lt;lgl&gt; | Overlap &lt;lgl&gt; | P.value &lt;lgl&gt; | Adjusted.P.value &lt;lgl&gt; | Old.P.value &lt;lgl&gt; | Old.Adjusted.P.value &lt;lgl&gt; | Odds.Ratio &lt;lgl&gt; | Combined.Score &lt;lgl&gt; | Genes &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "$Allen_Brain_Atlas_10x_scRNA_2021\n",
       "[1] Term                 Overlap              P.value             \n",
       "[4] Adjusted.P.value     Old.P.value          Old.Adjusted.P.value\n",
       "[7] Odds.Ratio           Combined.Score       Genes               \n",
       "<0 rows> (or 0-length row.names)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "1f0275fb-2705-4046-ac44-04113b2a1176",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 40 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>Human Inh L2 PAX6 FREM2 up         </td><td>7/328</td><td>0.0009632692</td><td>0.08264024</td><td>0</td><td>0</td><td> 4.853016</td><td> 33.705059</td><td>EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN</td></tr>\n",
       "\t<tr><td>Human Inh L1 PAX6 MIR101-1 up      </td><td>3/44 </td><td>0.0011940188</td><td>0.08264024</td><td>0</td><td>0</td><td>15.798515</td><td>106.330809</td><td>EFNB2;TRIM24;CNTLN                             </td></tr>\n",
       "\t<tr><td>Human Astro L1 FGFR3 SERPINI2 up   </td><td>3/101</td><td>0.0123712614</td><td>0.27676560</td><td>0</td><td>0</td><td> 6.590617</td><td> 28.948487</td><td>RFX4;CACHD1;ZFP36L1                            </td></tr>\n",
       "\t<tr><td>Human Inh L1-5 VIP CD27-AS1 up     </td><td>2/37 </td><td>0.0133465555</td><td>0.27676560</td><td>0</td><td>0</td><td>12.208909</td><td> 52.699720</td><td>FOXG1;RAB8B                                    </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 SST KLHL1 up        </td><td>2/40 </td><td>0.0154880314</td><td>0.27676560</td><td>0</td><td>0</td><td>11.243350</td><td> 46.858773</td><td>ARX;RAB8B                                      </td></tr>\n",
       "\t<tr><td>Human Endo L2-5 NOSTRIN SRGN up    </td><td>2/43 </td><td>0.0177664394</td><td>0.29610732</td><td>0</td><td>0</td><td>10.419093</td><td> 41.993569</td><td>YES1;SYNE2                                     </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 PVALB FAM150B up    </td><td>2/48 </td><td>0.0218566319</td><td>0.31041640</td><td>0</td><td>0</td><td> 9.284245</td><td> 35.495998</td><td>FILIP1;CREB5                                   </td></tr>\n",
       "\t<tr><td>Human Astro L1-6 FGFR3 PLCG1 up    </td><td>2/64 </td><td>0.0372052475</td><td>0.37459555</td><td>0</td><td>0</td><td> 6.882761</td><td> 22.653269</td><td>TTYH1;RFX4                                     </td></tr>\n",
       "\t<tr><td>Human Exc L6 FEZF2 KLK7 down       </td><td>1/11 </td><td>0.0510387981</td><td>0.37523538</td><td>0</td><td>0</td><td>21.164894</td><td> 62.969139</td><td>QKI                                            </td></tr>\n",
       "\t<tr><td>Human Exc L3-5 RORB LAMA4 down     </td><td>1/12 </td><td>0.0555488338</td><td>0.37523538</td><td>0</td><td>0</td><td>19.239845</td><td> 55.612633</td><td>QKI                                            </td></tr>\n",
       "\t<tr><td>Human Exc L2-3 LINC00507 DSG3 up   </td><td>3/184</td><td>0.0573841939</td><td>0.37523538</td><td>0</td><td>0</td><td> 3.553447</td><td> 10.155703</td><td>EFNB2;LINC01551;FOXG1                          </td></tr>\n",
       "\t<tr><td>Human Exc L2-3 LINC00507 DSG3 down </td><td>1/13 </td><td>0.0600376601</td><td>0.37523538</td><td>0</td><td>0</td><td>17.635638</td><td> 49.605228</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Exc L6 FEZF2 POGK down       </td><td>1/16 </td><td>0.0733778669</td><td>0.39915662</td><td>0</td><td>0</td><td>14.106383</td><td> 36.847747</td><td>QKI                                            </td></tr>\n",
       "\t<tr><td>Human Exc L3-5 FEZF2 ASGR2 down    </td><td>1/18 </td><td>0.0821670875</td><td>0.42492730</td><td>0</td><td>0</td><td>12.445557</td><td> 31.101452</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Inh L1 LAMP5 NMBR up         </td><td>1/19 </td><td>0.0865307131</td><td>0.42492730</td><td>0</td><td>0</td><td>11.753546</td><td> 28.763935</td><td>TRIM24                                         </td></tr>\n",
       "\t<tr><td>Human Inh L3-6 PAX6 LINC01497 up   </td><td>2/106</td><td>0.0904990018</td><td>0.42492730</td><td>0</td><td>0</td><td> 4.094500</td><td>  9.836693</td><td>TRIM24;RAB8B                                   </td></tr>\n",
       "\t<tr><td>Human Exc L5 FEZF2 NREP-AS1 down   </td><td>1/20 </td><td>0.0908738104</td><td>0.42492730</td><td>0</td><td>0</td><td>11.134378</td><td> 26.703395</td><td>QKI                                            </td></tr>\n",
       "\t<tr><td>Human Exc L3-5 FEZF2 LINC01107 down</td><td>1/21 </td><td>0.0951964750</td><td>0.42492730</td><td>0</td><td>0</td><td>10.577128</td><td> 24.875420</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Inh L5 PVALB LRIG3 up        </td><td>3/234</td><td>0.1004061115</td><td>0.42492730</td><td>0</td><td>0</td><td> 2.777244</td><td>  6.383586</td><td>PKIB;NOC3L;CNTLN                               </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 VIP COL4A3 up       </td><td>2/115</td><td>0.1037759572</td><td>0.42492730</td><td>0</td><td>0</td><td> 3.766676</td><td>  8.533484</td><td>RPL3;RAB8B                                     </td></tr>\n",
       "\t<tr><td>Human Inh L2-5 VIP SOX11 up        </td><td>2/119</td><td>0.1098390280</td><td>0.42492730</td><td>0</td><td>0</td><td> 3.637166</td><td>  8.033551</td><td>INTU;RAB8B                                     </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 PVALB ZFPM2-AS1 up  </td><td>2/120</td><td>0.1113693366</td><td>0.42492730</td><td>0</td><td>0</td><td> 3.606160</td><td>  7.915172</td><td>ARX;RAB8B                                      </td></tr>\n",
       "\t<tr><td>Human Inh L3-6 PAX6 LINC01497 down </td><td>1/25 </td><td>0.1122847006</td><td>0.42492730</td><td>0</td><td>0</td><td> 8.812500</td><td> 19.270449</td><td>LIMCH1                                         </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 SST DNAJC14 up      </td><td>5/543</td><td>0.1164224001</td><td>0.42492730</td><td>0</td><td>0</td><td> 1.999897</td><td>  4.300839</td><td>LINC01551;RPL3;RPS27L;ABCE1;RAB8B              </td></tr>\n",
       "\t<tr><td>Human Exc L5-6 FEZF2 CFTR down     </td><td>1/26 </td><td>0.1165066190</td><td>0.42492730</td><td>0</td><td>0</td><td> 8.459574</td><td> 18.186454</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Exc L5-6 FEZF2 IFNG-AS1 down </td><td>1/26 </td><td>0.1165066190</td><td>0.42492730</td><td>0</td><td>0</td><td> 8.459574</td><td> 18.186454</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Exc L5 FEZF2 RNF144A-AS1 down</td><td>1/28 </td><td>0.1248909420</td><td>0.42492730</td><td>0</td><td>0</td><td> 7.832151</td><td> 16.293337</td><td>QKI                                            </td></tr>\n",
       "\t<tr><td>Human Exc L2 LINC00507 ATP7B down  </td><td>1/30 </td><td>0.1331965279</td><td>0.42492730</td><td>0</td><td>0</td><td> 7.291269</td><td> 14.698685</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Exc L5-6 THEMIS SMYD1 down   </td><td>1/31 </td><td>0.1373200234</td><td>0.42492730</td><td>0</td><td>0</td><td> 7.047872</td><td> 13.993136</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Inh L1 PAX6 MIR101-1 down    </td><td>1/31 </td><td>0.1373200234</td><td>0.42492730</td><td>0</td><td>0</td><td> 7.047872</td><td> 13.993136</td><td>LIMCH1                                         </td></tr>\n",
       "\t<tr><td>Human Exc L3-5 RORB LNX2 up        </td><td>1/33 </td><td>0.1455088735</td><td>0.42492730</td><td>0</td><td>0</td><td> 6.606715</td><td> 12.734564</td><td>EFNB2                                          </td></tr>\n",
       "\t<tr><td>Human Inh L6 SST TH up             </td><td>3/279</td><td>0.1472417428</td><td>0.42492730</td><td>0</td><td>0</td><td> 2.319116</td><td>  4.442684</td><td>EFNB2;FOXG1;RPS27L                             </td></tr>\n",
       "\t<tr><td>Human Inh L3-6 VIP ZIM2-AS1 down   </td><td>1/37 </td><td>0.1616565266</td><td>0.43591061</td><td>0</td><td>0</td><td> 5.871454</td><td> 10.699441</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Inh L3-6 VIP UG0898H09 up    </td><td>1/38 </td><td>0.1656460329</td><td>0.43591061</td><td>0</td><td>0</td><td> 5.712478</td><td> 10.270477</td><td>RAB8B                                          </td></tr>\n",
       "\t<tr><td>Human Inh L1-6 LAMP5 CA1 up        </td><td>1/40 </td><td>0.1735687740</td><td>0.44464968</td><td>0</td><td>0</td><td> 5.418985</td><td>  9.489626</td><td>COL4A2                                         </td></tr>\n",
       "\t<tr><td>Human Inh L1-6 SST NPY down        </td><td>1/42 </td><td>0.1814170679</td><td>0.44464968</td><td>0</td><td>0</td><td> 5.154126</td><td>  8.797869</td><td>NCALD                                          </td></tr>\n",
       "\t<tr><td>Human Astro L1-6 FGFR3 AQP1 up     </td><td>1/45 </td><td>0.1930514329</td><td>0.44687832</td><td>0</td><td>0</td><td> 4.801983</td><td>  7.898294</td><td>RFX4                                           </td></tr>\n",
       "\t<tr><td>Human Inh L1-2 VIP HTR3A up        </td><td>1/45 </td><td>0.1930514329</td><td>0.44687832</td><td>0</td><td>0</td><td> 4.801983</td><td>  7.898294</td><td>INTU                                           </td></tr>\n",
       "\t<tr><td>Human Inh L5-6 PVALB SST CRHR2 up  </td><td>1/50 </td><td>0.2120795206</td><td>0.45316137</td><td>0</td><td>0</td><td> 4.310899</td><td>  6.685316</td><td>HEY1                                           </td></tr>\n",
       "\t<tr><td>Human OPC L1-6 PDGFRA COL20A1 up   </td><td>1/52 </td><td>0.2195658348</td><td>0.46127276</td><td>0</td><td>0</td><td> 4.141427</td><td>  6.278830</td><td>VCAN                                           </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 40 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t Human Inh L2 PAX6 FREM2 up          & 7/328 & 0.0009632692 & 0.08264024 & 0 & 0 &  4.853016 &  33.705059 & EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN\\\\\n",
       "\t Human Inh L1 PAX6 MIR101-1 up       & 3/44  & 0.0011940188 & 0.08264024 & 0 & 0 & 15.798515 & 106.330809 & EFNB2;TRIM24;CNTLN                             \\\\\n",
       "\t Human Astro L1 FGFR3 SERPINI2 up    & 3/101 & 0.0123712614 & 0.27676560 & 0 & 0 &  6.590617 &  28.948487 & RFX4;CACHD1;ZFP36L1                            \\\\\n",
       "\t Human Inh L1-5 VIP CD27-AS1 up      & 2/37  & 0.0133465555 & 0.27676560 & 0 & 0 & 12.208909 &  52.699720 & FOXG1;RAB8B                                    \\\\\n",
       "\t Human Inh L5-6 SST KLHL1 up         & 2/40  & 0.0154880314 & 0.27676560 & 0 & 0 & 11.243350 &  46.858773 & ARX;RAB8B                                      \\\\\n",
       "\t Human Endo L2-5 NOSTRIN SRGN up     & 2/43  & 0.0177664394 & 0.29610732 & 0 & 0 & 10.419093 &  41.993569 & YES1;SYNE2                                     \\\\\n",
       "\t Human Inh L5-6 PVALB FAM150B up     & 2/48  & 0.0218566319 & 0.31041640 & 0 & 0 &  9.284245 &  35.495998 & FILIP1;CREB5                                   \\\\\n",
       "\t Human Astro L1-6 FGFR3 PLCG1 up     & 2/64  & 0.0372052475 & 0.37459555 & 0 & 0 &  6.882761 &  22.653269 & TTYH1;RFX4                                     \\\\\n",
       "\t Human Exc L6 FEZF2 KLK7 down        & 1/11  & 0.0510387981 & 0.37523538 & 0 & 0 & 21.164894 &  62.969139 & QKI                                            \\\\\n",
       "\t Human Exc L3-5 RORB LAMA4 down      & 1/12  & 0.0555488338 & 0.37523538 & 0 & 0 & 19.239845 &  55.612633 & QKI                                            \\\\\n",
       "\t Human Exc L2-3 LINC00507 DSG3 up    & 3/184 & 0.0573841939 & 0.37523538 & 0 & 0 &  3.553447 &  10.155703 & EFNB2;LINC01551;FOXG1                          \\\\\n",
       "\t Human Exc L2-3 LINC00507 DSG3 down  & 1/13  & 0.0600376601 & 0.37523538 & 0 & 0 & 17.635638 &  49.605228 & NCALD                                          \\\\\n",
       "\t Human Exc L6 FEZF2 POGK down        & 1/16  & 0.0733778669 & 0.39915662 & 0 & 0 & 14.106383 &  36.847747 & QKI                                            \\\\\n",
       "\t Human Exc L3-5 FEZF2 ASGR2 down     & 1/18  & 0.0821670875 & 0.42492730 & 0 & 0 & 12.445557 &  31.101452 & NCALD                                          \\\\\n",
       "\t Human Inh L1 LAMP5 NMBR up          & 1/19  & 0.0865307131 & 0.42492730 & 0 & 0 & 11.753546 &  28.763935 & TRIM24                                         \\\\\n",
       "\t Human Inh L3-6 PAX6 LINC01497 up    & 2/106 & 0.0904990018 & 0.42492730 & 0 & 0 &  4.094500 &   9.836693 & TRIM24;RAB8B                                   \\\\\n",
       "\t Human Exc L5 FEZF2 NREP-AS1 down    & 1/20  & 0.0908738104 & 0.42492730 & 0 & 0 & 11.134378 &  26.703395 & QKI                                            \\\\\n",
       "\t Human Exc L3-5 FEZF2 LINC01107 down & 1/21  & 0.0951964750 & 0.42492730 & 0 & 0 & 10.577128 &  24.875420 & NCALD                                          \\\\\n",
       "\t Human Inh L5 PVALB LRIG3 up         & 3/234 & 0.1004061115 & 0.42492730 & 0 & 0 &  2.777244 &   6.383586 & PKIB;NOC3L;CNTLN                               \\\\\n",
       "\t Human Inh L5-6 VIP COL4A3 up        & 2/115 & 0.1037759572 & 0.42492730 & 0 & 0 &  3.766676 &   8.533484 & RPL3;RAB8B                                     \\\\\n",
       "\t Human Inh L2-5 VIP SOX11 up         & 2/119 & 0.1098390280 & 0.42492730 & 0 & 0 &  3.637166 &   8.033551 & INTU;RAB8B                                     \\\\\n",
       "\t Human Inh L5-6 PVALB ZFPM2-AS1 up   & 2/120 & 0.1113693366 & 0.42492730 & 0 & 0 &  3.606160 &   7.915172 & ARX;RAB8B                                      \\\\\n",
       "\t Human Inh L3-6 PAX6 LINC01497 down  & 1/25  & 0.1122847006 & 0.42492730 & 0 & 0 &  8.812500 &  19.270449 & LIMCH1                                         \\\\\n",
       "\t Human Inh L5-6 SST DNAJC14 up       & 5/543 & 0.1164224001 & 0.42492730 & 0 & 0 &  1.999897 &   4.300839 & LINC01551;RPL3;RPS27L;ABCE1;RAB8B              \\\\\n",
       "\t Human Exc L5-6 FEZF2 CFTR down      & 1/26  & 0.1165066190 & 0.42492730 & 0 & 0 &  8.459574 &  18.186454 & NCALD                                          \\\\\n",
       "\t Human Exc L5-6 FEZF2 IFNG-AS1 down  & 1/26  & 0.1165066190 & 0.42492730 & 0 & 0 &  8.459574 &  18.186454 & NCALD                                          \\\\\n",
       "\t Human Exc L5 FEZF2 RNF144A-AS1 down & 1/28  & 0.1248909420 & 0.42492730 & 0 & 0 &  7.832151 &  16.293337 & QKI                                            \\\\\n",
       "\t Human Exc L2 LINC00507 ATP7B down   & 1/30  & 0.1331965279 & 0.42492730 & 0 & 0 &  7.291269 &  14.698685 & NCALD                                          \\\\\n",
       "\t Human Exc L5-6 THEMIS SMYD1 down    & 1/31  & 0.1373200234 & 0.42492730 & 0 & 0 &  7.047872 &  13.993136 & NCALD                                          \\\\\n",
       "\t Human Inh L1 PAX6 MIR101-1 down     & 1/31  & 0.1373200234 & 0.42492730 & 0 & 0 &  7.047872 &  13.993136 & LIMCH1                                         \\\\\n",
       "\t Human Exc L3-5 RORB LNX2 up         & 1/33  & 0.1455088735 & 0.42492730 & 0 & 0 &  6.606715 &  12.734564 & EFNB2                                          \\\\\n",
       "\t Human Inh L6 SST TH up              & 3/279 & 0.1472417428 & 0.42492730 & 0 & 0 &  2.319116 &   4.442684 & EFNB2;FOXG1;RPS27L                             \\\\\n",
       "\t Human Inh L3-6 VIP ZIM2-AS1 down    & 1/37  & 0.1616565266 & 0.43591061 & 0 & 0 &  5.871454 &  10.699441 & NCALD                                          \\\\\n",
       "\t Human Inh L3-6 VIP UG0898H09 up     & 1/38  & 0.1656460329 & 0.43591061 & 0 & 0 &  5.712478 &  10.270477 & RAB8B                                          \\\\\n",
       "\t Human Inh L1-6 LAMP5 CA1 up         & 1/40  & 0.1735687740 & 0.44464968 & 0 & 0 &  5.418985 &   9.489626 & COL4A2                                         \\\\\n",
       "\t Human Inh L1-6 SST NPY down         & 1/42  & 0.1814170679 & 0.44464968 & 0 & 0 &  5.154126 &   8.797869 & NCALD                                          \\\\\n",
       "\t Human Astro L1-6 FGFR3 AQP1 up      & 1/45  & 0.1930514329 & 0.44687832 & 0 & 0 &  4.801983 &   7.898294 & RFX4                                           \\\\\n",
       "\t Human Inh L1-2 VIP HTR3A up         & 1/45  & 0.1930514329 & 0.44687832 & 0 & 0 &  4.801983 &   7.898294 & INTU                                           \\\\\n",
       "\t Human Inh L5-6 PVALB SST CRHR2 up   & 1/50  & 0.2120795206 & 0.45316137 & 0 & 0 &  4.310899 &   6.685316 & HEY1                                           \\\\\n",
       "\t Human OPC L1-6 PDGFRA COL20A1 up    & 1/52  & 0.2195658348 & 0.46127276 & 0 & 0 &  4.141427 &   6.278830 & VCAN                                           \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 40 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| Human Inh L2 PAX6 FREM2 up          | 7/328 | 0.0009632692 | 0.08264024 | 0 | 0 |  4.853016 |  33.705059 | EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN |\n",
       "| Human Inh L1 PAX6 MIR101-1 up       | 3/44  | 0.0011940188 | 0.08264024 | 0 | 0 | 15.798515 | 106.330809 | EFNB2;TRIM24;CNTLN                              |\n",
       "| Human Astro L1 FGFR3 SERPINI2 up    | 3/101 | 0.0123712614 | 0.27676560 | 0 | 0 |  6.590617 |  28.948487 | RFX4;CACHD1;ZFP36L1                             |\n",
       "| Human Inh L1-5 VIP CD27-AS1 up      | 2/37  | 0.0133465555 | 0.27676560 | 0 | 0 | 12.208909 |  52.699720 | FOXG1;RAB8B                                     |\n",
       "| Human Inh L5-6 SST KLHL1 up         | 2/40  | 0.0154880314 | 0.27676560 | 0 | 0 | 11.243350 |  46.858773 | ARX;RAB8B                                       |\n",
       "| Human Endo L2-5 NOSTRIN SRGN up     | 2/43  | 0.0177664394 | 0.29610732 | 0 | 0 | 10.419093 |  41.993569 | YES1;SYNE2                                      |\n",
       "| Human Inh L5-6 PVALB FAM150B up     | 2/48  | 0.0218566319 | 0.31041640 | 0 | 0 |  9.284245 |  35.495998 | FILIP1;CREB5                                    |\n",
       "| Human Astro L1-6 FGFR3 PLCG1 up     | 2/64  | 0.0372052475 | 0.37459555 | 0 | 0 |  6.882761 |  22.653269 | TTYH1;RFX4                                      |\n",
       "| Human Exc L6 FEZF2 KLK7 down        | 1/11  | 0.0510387981 | 0.37523538 | 0 | 0 | 21.164894 |  62.969139 | QKI                                             |\n",
       "| Human Exc L3-5 RORB LAMA4 down      | 1/12  | 0.0555488338 | 0.37523538 | 0 | 0 | 19.239845 |  55.612633 | QKI                                             |\n",
       "| Human Exc L2-3 LINC00507 DSG3 up    | 3/184 | 0.0573841939 | 0.37523538 | 0 | 0 |  3.553447 |  10.155703 | EFNB2;LINC01551;FOXG1                           |\n",
       "| Human Exc L2-3 LINC00507 DSG3 down  | 1/13  | 0.0600376601 | 0.37523538 | 0 | 0 | 17.635638 |  49.605228 | NCALD                                           |\n",
       "| Human Exc L6 FEZF2 POGK down        | 1/16  | 0.0733778669 | 0.39915662 | 0 | 0 | 14.106383 |  36.847747 | QKI                                             |\n",
       "| Human Exc L3-5 FEZF2 ASGR2 down     | 1/18  | 0.0821670875 | 0.42492730 | 0 | 0 | 12.445557 |  31.101452 | NCALD                                           |\n",
       "| Human Inh L1 LAMP5 NMBR up          | 1/19  | 0.0865307131 | 0.42492730 | 0 | 0 | 11.753546 |  28.763935 | TRIM24                                          |\n",
       "| Human Inh L3-6 PAX6 LINC01497 up    | 2/106 | 0.0904990018 | 0.42492730 | 0 | 0 |  4.094500 |   9.836693 | TRIM24;RAB8B                                    |\n",
       "| Human Exc L5 FEZF2 NREP-AS1 down    | 1/20  | 0.0908738104 | 0.42492730 | 0 | 0 | 11.134378 |  26.703395 | QKI                                             |\n",
       "| Human Exc L3-5 FEZF2 LINC01107 down | 1/21  | 0.0951964750 | 0.42492730 | 0 | 0 | 10.577128 |  24.875420 | NCALD                                           |\n",
       "| Human Inh L5 PVALB LRIG3 up         | 3/234 | 0.1004061115 | 0.42492730 | 0 | 0 |  2.777244 |   6.383586 | PKIB;NOC3L;CNTLN                                |\n",
       "| Human Inh L5-6 VIP COL4A3 up        | 2/115 | 0.1037759572 | 0.42492730 | 0 | 0 |  3.766676 |   8.533484 | RPL3;RAB8B                                      |\n",
       "| Human Inh L2-5 VIP SOX11 up         | 2/119 | 0.1098390280 | 0.42492730 | 0 | 0 |  3.637166 |   8.033551 | INTU;RAB8B                                      |\n",
       "| Human Inh L5-6 PVALB ZFPM2-AS1 up   | 2/120 | 0.1113693366 | 0.42492730 | 0 | 0 |  3.606160 |   7.915172 | ARX;RAB8B                                       |\n",
       "| Human Inh L3-6 PAX6 LINC01497 down  | 1/25  | 0.1122847006 | 0.42492730 | 0 | 0 |  8.812500 |  19.270449 | LIMCH1                                          |\n",
       "| Human Inh L5-6 SST DNAJC14 up       | 5/543 | 0.1164224001 | 0.42492730 | 0 | 0 |  1.999897 |   4.300839 | LINC01551;RPL3;RPS27L;ABCE1;RAB8B               |\n",
       "| Human Exc L5-6 FEZF2 CFTR down      | 1/26  | 0.1165066190 | 0.42492730 | 0 | 0 |  8.459574 |  18.186454 | NCALD                                           |\n",
       "| Human Exc L5-6 FEZF2 IFNG-AS1 down  | 1/26  | 0.1165066190 | 0.42492730 | 0 | 0 |  8.459574 |  18.186454 | NCALD                                           |\n",
       "| Human Exc L5 FEZF2 RNF144A-AS1 down | 1/28  | 0.1248909420 | 0.42492730 | 0 | 0 |  7.832151 |  16.293337 | QKI                                             |\n",
       "| Human Exc L2 LINC00507 ATP7B down   | 1/30  | 0.1331965279 | 0.42492730 | 0 | 0 |  7.291269 |  14.698685 | NCALD                                           |\n",
       "| Human Exc L5-6 THEMIS SMYD1 down    | 1/31  | 0.1373200234 | 0.42492730 | 0 | 0 |  7.047872 |  13.993136 | NCALD                                           |\n",
       "| Human Inh L1 PAX6 MIR101-1 down     | 1/31  | 0.1373200234 | 0.42492730 | 0 | 0 |  7.047872 |  13.993136 | LIMCH1                                          |\n",
       "| Human Exc L3-5 RORB LNX2 up         | 1/33  | 0.1455088735 | 0.42492730 | 0 | 0 |  6.606715 |  12.734564 | EFNB2                                           |\n",
       "| Human Inh L6 SST TH up              | 3/279 | 0.1472417428 | 0.42492730 | 0 | 0 |  2.319116 |   4.442684 | EFNB2;FOXG1;RPS27L                              |\n",
       "| Human Inh L3-6 VIP ZIM2-AS1 down    | 1/37  | 0.1616565266 | 0.43591061 | 0 | 0 |  5.871454 |  10.699441 | NCALD                                           |\n",
       "| Human Inh L3-6 VIP UG0898H09 up     | 1/38  | 0.1656460329 | 0.43591061 | 0 | 0 |  5.712478 |  10.270477 | RAB8B                                           |\n",
       "| Human Inh L1-6 LAMP5 CA1 up         | 1/40  | 0.1735687740 | 0.44464968 | 0 | 0 |  5.418985 |   9.489626 | COL4A2                                          |\n",
       "| Human Inh L1-6 SST NPY down         | 1/42  | 0.1814170679 | 0.44464968 | 0 | 0 |  5.154126 |   8.797869 | NCALD                                           |\n",
       "| Human Astro L1-6 FGFR3 AQP1 up      | 1/45  | 0.1930514329 | 0.44687832 | 0 | 0 |  4.801983 |   7.898294 | RFX4                                            |\n",
       "| Human Inh L1-2 VIP HTR3A up         | 1/45  | 0.1930514329 | 0.44687832 | 0 | 0 |  4.801983 |   7.898294 | INTU                                            |\n",
       "| Human Inh L5-6 PVALB SST CRHR2 up   | 1/50  | 0.2120795206 | 0.45316137 | 0 | 0 |  4.310899 |   6.685316 | HEY1                                            |\n",
       "| Human OPC L1-6 PDGFRA COL20A1 up    | 1/52  | 0.2195658348 | 0.46127276 | 0 | 0 |  4.141427 |   6.278830 | VCAN                                            |\n",
       "\n"
      ],
      "text/plain": [
       "   Term                                Overlap P.value      Adjusted.P.value\n",
       "1  Human Inh L2 PAX6 FREM2 up          7/328   0.0009632692 0.08264024      \n",
       "2  Human Inh L1 PAX6 MIR101-1 up       3/44    0.0011940188 0.08264024      \n",
       "3  Human Astro L1 FGFR3 SERPINI2 up    3/101   0.0123712614 0.27676560      \n",
       "4  Human Inh L1-5 VIP CD27-AS1 up      2/37    0.0133465555 0.27676560      \n",
       "5  Human Inh L5-6 SST KLHL1 up         2/40    0.0154880314 0.27676560      \n",
       "6  Human Endo L2-5 NOSTRIN SRGN up     2/43    0.0177664394 0.29610732      \n",
       "7  Human Inh L5-6 PVALB FAM150B up     2/48    0.0218566319 0.31041640      \n",
       "8  Human Astro L1-6 FGFR3 PLCG1 up     2/64    0.0372052475 0.37459555      \n",
       "9  Human Exc L6 FEZF2 KLK7 down        1/11    0.0510387981 0.37523538      \n",
       "10 Human Exc L3-5 RORB LAMA4 down      1/12    0.0555488338 0.37523538      \n",
       "11 Human Exc L2-3 LINC00507 DSG3 up    3/184   0.0573841939 0.37523538      \n",
       "12 Human Exc L2-3 LINC00507 DSG3 down  1/13    0.0600376601 0.37523538      \n",
       "13 Human Exc L6 FEZF2 POGK down        1/16    0.0733778669 0.39915662      \n",
       "14 Human Exc L3-5 FEZF2 ASGR2 down     1/18    0.0821670875 0.42492730      \n",
       "15 Human Inh L1 LAMP5 NMBR up          1/19    0.0865307131 0.42492730      \n",
       "16 Human Inh L3-6 PAX6 LINC01497 up    2/106   0.0904990018 0.42492730      \n",
       "17 Human Exc L5 FEZF2 NREP-AS1 down    1/20    0.0908738104 0.42492730      \n",
       "18 Human Exc L3-5 FEZF2 LINC01107 down 1/21    0.0951964750 0.42492730      \n",
       "19 Human Inh L5 PVALB LRIG3 up         3/234   0.1004061115 0.42492730      \n",
       "20 Human Inh L5-6 VIP COL4A3 up        2/115   0.1037759572 0.42492730      \n",
       "21 Human Inh L2-5 VIP SOX11 up         2/119   0.1098390280 0.42492730      \n",
       "22 Human Inh L5-6 PVALB ZFPM2-AS1 up   2/120   0.1113693366 0.42492730      \n",
       "23 Human Inh L3-6 PAX6 LINC01497 down  1/25    0.1122847006 0.42492730      \n",
       "24 Human Inh L5-6 SST DNAJC14 up       5/543   0.1164224001 0.42492730      \n",
       "25 Human Exc L5-6 FEZF2 CFTR down      1/26    0.1165066190 0.42492730      \n",
       "26 Human Exc L5-6 FEZF2 IFNG-AS1 down  1/26    0.1165066190 0.42492730      \n",
       "27 Human Exc L5 FEZF2 RNF144A-AS1 down 1/28    0.1248909420 0.42492730      \n",
       "28 Human Exc L2 LINC00507 ATP7B down   1/30    0.1331965279 0.42492730      \n",
       "29 Human Exc L5-6 THEMIS SMYD1 down    1/31    0.1373200234 0.42492730      \n",
       "30 Human Inh L1 PAX6 MIR101-1 down     1/31    0.1373200234 0.42492730      \n",
       "31 Human Exc L3-5 RORB LNX2 up         1/33    0.1455088735 0.42492730      \n",
       "32 Human Inh L6 SST TH up              3/279   0.1472417428 0.42492730      \n",
       "33 Human Inh L3-6 VIP ZIM2-AS1 down    1/37    0.1616565266 0.43591061      \n",
       "34 Human Inh L3-6 VIP UG0898H09 up     1/38    0.1656460329 0.43591061      \n",
       "35 Human Inh L1-6 LAMP5 CA1 up         1/40    0.1735687740 0.44464968      \n",
       "36 Human Inh L1-6 SST NPY down         1/42    0.1814170679 0.44464968      \n",
       "37 Human Astro L1-6 FGFR3 AQP1 up      1/45    0.1930514329 0.44687832      \n",
       "38 Human Inh L1-2 VIP HTR3A up         1/45    0.1930514329 0.44687832      \n",
       "39 Human Inh L5-6 PVALB SST CRHR2 up   1/50    0.2120795206 0.45316137      \n",
       "40 Human OPC L1-6 PDGFRA COL20A1 up    1/52    0.2195658348 0.46127276      \n",
       "   Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score\n",
       "1  0           0                     4.853016   33.705059    \n",
       "2  0           0                    15.798515  106.330809    \n",
       "3  0           0                     6.590617   28.948487    \n",
       "4  0           0                    12.208909   52.699720    \n",
       "5  0           0                    11.243350   46.858773    \n",
       "6  0           0                    10.419093   41.993569    \n",
       "7  0           0                     9.284245   35.495998    \n",
       "8  0           0                     6.882761   22.653269    \n",
       "9  0           0                    21.164894   62.969139    \n",
       "10 0           0                    19.239845   55.612633    \n",
       "11 0           0                     3.553447   10.155703    \n",
       "12 0           0                    17.635638   49.605228    \n",
       "13 0           0                    14.106383   36.847747    \n",
       "14 0           0                    12.445557   31.101452    \n",
       "15 0           0                    11.753546   28.763935    \n",
       "16 0           0                     4.094500    9.836693    \n",
       "17 0           0                    11.134378   26.703395    \n",
       "18 0           0                    10.577128   24.875420    \n",
       "19 0           0                     2.777244    6.383586    \n",
       "20 0           0                     3.766676    8.533484    \n",
       "21 0           0                     3.637166    8.033551    \n",
       "22 0           0                     3.606160    7.915172    \n",
       "23 0           0                     8.812500   19.270449    \n",
       "24 0           0                     1.999897    4.300839    \n",
       "25 0           0                     8.459574   18.186454    \n",
       "26 0           0                     8.459574   18.186454    \n",
       "27 0           0                     7.832151   16.293337    \n",
       "28 0           0                     7.291269   14.698685    \n",
       "29 0           0                     7.047872   13.993136    \n",
       "30 0           0                     7.047872   13.993136    \n",
       "31 0           0                     6.606715   12.734564    \n",
       "32 0           0                     2.319116    4.442684    \n",
       "33 0           0                     5.871454   10.699441    \n",
       "34 0           0                     5.712478   10.270477    \n",
       "35 0           0                     5.418985    9.489626    \n",
       "36 0           0                     5.154126    8.797869    \n",
       "37 0           0                     4.801983    7.898294    \n",
       "38 0           0                     4.801983    7.898294    \n",
       "39 0           0                     4.310899    6.685316    \n",
       "40 0           0                     4.141427    6.278830    \n",
       "   Genes                                          \n",
       "1  EFNB2;LINC01551;FOXG1;TRIM24;CACHD1;RAB8B;CNTLN\n",
       "2  EFNB2;TRIM24;CNTLN                             \n",
       "3  RFX4;CACHD1;ZFP36L1                            \n",
       "4  FOXG1;RAB8B                                    \n",
       "5  ARX;RAB8B                                      \n",
       "6  YES1;SYNE2                                     \n",
       "7  FILIP1;CREB5                                   \n",
       "8  TTYH1;RFX4                                     \n",
       "9  QKI                                            \n",
       "10 QKI                                            \n",
       "11 EFNB2;LINC01551;FOXG1                          \n",
       "12 NCALD                                          \n",
       "13 QKI                                            \n",
       "14 NCALD                                          \n",
       "15 TRIM24                                         \n",
       "16 TRIM24;RAB8B                                   \n",
       "17 QKI                                            \n",
       "18 NCALD                                          \n",
       "19 PKIB;NOC3L;CNTLN                               \n",
       "20 RPL3;RAB8B                                     \n",
       "21 INTU;RAB8B                                     \n",
       "22 ARX;RAB8B                                      \n",
       "23 LIMCH1                                         \n",
       "24 LINC01551;RPL3;RPS27L;ABCE1;RAB8B              \n",
       "25 NCALD                                          \n",
       "26 NCALD                                          \n",
       "27 QKI                                            \n",
       "28 NCALD                                          \n",
       "29 NCALD                                          \n",
       "30 LIMCH1                                         \n",
       "31 EFNB2                                          \n",
       "32 EFNB2;FOXG1;RPS27L                             \n",
       "33 NCALD                                          \n",
       "34 RAB8B                                          \n",
       "35 COL4A2                                         \n",
       "36 NCALD                                          \n",
       "37 RFX4                                           \n",
       "38 INTU                                           \n",
       "39 HEY1                                           \n",
       "40 VCAN                                           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "annotation_list$'0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c94f8948-e9ee-49d9-8524-4f7d3134b8b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n"
     ]
    }
   ],
   "source": [
    "genes_cluster <- top_genes %>% filter(cluster == 1) %>% pull(entrez)\n",
    "# Optionally, use Entrez IDs instead:\n",
    "# genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "\n",
    "# Perform enrichment analysis\n",
    "enriched <- enrichr(genes_cluster, databases = \"Allen_Brain_Atlas_10x_scRNA_2021\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cb9b5317-da1b-4087-ba22-87b7306133a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>$Allen_Brain_Atlas_10x_scRNA_2021</strong> = <table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\textbf{\\$Allen\\_Brain\\_Atlas\\_10x\\_scRNA\\_2021} = A data.frame: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "**$Allen_Brain_Atlas_10x_scRNA_2021** = \n",
       "A data.frame: 0 × 9\n",
       "\n",
       "| Term &lt;lgl&gt; | Overlap &lt;lgl&gt; | P.value &lt;lgl&gt; | Adjusted.P.value &lt;lgl&gt; | Old.P.value &lt;lgl&gt; | Old.Adjusted.P.value &lt;lgl&gt; | Odds.Ratio &lt;lgl&gt; | Combined.Score &lt;lgl&gt; | Genes &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "$Allen_Brain_Atlas_10x_scRNA_2021\n",
       "[1] Term                 Overlap              P.value             \n",
       "[4] Adjusted.P.value     Old.P.value          Old.Adjusted.P.value\n",
       "[7] Odds.Ratio           Combined.Score       Genes               \n",
       "<0 rows> (or 0-length row.names)\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "8f050580-5eb7-4d3f-a23d-0f6bcef96439",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in enriched$Allen_Brain_Atlas_10x_scRNA_2021: $ operator is invalid for atomic vectors\n",
     "output_type": "error",
     "traceback": [
      "Error in enriched$Allen_Brain_Atlas_10x_scRNA_2021: $ operator is invalid for atomic vectors\nTraceback:\n",
      "1. enriched$Allen_Brain_Atlas_10x_scRNA_2021 %>% as.data.table()",
      "2. as.data.table(.)"
     ]
    }
   ],
   "source": [
    "annotation <- enriched$Allen_Brain_Atlas_10x_scRNA_2021 %>% as.data.table()\n",
    "annotation$cluster <- 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e720ea52-811e-427a-8585-e9543ce0dda5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying Allen_Brain_Atlas_10x_scRNA_2021... Done.\n",
      "Parsing results... Done.\n",
      "EnrichR: 65.565 sec elapsed\n"
     ]
    }
   ],
   "source": [
    "enrichR_list <- annotation.enrichR()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8e93262a-92c2-44c7-adbf-c6aa85ac0f02",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$`0`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`1`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`2`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`3`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`4`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`5`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`6`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`7`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`8`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`9`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`10`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`11`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`12`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`13`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`14`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`15`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "\t<dt>$`16`</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.table: 0 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Term</th><th scope=col>Overlap</th><th scope=col>P.value</th><th scope=col>Adjusted.P.value</th><th scope=col>Old.P.value</th><th scope=col>Old.Adjusted.P.value</th><th scope=col>Odds.Ratio</th><th scope=col>Combined.Score</th><th scope=col>Genes</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$`0`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`1`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`2`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`3`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`4`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`5`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`6`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`7`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`8`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`9`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`10`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`11`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`12`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`13`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`14`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`15`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\item[\\$`16`] A data.table: 0 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " Term & Overlap & P.value & Adjusted.P.value & Old.P.value & Old.Adjusted.P.value & Odds.Ratio & Combined.Score & Genes\\\\\n",
       " <chr> & <chr> & <dbl> & <dbl> & <int> & <int> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$`0`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`1`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`2`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`3`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`4`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`5`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`6`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`7`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`8`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`9`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`10`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`11`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`12`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`13`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`14`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`15`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "$`16`\n",
       ":   \n",
       "A data.table: 0 × 9\n",
       "\n",
       "| Term &lt;chr&gt; | Overlap &lt;chr&gt; | P.value &lt;dbl&gt; | Adjusted.P.value &lt;dbl&gt; | Old.P.value &lt;int&gt; | Old.Adjusted.P.value &lt;int&gt; | Odds.Ratio &lt;dbl&gt; | Combined.Score &lt;dbl&gt; | Genes &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$`0`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`1`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`2`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`3`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`4`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`5`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`6`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`7`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`8`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`9`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`10`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`11`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`12`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`13`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`14`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`15`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n",
       "\n",
       "$`16`\n",
       "Empty data.table (0 rows and 9 cols): Term,Overlap,P.value,Adjusted.P.value,Old.P.value,Old.Adjusted.P.value...\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "enrichR_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b47008c9-9b0b-4078-b0af-0481f17fb29e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(enrichR)\n",
    "\n",
    "clusters <- unique(top_genes$cluster)\n",
    "all_results <- list()\n",
    "\n",
    "for (cl in clusters) {\n",
    "  gene_list <- top_genes %>% filter(cluster == cl) %>% pull(gene)\n",
    "  enr <- enrichr(gene_list, databases = \"Allen_Brain_Atlas_10x_scRNA_2021\")\n",
    "  res <- enr$Allen_Brain_Atlas_10x_scRNA_2021\n",
    "  res$cluster <- cl\n",
    "  all_results[[cl]] <- res\n",
    "}\n",
    "\n",
    "final_df <- bind_rows(all_results)\n",
    "write.csv(final_df, \"enrichr_all_clusters.csv\", row.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "289cbb99-9171-4cb1-8ac5-33e5540d3eb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "capture.output(enrichR_list, file = \"aa.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "7ace52d7-8534-4b89-bc76-f88a37bb2648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Human Inh L5-6 SST C4orf26 up'</li><li>'Human Inh L5-6 SST KLHL1 up'</li><li>'Human Inh L6 SST TH up'</li><li>'Human Inh L5-6 SST DNAJC14 up'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Human Inh L5-6 SST C4orf26 up'\n",
       "\\item 'Human Inh L5-6 SST KLHL1 up'\n",
       "\\item 'Human Inh L6 SST TH up'\n",
       "\\item 'Human Inh L5-6 SST DNAJC14 up'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Human Inh L5-6 SST C4orf26 up'\n",
       "2. 'Human Inh L5-6 SST KLHL1 up'\n",
       "3. 'Human Inh L6 SST TH up'\n",
       "4. 'Human Inh L5-6 SST DNAJC14 up'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"Human Inh L5-6 SST C4orf26 up\" \"Human Inh L5-6 SST KLHL1 up\"  \n",
       "[3] \"Human Inh L6 SST TH up\"        \"Human Inh L5-6 SST DNAJC14 up\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "enrichR_list[[2]]$Term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a581e3bf-f4ef-40b1-a9ae-99b588ec123e",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "tic(\"Perform enrichment analysis\")\n",
    "for (cl in unique(top_genes$cluster)) {\n",
    "    genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "    # Optionally, use Entrez IDs instead:\n",
    "    # genes_cluster <- top_genes %>% filter(cluster == cl) %>% pull(entrez)\n",
    "    \n",
    "    # Perform enrichment analysis\n",
    "    enriched <- enrichr(genes_cluster, databases = \"Allen_Brain_Atlas_10x_scRNA_2021\")\n",
    "    \n",
    "\n",
    "}\n",
    "\n",
    "toc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2a2101ef-6d06-4065-b356-0f90034e1911",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\"><tr><td>enrichr {enrichR}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Gene enrichment using Enrichr</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Gene enrichment using Enrichr\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>enrichr(\n",
       "  genes,\n",
       "  databases = NULL,\n",
       "  background = NULL,\n",
       "  include_overlap = FALSE,\n",
       "  sleepTime = 1\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr valign=\"top\"><td><code>genes</code></td>\n",
       "<td>\n",
       "<p>(Required). Character vector of Entrez gene symbols as input. A data.frame\n",
       "of gene symbols in first column is also acceptable, optionally a score denoting the\n",
       "degree of membership between 0 and 1 in the second column.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>databases</code></td>\n",
       "<td>\n",
       "<p>(Required). Character vector of databases to search.\n",
       "See https://maayanlab.cloud/Enrichr/ for available databases.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>background</code></td>\n",
       "<td>\n",
       "<p>(Optional). Character vector of Entrez gene symbols to be used as\n",
       "background. A data.frame of gene symbols in first column is also acceptable. \n",
       "Default is <code>\"NULL\"</code>. Enrichment analysis with background genes is only available \n",
       "on the main site (Enrichr). Also, it is using a different API service (Speedrichr), \n",
       "hence it is a little slower to complete and return the results.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>include_overlap</code></td>\n",
       "<td>\n",
       "<p>(Optional). Download database in GMT format to include 'Overlap'\n",
       "in the resulting data.frame when analysing with a background. Default is <code>\"FALSE\"</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>sleepTime</code></td>\n",
       "<td>\n",
       "<p>(Optional) Time to wait (in seconds) between sending requests to the server to prevent the same results being returned as the previous request. Default is 1.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>Gene enrichment using Enrichr, also, you can now try adding a background.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>Returns a list of data.frame of enrichment terms, p-values, ...\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Wajid Jawaid <a href=\"mailto:wajid.jawaid@gmail.com\">wajid.jawaid@gmail.com</a>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'># data(input) # Load example input genes\n",
       "# data(background) # Load example background genes\n",
       "# dbs &lt;- c(\"GO_Molecular_Function_2023\", \"GO_Cellular_Component_2023\",\n",
       "#          \"GO_Biological_Process_2023\")\n",
       "# if (getOption(\"enrichR.live\")) {\n",
       "#   enriched1 &lt;- enrichr(input, dbs)\n",
       "#   print(head(enriched1[[1]]))\n",
       "\n",
       "#   # Include background\n",
       "#   enriched2 &lt;- enrichr(input, dbs, background = background)\n",
       "#   print(head(enriched2[[1]]))\n",
       "\n",
       "#   # Include background and add 'Overlap' info\n",
       "#   enriched3 &lt;- enrichr(input, dbs, background = background, include_overlap = TRUE)\n",
       "#   print(head(enriched3[[1]]))\n",
       "# }\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>enrichR</em> version 3.4 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{enrichr}{Gene enrichment using Enrichr}{enrichr}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Gene enrichment using Enrichr\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "enrichr(\n",
       "  genes,\n",
       "  databases = NULL,\n",
       "  background = NULL,\n",
       "  include_overlap = FALSE,\n",
       "  sleepTime = 1\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{genes}] (Required). Character vector of Entrez gene symbols as input. A data.frame\n",
       "of gene symbols in first column is also acceptable, optionally a score denoting the\n",
       "degree of membership between 0 and 1 in the second column.\n",
       "\n",
       "\\item[\\code{databases}] (Required). Character vector of databases to search.\n",
       "See https://maayanlab.cloud/Enrichr/ for available databases.\n",
       "\n",
       "\\item[\\code{background}] (Optional). Character vector of Entrez gene symbols to be used as\n",
       "background. A data.frame of gene symbols in first column is also acceptable. \n",
       "Default is \\code{\"NULL\"}. Enrichment analysis with background genes is only available \n",
       "on the main site (Enrichr). Also, it is using a different API service (Speedrichr), \n",
       "hence it is a little slower to complete and return the results.\n",
       "\n",
       "\\item[\\code{include\\_overlap}] (Optional). Download database in GMT format to include 'Overlap'\n",
       "in the resulting data.frame when analysing with a background. Default is \\code{\"FALSE\"}.\n",
       "\n",
       "\\item[\\code{sleepTime}] (Optional) Time to wait (in seconds) between sending requests to the server to prevent the same results being returned as the previous request. Default is 1.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "Gene enrichment using Enrichr, also, you can now try adding a background.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "Returns a list of data.frame of enrichment terms, p-values, ...\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Wajid Jawaid \\email{wajid.jawaid@gmail.com}\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "# data(input) # Load example input genes\n",
       "# data(background) # Load example background genes\n",
       "# dbs <- c(\"GO_Molecular_Function_2023\", \"GO_Cellular_Component_2023\",\n",
       "#          \"GO_Biological_Process_2023\")\n",
       "# if (getOption(\"enrichR.live\")) {\n",
       "#   enriched1 <- enrichr(input, dbs)\n",
       "#   print(head(enriched1[[1]]))\n",
       "\n",
       "#   # Include background\n",
       "#   enriched2 <- enrichr(input, dbs, background = background)\n",
       "#   print(head(enriched2[[1]]))\n",
       "\n",
       "#   # Include background and add 'Overlap' info\n",
       "#   enriched3 <- enrichr(input, dbs, background = background, include_overlap = TRUE)\n",
       "#   print(head(enriched3[[1]]))\n",
       "# }\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "enrichr                package:enrichR                 R Documentation\n",
       "\n",
       "_\bG_\be_\bn_\be _\be_\bn_\br_\bi_\bc_\bh_\bm_\be_\bn_\bt _\bu_\bs_\bi_\bn_\bg _\bE_\bn_\br_\bi_\bc_\bh_\br\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Gene enrichment using Enrichr\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     enrichr(\n",
       "       genes,\n",
       "       databases = NULL,\n",
       "       background = NULL,\n",
       "       include_overlap = FALSE,\n",
       "       sleepTime = 1\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   genes: (Required). Character vector of Entrez gene symbols as input.\n",
       "          A data.frame of gene symbols in first column is also\n",
       "          acceptable, optionally a score denoting the degree of\n",
       "          membership between 0 and 1 in the second column.\n",
       "\n",
       "databases: (Required). Character vector of databases to search. See\n",
       "          https://maayanlab.cloud/Enrichr/ for available databases.\n",
       "\n",
       "background: (Optional). Character vector of Entrez gene symbols to be\n",
       "          used as background. A data.frame of gene symbols in first\n",
       "          column is also acceptable.  Default is ‘\"NULL\"’. Enrichment\n",
       "          analysis with background genes is only available on the main\n",
       "          site (Enrichr). Also, it is using a different API service\n",
       "          (Speedrichr), hence it is a little slower to complete and\n",
       "          return the results.\n",
       "\n",
       "include_overlap: (Optional). Download database in GMT format to include\n",
       "          'Overlap' in the resulting data.frame when analysing with a\n",
       "          background. Default is ‘\"FALSE\"’.\n",
       "\n",
       "sleepTime: (Optional) Time to wait (in seconds) between sending\n",
       "          requests to the server to prevent the same results being\n",
       "          returned as the previous request. Default is 1.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     Gene enrichment using Enrichr, also, you can now try adding a\n",
       "     background.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     Returns a list of data.frame of enrichment terms, p-values, ...\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Wajid Jawaid <mailto:wajid.jawaid@gmail.com>\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     # data(input) # Load example input genes\n",
       "     # data(background) # Load example background genes\n",
       "     # dbs <- c(\"GO_Molecular_Function_2023\", \"GO_Cellular_Component_2023\",\n",
       "     #          \"GO_Biological_Process_2023\")\n",
       "     # if (getOption(\"enrichR.live\")) {\n",
       "     #   enriched1 <- enrichr(input, dbs)\n",
       "     #   print(head(enriched1[[1]]))\n",
       "     \n",
       "     #   # Include background\n",
       "     #   enriched2 <- enrichr(input, dbs, background = background)\n",
       "     #   print(head(enriched2[[1]]))\n",
       "     \n",
       "     #   # Include background and add 'Overlap' info\n",
       "     #   enriched3 <- enrichr(input, dbs, background = background, include_overlap = TRUE)\n",
       "     #   print(head(enriched3[[1]]))\n",
       "     # }\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?enrichr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd19d997-b381-4968-8345-5f47e4136d50",
   "metadata": {},
   "source": [
    "### singleR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fd6176b9-9266-49e0-86f5-7c7b3b85de3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.16 (BiocManager 1.30.22), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Installing package(s) 'devtools'\n",
      "\n",
      "also installing the dependencies ‘credentials’, ‘zip’, ‘gitcreds’, ‘ini’, ‘diffobj’, ‘gert’, ‘gh’, ‘whisker’, ‘downlit’, ‘httr2’, ‘rmarkdown’, ‘xopen’, ‘brew’, ‘brio’, ‘evaluate’, ‘praise’, ‘waldo’, ‘usethis’, ‘desc’, ‘pkgbuild’, ‘pkgdown’, ‘pkgload’, ‘profvis’, ‘rcmdcheck’, ‘roxygen2’, ‘rversions’, ‘sessioninfo’, ‘testthat’, ‘urlchecker’\n",
      "\n",
      "\n",
      "Old packages: 'abind', 'ape', 'BH', 'BiocManager', 'bit', 'bit64', 'bitops',\n",
      "  'bslib', 'cachem', 'caTools', 'colorspace', 'commonmark', 'cowplot',\n",
      "  'crayon', 'crosstalk', 'curl', 'data.table', 'DBI', 'deldir', 'DEoptimR',\n",
      "  'digest', 'docopt', 'dotCall64', 'dqrng', 'farver', 'fastDummies', 'fastICA',\n",
      "  'fastmap', 'fitdistrplus', 'FNN', 'fontawesome', 'fs', 'future',\n",
      "  'future.apply', 'ggplot2', 'ggrepel', 'ggridges', 'globals', 'gplots',\n",
      "  'gtable', 'gtools', 'hdf5r', 'highr', 'htmltools', 'htmlwidgets', 'httpuv',\n",
      "  'igraph', 'kernlab', 'knitr', 'later', 'leiden', 'leidenbase', 'listenv',\n",
      "  'littler', 'locfit', 'mathjaxr', 'MatrixModels', 'matrixStats', 'metap',\n",
      "  'miniUI', 'mixtools', 'multcomp', 'munsell', 'mvtnorm', 'parallelly',\n",
      "  'patchwork', 'pheatmap', 'plotly', 'plotrix', 'polyclip', 'progress',\n",
      "  'progressr', 'promises', 'qlcMatrix', 'quantreg', 'R.oo', 'R.utils', 'RANN',\n",
      "  'rbibutils', 'Rcpp', 'RcppAnnoy', 'RcppArmadillo', 'RcppEigen', 'RcppHNSW',\n",
      "  'RcppParallel', 'RcppTOML', 'RcppZiggurat', 'RCurl', 'Rdpack', 'remotes',\n",
      "  'restfulr', 'reticulate', 'Rfast', 'Rfast2', 'rjson', 'robustbase',\n",
      "  'rprojroot', 'RSpectra', 'RSQLite', 'Rtsne', 'RUnit', 'sandwich', 'sass',\n",
      "  'scales', 'sctransform', 'segmented', 'Seurat', 'SeuratObject', 'shiny',\n",
      "  'slam', 'sp', 'spam', 'SparseM', 'spatstat.data', 'spatstat.explore',\n",
      "  'spatstat.geom', 'spatstat.random', 'spatstat.sparse', 'spatstat.utils',\n",
      "  'tensor', 'TH.data', 'tibble', 'tinytex', 'utf8', 'uwot', 'VGAM', 'viridis',\n",
      "  'WriteXLS', 'xfun', 'XML', 'yaml', 'zoo', 'boot', 'class', 'cluster',\n",
      "  'codetools', 'foreign', 'KernSmooth', 'lattice', 'mgcv', 'nlme', 'nnet',\n",
      "  'rpart', 'spatial', 'survival'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"devtools\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c3fa7d35-a3a0-4288-8aaf-e94b3936208a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo dviraran/SingleR@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error in utils::download.file(url, path, method = method, quiet = quiet,  : \n",
      "  download from 'https://api.github.com/repos/dviraran/SingleR/tarball/HEAD' failed\n"
     ]
    }
   ],
   "source": [
    "devtools::install_github('dviraran/SingleR')\n",
    "# this might take long, though mostly because of the installation of Seurat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b245a16d-2912-4970-8542-8adf132d705c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.16 (BiocManager 1.30.22), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Installing package(s) 'SingleR'\n",
      "\n",
      "also installing the dependencies ‘ScaledMatrix’, ‘rsvd’, ‘BiocSingular’, ‘BiocNeighbors’\n",
      "\n",
      "\n",
      "Old packages: 'abind', 'ape', 'BH', 'BiocManager', 'bit', 'bit64', 'bitops',\n",
      "  'bslib', 'cachem', 'caTools', 'colorspace', 'commonmark', 'cowplot',\n",
      "  'crayon', 'crosstalk', 'curl', 'data.table', 'DBI', 'deldir', 'DEoptimR',\n",
      "  'digest', 'docopt', 'dotCall64', 'dqrng', 'farver', 'fastDummies', 'fastICA',\n",
      "  'fastmap', 'fitdistrplus', 'FNN', 'fontawesome', 'fs', 'future',\n",
      "  'future.apply', 'ggplot2', 'ggrepel', 'ggridges', 'globals', 'gplots',\n",
      "  'gtable', 'gtools', 'hdf5r', 'highr', 'htmltools', 'htmlwidgets', 'httpuv',\n",
      "  'igraph', 'kernlab', 'knitr', 'later', 'leiden', 'leidenbase', 'listenv',\n",
      "  'littler', 'locfit', 'mathjaxr', 'MatrixModels', 'matrixStats', 'metap',\n",
      "  'miniUI', 'mixtools', 'multcomp', 'munsell', 'mvtnorm', 'parallelly',\n",
      "  'patchwork', 'pheatmap', 'plotly', 'plotrix', 'polyclip', 'progress',\n",
      "  'progressr', 'promises', 'qlcMatrix', 'quantreg', 'R.oo', 'R.utils', 'RANN',\n",
      "  'rbibutils', 'Rcpp', 'RcppAnnoy', 'RcppArmadillo', 'RcppEigen', 'RcppHNSW',\n",
      "  'RcppParallel', 'RcppTOML', 'RcppZiggurat', 'RCurl', 'Rdpack', 'remotes',\n",
      "  'restfulr', 'reticulate', 'Rfast', 'Rfast2', 'rjson', 'robustbase',\n",
      "  'rprojroot', 'RSpectra', 'RSQLite', 'Rtsne', 'RUnit', 'sandwich', 'sass',\n",
      "  'scales', 'sctransform', 'segmented', 'Seurat', 'SeuratObject', 'shiny',\n",
      "  'slam', 'sp', 'spam', 'SparseM', 'spatstat.data', 'spatstat.explore',\n",
      "  'spatstat.geom', 'spatstat.random', 'spatstat.sparse', 'spatstat.utils',\n",
      "  'tensor', 'TH.data', 'tibble', 'tinytex', 'utf8', 'uwot', 'VGAM', 'viridis',\n",
      "  'WriteXLS', 'xfun', 'XML', 'yaml', 'zoo', 'boot', 'class', 'cluster',\n",
      "  'codetools', 'foreign', 'KernSmooth', 'lattice', 'mgcv', 'nlme', 'nnet',\n",
      "  'rpart', 'spatial', 'survival'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(\"SingleR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7c641651-41db-4cb5-8d44-06a845955053",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(SingleR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e0208f8-6381-45b1-922f-403e821746e0",
   "metadata": {},
   "source": [
    "### GPTCelltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7e0d10d-f8f2-4762-95fe-2e5b15e2cfa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "bdcf9498-d9a0-4c2d-b5ba-4a448574f1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.setenv(OPENAI_API_KEY = \"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\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "72ef24ba-cfe9-4564-9148-964e7f1eaa36",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl>\n",
       "\t<dt>$object</dt>\n",
       "\t\t<dd>'list'</dd>\n",
       "\t<dt>$data</dt>\n",
       "\t\t<dd><table class=\"dataframe\">\n",
       "<caption>A data.frame: 50 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>object</th><th scope=col>created</th><th scope=col>owned_by</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>text-embedding-ada-002               </td><td>model</td><td>1671217299</td><td>openai-internal</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>whisper-1                            </td><td>model</td><td>1677532384</td><td>openai-internal</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>gpt-3.5-turbo                        </td><td>model</td><td>1677610602</td><td>openai         </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>tts-1                                </td><td>model</td><td>1681940951</td><td>openai-internal</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>gpt-3.5-turbo-16k                    </td><td>model</td><td>1683758102</td><td>openai-internal</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>davinci-002                          </td><td>model</td><td>1692634301</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>7</th><td>babbage-002                          </td><td>model</td><td>1692634615</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>8</th><td>gpt-3.5-turbo-instruct               </td><td>model</td><td>1692901427</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>9</th><td>gpt-3.5-turbo-instruct-0914          </td><td>model</td><td>1694122472</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>10</th><td>dall-e-3                             </td><td>model</td><td>1698785189</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>11</th><td>dall-e-2                             </td><td>model</td><td>1698798177</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>12</th><td>gpt-3.5-turbo-1106                   </td><td>model</td><td>1698959748</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>13</th><td>tts-1-hd                             </td><td>model</td><td>1699046015</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>14</th><td>tts-1-1106                           </td><td>model</td><td>1699053241</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>15</th><td>tts-1-hd-1106                        </td><td>model</td><td>1699053533</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>16</th><td>text-embedding-3-small               </td><td>model</td><td>1705948997</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>17</th><td>text-embedding-3-large               </td><td>model</td><td>1705953180</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>18</th><td>gpt-3.5-turbo-0125                   </td><td>model</td><td>1706048358</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>19</th><td>gpt-4o                               </td><td>model</td><td>1715367049</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>20</th><td>gpt-4o-2024-05-13                    </td><td>model</td><td>1715368132</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>21</th><td>gpt-4o-mini-2024-07-18               </td><td>model</td><td>1721172717</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>22</th><td>gpt-4o-mini                          </td><td>model</td><td>1721172741</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>23</th><td>gpt-4o-2024-08-06                    </td><td>model</td><td>1722814719</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>24</th><td>o1-preview-2024-09-12                </td><td>model</td><td>1725648865</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>25</th><td>o1-preview                           </td><td>model</td><td>1725648897</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>26</th><td>o1-mini-2024-09-12                   </td><td>model</td><td>1725648979</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>27</th><td>o1-mini                              </td><td>model</td><td>1725649008</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>28</th><td>gpt-4o-audio-preview-2024-10-01      </td><td>model</td><td>1727389042</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>29</th><td>gpt-4o-audio-preview                 </td><td>model</td><td>1727460443</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>30</th><td>omni-moderation-latest               </td><td>model</td><td>1731689265</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>31</th><td>omni-moderation-2024-09-26           </td><td>model</td><td>1732734466</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>32</th><td>gpt-4o-mini-audio-preview-2024-12-17 </td><td>model</td><td>1734115920</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>33</th><td>gpt-4o-mini-audio-preview            </td><td>model</td><td>1734387424</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>34</th><td>gpt-4o-2024-11-20                    </td><td>model</td><td>1739331543</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>35</th><td>gpt-4.5-preview                      </td><td>model</td><td>1740623059</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>36</th><td>gpt-4.5-preview-2025-02-27           </td><td>model</td><td>1740623304</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>37</th><td>gpt-4o-search-preview-2025-03-11     </td><td>model</td><td>1741388170</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>38</th><td>gpt-4o-search-preview                </td><td>model</td><td>1741388720</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>39</th><td>gpt-4o-mini-search-preview-2025-03-11</td><td>model</td><td>1741390858</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>40</th><td>gpt-4o-mini-search-preview           </td><td>model</td><td>1741391161</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>41</th><td>gpt-4o-transcribe                    </td><td>model</td><td>1742068463</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>42</th><td>gpt-4o-mini-transcribe               </td><td>model</td><td>1742068596</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>43</th><td>gpt-4o-mini-tts                      </td><td>model</td><td>1742403959</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>44</th><td>gpt-4.1-2025-04-14                   </td><td>model</td><td>1744315746</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>45</th><td>gpt-4.1                              </td><td>model</td><td>1744316542</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>46</th><td>gpt-4.1-mini-2025-04-14              </td><td>model</td><td>1744317547</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>47</th><td>gpt-4.1-mini                         </td><td>model</td><td>1744318173</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>48</th><td>gpt-4.1-nano-2025-04-14              </td><td>model</td><td>1744321025</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>49</th><td>gpt-4.1-nano                         </td><td>model</td><td>1744321707</td><td>system         </td></tr>\n",
       "\t<tr><th scope=row>50</th><td>gpt-image-1                          </td><td>model</td><td>1745517030</td><td>system         </td></tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</dd>\n",
       "</dl>\n"
      ],
      "text/latex": [
       "\\begin{description}\n",
       "\\item[\\$object] 'list'\n",
       "\\item[\\$data] A data.frame: 50 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & id & object & created & owned\\_by\\\\\n",
       "  & <chr> & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & text-embedding-ada-002                & model & 1671217299 & openai-internal\\\\\n",
       "\t2 & whisper-1                             & model & 1677532384 & openai-internal\\\\\n",
       "\t3 & gpt-3.5-turbo                         & model & 1677610602 & openai         \\\\\n",
       "\t4 & tts-1                                 & model & 1681940951 & openai-internal\\\\\n",
       "\t5 & gpt-3.5-turbo-16k                     & model & 1683758102 & openai-internal\\\\\n",
       "\t6 & davinci-002                           & model & 1692634301 & system         \\\\\n",
       "\t7 & babbage-002                           & model & 1692634615 & system         \\\\\n",
       "\t8 & gpt-3.5-turbo-instruct                & model & 1692901427 & system         \\\\\n",
       "\t9 & gpt-3.5-turbo-instruct-0914           & model & 1694122472 & system         \\\\\n",
       "\t10 & dall-e-3                              & model & 1698785189 & system         \\\\\n",
       "\t11 & dall-e-2                              & model & 1698798177 & system         \\\\\n",
       "\t12 & gpt-3.5-turbo-1106                    & model & 1698959748 & system         \\\\\n",
       "\t13 & tts-1-hd                              & model & 1699046015 & system         \\\\\n",
       "\t14 & tts-1-1106                            & model & 1699053241 & system         \\\\\n",
       "\t15 & tts-1-hd-1106                         & model & 1699053533 & system         \\\\\n",
       "\t16 & text-embedding-3-small                & model & 1705948997 & system         \\\\\n",
       "\t17 & text-embedding-3-large                & model & 1705953180 & system         \\\\\n",
       "\t18 & gpt-3.5-turbo-0125                    & model & 1706048358 & system         \\\\\n",
       "\t19 & gpt-4o                                & model & 1715367049 & system         \\\\\n",
       "\t20 & gpt-4o-2024-05-13                     & model & 1715368132 & system         \\\\\n",
       "\t21 & gpt-4o-mini-2024-07-18                & model & 1721172717 & system         \\\\\n",
       "\t22 & gpt-4o-mini                           & model & 1721172741 & system         \\\\\n",
       "\t23 & gpt-4o-2024-08-06                     & model & 1722814719 & system         \\\\\n",
       "\t24 & o1-preview-2024-09-12                 & model & 1725648865 & system         \\\\\n",
       "\t25 & o1-preview                            & model & 1725648897 & system         \\\\\n",
       "\t26 & o1-mini-2024-09-12                    & model & 1725648979 & system         \\\\\n",
       "\t27 & o1-mini                               & model & 1725649008 & system         \\\\\n",
       "\t28 & gpt-4o-audio-preview-2024-10-01       & model & 1727389042 & system         \\\\\n",
       "\t29 & gpt-4o-audio-preview                  & model & 1727460443 & system         \\\\\n",
       "\t30 & omni-moderation-latest                & model & 1731689265 & system         \\\\\n",
       "\t31 & omni-moderation-2024-09-26            & model & 1732734466 & system         \\\\\n",
       "\t32 & gpt-4o-mini-audio-preview-2024-12-17  & model & 1734115920 & system         \\\\\n",
       "\t33 & gpt-4o-mini-audio-preview             & model & 1734387424 & system         \\\\\n",
       "\t34 & gpt-4o-2024-11-20                     & model & 1739331543 & system         \\\\\n",
       "\t35 & gpt-4.5-preview                       & model & 1740623059 & system         \\\\\n",
       "\t36 & gpt-4.5-preview-2025-02-27            & model & 1740623304 & system         \\\\\n",
       "\t37 & gpt-4o-search-preview-2025-03-11      & model & 1741388170 & system         \\\\\n",
       "\t38 & gpt-4o-search-preview                 & model & 1741388720 & system         \\\\\n",
       "\t39 & gpt-4o-mini-search-preview-2025-03-11 & model & 1741390858 & system         \\\\\n",
       "\t40 & gpt-4o-mini-search-preview            & model & 1741391161 & system         \\\\\n",
       "\t41 & gpt-4o-transcribe                     & model & 1742068463 & system         \\\\\n",
       "\t42 & gpt-4o-mini-transcribe                & model & 1742068596 & system         \\\\\n",
       "\t43 & gpt-4o-mini-tts                       & model & 1742403959 & system         \\\\\n",
       "\t44 & gpt-4.1-2025-04-14                    & model & 1744315746 & system         \\\\\n",
       "\t45 & gpt-4.1                               & model & 1744316542 & system         \\\\\n",
       "\t46 & gpt-4.1-mini-2025-04-14               & model & 1744317547 & system         \\\\\n",
       "\t47 & gpt-4.1-mini                          & model & 1744318173 & system         \\\\\n",
       "\t48 & gpt-4.1-nano-2025-04-14               & model & 1744321025 & system         \\\\\n",
       "\t49 & gpt-4.1-nano                          & model & 1744321707 & system         \\\\\n",
       "\t50 & gpt-image-1                           & model & 1745517030 & system         \\\\\n",
       "\\end{tabular}\n",
       "\n",
       "\\end{description}\n"
      ],
      "text/markdown": [
       "$object\n",
       ":   'list'\n",
       "$data\n",
       ":   \n",
       "A data.frame: 50 × 4\n",
       "\n",
       "| <!--/--> | id &lt;chr&gt; | object &lt;chr&gt; | created &lt;int&gt; | owned_by &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 1 | text-embedding-ada-002                | model | 1671217299 | openai-internal |\n",
       "| 2 | whisper-1                             | model | 1677532384 | openai-internal |\n",
       "| 3 | gpt-3.5-turbo                         | model | 1677610602 | openai          |\n",
       "| 4 | tts-1                                 | model | 1681940951 | openai-internal |\n",
       "| 5 | gpt-3.5-turbo-16k                     | model | 1683758102 | openai-internal |\n",
       "| 6 | davinci-002                           | model | 1692634301 | system          |\n",
       "| 7 | babbage-002                           | model | 1692634615 | system          |\n",
       "| 8 | gpt-3.5-turbo-instruct                | model | 1692901427 | system          |\n",
       "| 9 | gpt-3.5-turbo-instruct-0914           | model | 1694122472 | system          |\n",
       "| 10 | dall-e-3                              | model | 1698785189 | system          |\n",
       "| 11 | dall-e-2                              | model | 1698798177 | system          |\n",
       "| 12 | gpt-3.5-turbo-1106                    | model | 1698959748 | system          |\n",
       "| 13 | tts-1-hd                              | model | 1699046015 | system          |\n",
       "| 14 | tts-1-1106                            | model | 1699053241 | system          |\n",
       "| 15 | tts-1-hd-1106                         | model | 1699053533 | system          |\n",
       "| 16 | text-embedding-3-small                | model | 1705948997 | system          |\n",
       "| 17 | text-embedding-3-large                | model | 1705953180 | system          |\n",
       "| 18 | gpt-3.5-turbo-0125                    | model | 1706048358 | system          |\n",
       "| 19 | gpt-4o                                | model | 1715367049 | system          |\n",
       "| 20 | gpt-4o-2024-05-13                     | model | 1715368132 | system          |\n",
       "| 21 | gpt-4o-mini-2024-07-18                | model | 1721172717 | system          |\n",
       "| 22 | gpt-4o-mini                           | model | 1721172741 | system          |\n",
       "| 23 | gpt-4o-2024-08-06                     | model | 1722814719 | system          |\n",
       "| 24 | o1-preview-2024-09-12                 | model | 1725648865 | system          |\n",
       "| 25 | o1-preview                            | model | 1725648897 | system          |\n",
       "| 26 | o1-mini-2024-09-12                    | model | 1725648979 | system          |\n",
       "| 27 | o1-mini                               | model | 1725649008 | system          |\n",
       "| 28 | gpt-4o-audio-preview-2024-10-01       | model | 1727389042 | system          |\n",
       "| 29 | gpt-4o-audio-preview                  | model | 1727460443 | system          |\n",
       "| 30 | omni-moderation-latest                | model | 1731689265 | system          |\n",
       "| 31 | omni-moderation-2024-09-26            | model | 1732734466 | system          |\n",
       "| 32 | gpt-4o-mini-audio-preview-2024-12-17  | model | 1734115920 | system          |\n",
       "| 33 | gpt-4o-mini-audio-preview             | model | 1734387424 | system          |\n",
       "| 34 | gpt-4o-2024-11-20                     | model | 1739331543 | system          |\n",
       "| 35 | gpt-4.5-preview                       | model | 1740623059 | system          |\n",
       "| 36 | gpt-4.5-preview-2025-02-27            | model | 1740623304 | system          |\n",
       "| 37 | gpt-4o-search-preview-2025-03-11      | model | 1741388170 | system          |\n",
       "| 38 | gpt-4o-search-preview                 | model | 1741388720 | system          |\n",
       "| 39 | gpt-4o-mini-search-preview-2025-03-11 | model | 1741390858 | system          |\n",
       "| 40 | gpt-4o-mini-search-preview            | model | 1741391161 | system          |\n",
       "| 41 | gpt-4o-transcribe                     | model | 1742068463 | system          |\n",
       "| 42 | gpt-4o-mini-transcribe                | model | 1742068596 | system          |\n",
       "| 43 | gpt-4o-mini-tts                       | model | 1742403959 | system          |\n",
       "| 44 | gpt-4.1-2025-04-14                    | model | 1744315746 | system          |\n",
       "| 45 | gpt-4.1                               | model | 1744316542 | system          |\n",
       "| 46 | gpt-4.1-mini-2025-04-14               | model | 1744317547 | system          |\n",
       "| 47 | gpt-4.1-mini                          | model | 1744318173 | system          |\n",
       "| 48 | gpt-4.1-nano-2025-04-14               | model | 1744321025 | system          |\n",
       "| 49 | gpt-4.1-nano                          | model | 1744321707 | system          |\n",
       "| 50 | gpt-image-1                           | model | 1745517030 | system          |\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "$object\n",
       "[1] \"list\"\n",
       "\n",
       "$data\n",
       "                                      id object    created        owned_by\n",
       "1                 text-embedding-ada-002  model 1671217299 openai-internal\n",
       "2                              whisper-1  model 1677532384 openai-internal\n",
       "3                          gpt-3.5-turbo  model 1677610602          openai\n",
       "4                                  tts-1  model 1681940951 openai-internal\n",
       "5                      gpt-3.5-turbo-16k  model 1683758102 openai-internal\n",
       "6                            davinci-002  model 1692634301          system\n",
       "7                            babbage-002  model 1692634615          system\n",
       "8                 gpt-3.5-turbo-instruct  model 1692901427          system\n",
       "9            gpt-3.5-turbo-instruct-0914  model 1694122472          system\n",
       "10                              dall-e-3  model 1698785189          system\n",
       "11                              dall-e-2  model 1698798177          system\n",
       "12                    gpt-3.5-turbo-1106  model 1698959748          system\n",
       "13                              tts-1-hd  model 1699046015          system\n",
       "14                            tts-1-1106  model 1699053241          system\n",
       "15                         tts-1-hd-1106  model 1699053533          system\n",
       "16                text-embedding-3-small  model 1705948997          system\n",
       "17                text-embedding-3-large  model 1705953180          system\n",
       "18                    gpt-3.5-turbo-0125  model 1706048358          system\n",
       "19                                gpt-4o  model 1715367049          system\n",
       "20                     gpt-4o-2024-05-13  model 1715368132          system\n",
       "21                gpt-4o-mini-2024-07-18  model 1721172717          system\n",
       "22                           gpt-4o-mini  model 1721172741          system\n",
       "23                     gpt-4o-2024-08-06  model 1722814719          system\n",
       "24                 o1-preview-2024-09-12  model 1725648865          system\n",
       "25                            o1-preview  model 1725648897          system\n",
       "26                    o1-mini-2024-09-12  model 1725648979          system\n",
       "27                               o1-mini  model 1725649008          system\n",
       "28       gpt-4o-audio-preview-2024-10-01  model 1727389042          system\n",
       "29                  gpt-4o-audio-preview  model 1727460443          system\n",
       "30                omni-moderation-latest  model 1731689265          system\n",
       "31            omni-moderation-2024-09-26  model 1732734466          system\n",
       "32  gpt-4o-mini-audio-preview-2024-12-17  model 1734115920          system\n",
       "33             gpt-4o-mini-audio-preview  model 1734387424          system\n",
       "34                     gpt-4o-2024-11-20  model 1739331543          system\n",
       "35                       gpt-4.5-preview  model 1740623059          system\n",
       "36            gpt-4.5-preview-2025-02-27  model 1740623304          system\n",
       "37      gpt-4o-search-preview-2025-03-11  model 1741388170          system\n",
       "38                 gpt-4o-search-preview  model 1741388720          system\n",
       "39 gpt-4o-mini-search-preview-2025-03-11  model 1741390858          system\n",
       "40            gpt-4o-mini-search-preview  model 1741391161          system\n",
       "41                     gpt-4o-transcribe  model 1742068463          system\n",
       "42                gpt-4o-mini-transcribe  model 1742068596          system\n",
       "43                       gpt-4o-mini-tts  model 1742403959          system\n",
       "44                    gpt-4.1-2025-04-14  model 1744315746          system\n",
       "45                               gpt-4.1  model 1744316542          system\n",
       "46               gpt-4.1-mini-2025-04-14  model 1744317547          system\n",
       "47                          gpt-4.1-mini  model 1744318173          system\n",
       "48               gpt-4.1-nano-2025-04-14  model 1744321025          system\n",
       "49                          gpt-4.1-nano  model 1744321707          system\n",
       "50                           gpt-image-1  model 1745517030          system\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "openai::list_models()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "b9c989fb-6fe0-4e76-8f60-221ad4ff85d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "22758"
      ],
      "text/latex": [
       "22758"
      ],
      "text/markdown": [
       "22758"
      ],
      "text/plain": [
       "[1] 22758"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "8690"
      ],
      "text/latex": [
       "8690"
      ],
      "text/markdown": [
       "8690"
      ],
      "text/plain": [
       "[1] 8690"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(cluster_markers)\n",
    "dt <- cluster_markers[cluster_markers$cluster %in% c(0:10),]\n",
    "nrow(dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "5399c216-efa9-4d5f-a141-0cadc2885220",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "17"
      ],
      "text/latex": [
       "17"
      ],
      "text/markdown": [
       "17"
      ],
      "text/plain": [
       "[1] 17"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(dt$cluster))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "c8b1fce1-0f2d-41de-b117-622170bd2340",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "850"
      ],
      "text/latex": [
       "850"
      ],
      "text/markdown": [
       "850"
      ],
      "text/plain": [
       "[1] 850"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "50"
      ],
      "text/latex": [
       "50"
      ],
      "text/markdown": [
       "50"
      ],
      "text/plain": [
       "[1] 50"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dt <- cluster_markers %>% group_by(cluster) %>% top_n(n = 50, wt = avg_log2FC) %>% as.data.frame()\n",
    "nrow(dt)\n",
    "dt <- dt[dt$cluster %in% c(0),]\n",
    "nrow(dt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "a6f61444-ad19-4266-8330-cf36e1512405",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated tokens: 837.5 \n"
     ]
    }
   ],
   "source": [
    "# Suppose df is your FindAllMarkers result\n",
    "csv_string <- paste(capture.output(write.csv(dt, row.names = FALSE)), collapse = \"\\n\")\n",
    "char_count <- nchar(csv_string)\n",
    "token_estimate <- char_count / 4\n",
    "cat(\"Estimated tokens:\", token_estimate, \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "9276c465-8ad6-4b57-b18d-9da1a4be1530",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated tokens: 272570.2 \n"
     ]
    }
   ],
   "source": [
    "# Suppose df is your FindAllMarkers result\n",
    "csv_string <- paste(capture.output(write.csv(cluster_markers, row.names = FALSE)), collapse = \"\\n\")\n",
    "char_count <- nchar(csv_string)\n",
    "token_estimate <- char_count / 4\n",
    "cat(\"Estimated tokens:\", token_estimate, \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "e473449e-89e4-4ffd-a0ce-2af6ec69f0d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Note: OpenAI API key found: returning the cell type annotations.\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error: OpenAI API request failed [429]:\n\nYou exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.\n",
     "output_type": "error",
     "traceback": [
      "Error: OpenAI API request failed [429]:\n\nYou exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.\nTraceback:\n",
      "1. gptcelltype(cluster_markers, tissuename = \"Human brain\", model = \"gpt-3.5-turbo\")",
      "2. sapply(1:cutnum, function(i) {\n .     id <- which(cid == i)\n .     flag <- 0\n .     while (flag == 0) {\n .         k <- openai::create_chat_completion(model = model, message = list(list(role = \"user\", \n .             content = paste0(\"Identify cell types of \", tissuename, \n .                 \" cells using the following markers separately for each\\n row. Only provide the cell type name. Do not show numbers before the name.\\n Some can be a mixture of multiple cell types.\\n\", \n .                 paste(input[id], collapse = \"\\n\")))))\n .         res <- strsplit(k$choices[, \"message.content\"], \"\\n\")[[1]]\n .         if (length(res) == length(id)) \n .             flag <- 1\n .     }\n .     names(res) <- names(input)[id]\n .     res\n . }, simplify = F)",
      "3. lapply(X = X, FUN = FUN, ...)",
      "4. FUN(X[[i]], ...)",
      "5. openai::create_chat_completion(model = model, message = list(list(role = \"user\", \n .     content = paste0(\"Identify cell types of \", tissuename, \" cells using the following markers separately for each\\n row. Only provide the cell type name. Do not show numbers before the name.\\n Some can be a mixture of multiple cell types.\\n\", \n .         paste(input[id], collapse = \"\\n\")))))",
      "6. paste0(\"OpenAI API request failed [\", httr::status_code(response), \n .     \"]:\\n\\n\", parsed$error$message) %>% stop(call. = FALSE)",
      "7. stop(., call. = FALSE)"
     ]
    }
   ],
   "source": [
    "library(GPTCelltype)\n",
    "library(openai)\n",
    "\n",
    "tic()\n",
    "res <- gptcelltype(\n",
    "    cluster_markers, tissuename = \"Human brain\",\n",
    "    model = \"gpt-4\"\n",
    ")\n",
    "toc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "80e2c613-babe-43c0-bd54-1d0cbecd2ec4",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: 'num_tokens_from_messages' is not an exported object from 'namespace:openai'\n",
     "output_type": "error",
     "traceback": [
      "Error: 'num_tokens_from_messages' is not an exported object from 'namespace:openai'\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "openai::num_tokens_from_messages(\n",
    "  messages = list(list(role = \"user\", content = \"how can i calculate how many tocken will i need to do soemthing with the gpt 4 api? I what to use gptcelltype to annotate some cells. I need to do it a few times\")),\n",
    "  model = \"gpt-4\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "3b608d15-42d8-489b-b873-38f15c3f3f91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\"><tr><td>gptcelltype {GPTCelltype}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Cell type annotation with OpenAI GPT models</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>Cell type annotation with OpenAI GPT models\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>gptcelltype(input, tissuename = NULL, model = \"gpt-4\", topgenenumber = 10)\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr valign=\"top\"><td><code>input</code></td>\n",
       "<td>\n",
       "<p>Either the differential gene table returned by Seurat FindAllMarkers() function, or a list of genes.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>tissuename</code></td>\n",
       "<td>\n",
       "<p>Optional input of tissue name.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>model</code></td>\n",
       "<td>\n",
       "<p>A valid GPT-4 or GPT-3.5 model name list on https://platform.openai.com/docs/models. Default is 'gpt-4-32k'.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>topgenenumber</code></td>\n",
       "<td>\n",
       "<p>Number of top differential genes to be used if input is Seurat differential genes.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>openai_key</code></td>\n",
       "<td>\n",
       "<p>The OpenAI key obtained from https://platform.openai.com/account/api-keys The default is NA, which will resulting outputing the prompt itself. If an actual key is provided, then the output will be the celltype annotations from the GPT model specified by the user.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>Annotate cell types by OpenAI GPT models in a Seurat pipeline or with a custom gene list. If used in a Seurat pipeline, Seurat FindAllMarkers() function needs to be run first and the differential gene table generated by Seurat will serve as the input. If the input is a custom list of genes, one cell type is identified for each element in the list.  Note: system environment should have variable OPENAI_API_KEY = 'your_api_key' or &rdquo;. The OpenAI key can be obtained from https://platform.openai.com/account/api-keys. If &rdquo;, then output the prompt itself. If an actual key is provided, then the output will be the celltype annotations from the GPT model specified by the user.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A vector of cell types when the user provide openai_key or the prompt itself when openai_key = NA (default)\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Author(s)</h3>\n",
       "\n",
       "<p>Wenpin Hou &lt;wh2526@cumc.columbia.edu&gt;, Zhicheng Ji &lt;zhicheng.ji@duke.edu&gt;\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre><code class='language-R'># Gene list as input, OPENAI_API_KEY = NA (default)\n",
       "Sys.setenv(OPENAI_API_KEY = '')\n",
       "print(Sys.getenv('OPENAI_API_KEY'))  \n",
       "gptcelltype(input = list(gs1=c('CD4,CD3D'),gs2='CD14'), tissuename=NULL, model='gpt-4', topgenenumber = 10) \n",
       "# Seurat object as input, openai_key = NA (default)\n",
       "#You have to install Seurat package first to run this example\n",
       "library(Seurat)\n",
       "data(\"pbmc_small\")\n",
       "all.markers &lt;- FindAllMarkers(object = pbmc_small)\n",
       "Sys.setenv(OPENAI_API_KEY = '')\n",
       "print(Sys.getenv('OPENAI_API_KEY'))  \n",
       "prompt &lt;- gptcelltype(all.markers, tissuename='human PBMC')\n",
       "return(prompt)\n",
       "</code></pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>GPTCelltype</em> version 1.0.1 ]</div>\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{gptcelltype}{Cell type annotation with OpenAI GPT models}{gptcelltype}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "Cell type annotation with OpenAI GPT models\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "gptcelltype(input, tissuename = NULL, model = \"gpt-4\", topgenenumber = 10)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{input}] Either the differential gene table returned by Seurat FindAllMarkers() function, or a list of genes.\n",
       "\n",
       "\\item[\\code{tissuename}] Optional input of tissue name.\n",
       "\n",
       "\\item[\\code{model}] A valid GPT-4 or GPT-3.5 model name list on https://platform.openai.com/docs/models. Default is 'gpt-4-32k'.\n",
       "\n",
       "\\item[\\code{topgenenumber}] Number of top differential genes to be used if input is Seurat differential genes.\n",
       "\n",
       "\\item[\\code{openai\\_key}] The OpenAI key obtained from https://platform.openai.com/account/api-keys The default is NA, which will resulting outputing the prompt itself. If an actual key is provided, then the output will be the celltype annotations from the GPT model specified by the user.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "Annotate cell types by OpenAI GPT models in a Seurat pipeline or with a custom gene list. If used in a Seurat pipeline, Seurat FindAllMarkers() function needs to be run first and the differential gene table generated by Seurat will serve as the input. If the input is a custom list of genes, one cell type is identified for each element in the list.  Note: system environment should have variable OPENAI\\_API\\_KEY = 'your\\_api\\_key' or ''. The OpenAI key can be obtained from https://platform.openai.com/account/api-keys. If '', then output the prompt itself. If an actual key is provided, then the output will be the celltype annotations from the GPT model specified by the user.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "A vector of cell types when the user provide openai\\_key or the prompt itself when openai\\_key = NA (default)\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Author}\\relax\n",
       "Wenpin Hou <wh2526@cumc.columbia.edu>, Zhicheng Ji <zhicheng.ji@duke.edu>\n",
       "\\end{Author}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "# Gene list as input, OPENAI_API_KEY = NA (default)\n",
       "Sys.setenv(OPENAI_API_KEY = '')\n",
       "print(Sys.getenv('OPENAI_API_KEY'))  \n",
       "gptcelltype(input = list(gs1=c('CD4,CD3D'),gs2='CD14'), tissuename=NULL, model='gpt-4', topgenenumber = 10) \n",
       "# Seurat object as input, openai_key = NA (default)\n",
       "#You have to install Seurat package first to run this example\n",
       "library(Seurat)\n",
       "data(\"pbmc_small\")\n",
       "all.markers <- FindAllMarkers(object = pbmc_small)\n",
       "Sys.setenv(OPENAI_API_KEY = '')\n",
       "print(Sys.getenv('OPENAI_API_KEY'))  \n",
       "prompt <- gptcelltype(all.markers, tissuename='human PBMC')\n",
       "return(prompt)\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "gptcelltype            package:GPTCelltype             R Documentation\n",
       "\n",
       "_\bC_\be_\bl_\bl _\bt_\by_\bp_\be _\ba_\bn_\bn_\bo_\bt_\ba_\bt_\bi_\bo_\bn _\bw_\bi_\bt_\bh _\bO_\bp_\be_\bn_\bA_\bI _\bG_\bP_\bT _\bm_\bo_\bd_\be_\bl_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     Cell type annotation with OpenAI GPT models\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     gptcelltype(input, tissuename = NULL, model = \"gpt-4\", topgenenumber = 10)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "   input: Either the differential gene table returned by Seurat\n",
       "          FindAllMarkers() function, or a list of genes.\n",
       "\n",
       "tissuename: Optional input of tissue name.\n",
       "\n",
       "   model: A valid GPT-4 or GPT-3.5 model name list on\n",
       "          https://platform.openai.com/docs/models. Default is\n",
       "          'gpt-4-32k'.\n",
       "\n",
       "topgenenumber: Number of top differential genes to be used if input is\n",
       "          Seurat differential genes.\n",
       "\n",
       "openai_key: The OpenAI key obtained from\n",
       "          https://platform.openai.com/account/api-keys The default is\n",
       "          NA, which will resulting outputing the prompt itself. If an\n",
       "          actual key is provided, then the output will be the celltype\n",
       "          annotations from the GPT model specified by the user.\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     Annotate cell types by OpenAI GPT models in a Seurat pipeline or\n",
       "     with a custom gene list. If used in a Seurat pipeline, Seurat\n",
       "     FindAllMarkers() function needs to be run first and the\n",
       "     differential gene table generated by Seurat will serve as the\n",
       "     input. If the input is a custom list of genes, one cell type is\n",
       "     identified for each element in the list.  Note: system environment\n",
       "     should have variable OPENAI_API_KEY = 'your_api_key' or ''. The\n",
       "     OpenAI key can be obtained from\n",
       "     https://platform.openai.com/account/api-keys. If '', then output\n",
       "     the prompt itself. If an actual key is provided, then the output\n",
       "     will be the celltype annotations from the GPT model specified by\n",
       "     the user.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A vector of cell types when the user provide openai_key or the\n",
       "     prompt itself when openai_key = NA (default)\n",
       "\n",
       "_\bA_\bu_\bt_\bh_\bo_\br(_\bs):\n",
       "\n",
       "     Wenpin Hou <wh2526@cumc.columbia.edu>, Zhicheng Ji\n",
       "     <zhicheng.ji@duke.edu>\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     # Gene list as input, OPENAI_API_KEY = NA (default)\n",
       "     Sys.setenv(OPENAI_API_KEY = '')\n",
       "     print(Sys.getenv('OPENAI_API_KEY'))  \n",
       "     gptcelltype(input = list(gs1=c('CD4,CD3D'),gs2='CD14'), tissuename=NULL, model='gpt-4', topgenenumber = 10) \n",
       "     # Seurat object as input, openai_key = NA (default)\n",
       "     #You have to install Seurat package first to run this example\n",
       "     library(Seurat)\n",
       "     data(\"pbmc_small\")\n",
       "     all.markers <- FindAllMarkers(object = pbmc_small)\n",
       "     Sys.setenv(OPENAI_API_KEY = '')\n",
       "     print(Sys.getenv('OPENAI_API_KEY'))  \n",
       "     prompt <- gptcelltype(all.markers, tissuename='human PBMC')\n",
       "     return(prompt)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?gptcelltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793d5d40-f932-471b-8d94-26be9b657cc5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
